<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="group and sample level comparisons for single cell data with mixed effects models ">
<title>Sample-level Single-cell Generalized Linear Multilevel Models in R  • scglmmr</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Sample-level Single-cell Generalized Linear Multilevel Models in R ">
<meta property="og:description" content="group and sample level comparisons for single cell data with mixed effects models ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">scglmmr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/networks.html">Relationships between cell subsets with latent information correlation network analysis</a>
    <a class="dropdown-item" href="articles/pseudobulk_mixed_effects.html">Analysis of cohort perturbation experiments including nested group designs - statistical models to multicell correlation networks</a>
    <a class="dropdown-item" href="articles/scglmmr_singlecell.html">Single cell gene and module differential expression with nested group repeated measures experiment designs</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MattPM/scglmmr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><p>sample-level single cell generalized linear mixed models and phenotype networks in R ================</p>
<!-- README.md is generated from README.Rmd. Please edit that file -->
<p><a href="https://zenodo.org/badge/latestdoi/330215343" class="external-link"><img src="https://zenodo.org/badge/330215343.svg" alt="DOI"></a></p>
<p>This R package supports analysis of single cell data from multi-subject experiments. The main use case is for <strong>testing <em>within</em> each cell type / cluster</strong>, the effects of treatment, differences in treatment effect between groups, baseline differences between groups, and intrinsic effects such as age or sex on gene expression variation. After multivariate statistical models define these interpretable effects <strong>relationships between cell types</strong> can be defined using multi cell correlation network based approaches also provided here. One can use these approaches in isolation or together to interpret the effect of perturbation on the “wiring” of the system being profiled. These networks and phenotypes derived are interpretable as they are based on robust multivariate and mixed effects models that can flexibly adapt to different experiment designs.</p>
<p><em>under active development</em><br><strong>Examples of manuscripts using this package:</strong><br><a href="https://www.medrxiv.org/content/10.1101/2023.03.20.23287474v1" class="external-link">Mulè et al <em>Immunity</em> (2024 in press)</a><br><a href="https://www.biorxiv.org/content/10.1101/2022.06.05.494592v1.full" class="external-link">Mulè et al <em>Biorxiv</em> (2022)</a></p>
<p>An overview of methods provided below-see vignettes for more details.</p>
<div class="section level3">
<h3 id="id_1-aggregated-models-with-correlation-network-analysis-of-effects-across-cell-types">1. Aggregated models with correlation network analysis of effects across cell types<a class="anchor" aria-label="anchor" href="#id_1-aggregated-models-with-correlation-network-analysis-of-effects-across-cell-types"></a>
</h3>
<p align="center">
<img src="reference/figures/scglmmr_workflow_pb.png"></p>
<p>Overview of the workflow starting from a single cell data object. For experiments with repeated measures from the same individuals, mixed effects models are used to account for non independence of repeated measures from the same donor. For experiments with a single measure for each subject, fixed effects models are used. The statistical effect size of covariates or contrasts from these models can be extracted for gene set enrichment using <code>fgsea</code>. The sample level scores for expression of the leading edge genes driving each enriched signal are then calculated and can be used for a correlation network analysis. The network correlation methods add fdr control and additional adjustments for gene sharing between leading edge genes from different signatures within the same cell types.</p>
<p>This workflow uses wrapper functions around well supported methods created by other groups. These should be cited as outlined below:</p>
<p>Cite <a href="https://doi.org/10.1093/bioinformatics/btaa687" class="external-link">Hoffman et al <em>Bioinformatics</em> 2020</a> is using the function <code><a href="reference/FitDream.html">FitDream()</a></code>, a wrapper around the method <code>dream</code>. This method implements mixed effects models for transcriptome data by incorporating observational weights and empirical Bayes methods tailored to models fit with <a href="https://www.jstatsoft.org/article/view/v067i01" class="external-link">lme4</a>.</p>
<p>Cite <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29" class="external-link">Law et al <em>Genome Biology</em> 2014</a> if using the function <code>RunVoomLimma</code>, a wrapper around the fixed effects model <code>Limma + voom</code> which introduced observational weights to allow normal-based methods for transcriptome count data.</p>
<p>Cite <a href="https://www.biorxiv.org/content/10.1101/060012v3" class="external-link">Korotkevich et al <em>Biorxiv</em> 2021</a> if using the function <code><a href="reference/FgseaList.html">FgseaList()</a></code>, a wrapper around <code>fgsea</code>, a novel Monte Carlo algorithm for enrichment with exact statistics.</p>
</div>
<div class="section level3">
<h3 id="id_2-single-cell-gene-or-module-level-models">2. Single cell gene or module level models<a class="anchor" aria-label="anchor" href="#id_2-single-cell-gene-or-module-level-models"></a>
</h3>
<p>A set of functions test single cell level module activity scores to derive treatment effects, treatment effect differences between groups and baseline differences within cell types. As above a random intercept term for donor accommodates non-independence of expression for repeated timepoints from each subject. A gene level single cell Poisson model is also provided. A custom function for experiments with a single perturbation applied to donors nested in 2 groups, e.g. responders and non-responders, is also provided for this common experiment design.</p>
</div>
<div class="section level3">
<h3 id="id_3-enrichment-and-visualization-testing">3. Enrichment and visualization testing<a class="anchor" aria-label="anchor" href="#id_3-enrichment-and-visualization-testing"></a>
</h3>
<p>Several functions are provided for interacting with model fits, extracting effects of interest from all cell types simultaneously, comparing cell type specific or shared effects, creating lists of gene signatures, and visualizing results.</p>
</div>
<div class="section level3">
<h3 id="purpose--philosophy">Purpose &amp; philosophy<a class="anchor" aria-label="anchor" href="#purpose--philosophy"></a>
</h3>
<p>Multi sample / subject experiments from human cohorts including many individuals pre and post treatment nested into different ‘response’ groups or treated with different drug perturbations are becoming more common. Common workflows for single cell analysis focus on comparing differences between cell type clusters and do not provide support for complex highly nested data. The approach outlined hereenables integrating human population variations with inference about differences due to group status or treatment effects and single cell variations. For example, assumptions about independence made by many methods methods that do not adjust for repeated measures (cells) from the same donor are violated with complex nesting. This package supports this complex structure. Consider the structure and effects shown below. Instead of ignoring or removing subject to subject variations, multivariate mixed effects models enable integration and quantification of human population variation inherent in these multi-subject experiments. The effects of age, sex, ethnicity can all be extracted and analyzed for enrichment of biological pathways. In addition, by using correlation network analysis and a shared latent information metric one can understand relationships between cell phenotypes across the entire system profiled.</p>
<p><img src="reference/figures/scglmmr.overview.png" width="750" height="550"></p>
</div>
<div class="section level3">
<h3 id="additional-details">Additional details<a class="anchor" aria-label="anchor" href="#additional-details"></a>
</h3>
<p>We model expression within each cluster/ cell type independently in order to capture perturbation effects of cell type specific genes as well as genes that are expressed by multiple cell types. This is necessary in order to implement linear methods which first model the mean variance trend after filtering out genes with near 0 expression across all individuals for a given cell type. Transcripts are therefore tested for perturbation effects within only in the cell types that express the genes. Enrichment and fdr testing is then done within cell type enabling the extraction of cell type specific signatures of perturbation response. The package provides simple methods for extracting such signatures.</p>
<p>The approach in this package is particularly well suited for multimodal single cell data where separate interpretable information (for example surface proteins with CITE-seq) are first used for cell type classification, while mRNA within these defined clusters can be analyzed for perturbation effects. A workflow we have used in many manuscripts includes denoising and normalizing ADT data using <a href="https://github.com/niaid/dsb" class="external-link">dsb</a>, clustering cells based on only normalized protein levels, then using the workflows described in this package to integrate human population variation with transcriptome variations, inferring e.g. drug treatment effects within each protein based subset.</p>
<!-- badges: start -->
<!-- badges: end -->
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/MattPM/scglmmr/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/MattPM/scglmmr/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>NA</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing scglmmr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Matthew Mulè <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-8457-2716" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Matthew Mulè.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
