<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="group and sample level comparisons for single cell data with multlevel models ">
<title>Sample-level Single-cell Generalized Linear Multilevel Models in R  • scglmmr</title>
<script src="deps/jquery-3.5.1/jquery-3.5.1.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Sample-level Single-cell Generalized Linear Multilevel Models in R ">
<meta property="og:description" content="group and sample level comparisons for single cell data with multlevel models ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">scglmmr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/pseudobulk_mixed_effects.html">nested group repeated measures pseudobulk differential expression testing</a>
    <a class="dropdown-item" href="articles/scglmmr_pseudobulk.html">nested group repeated measures pseudobulk differential expression testing</a>
    <a class="dropdown-item" href="articles/scglmmr_singlecell.html">Single cell gene and module level differential expression with nested group repeated measures experiment designs using scglmmr</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MattPM/scglmmr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><p>scglmmr <em>S</em>ample-level Single-<em>C</em>ell
<em>G</em>eneralized <em>L</em>inear <em>M</em>ultilevel <em>M</em>odels
in <em>R</em> ================</p>
<!-- README.md is generated from README.Rmd. Please edit that file -->
<p>An R package for implementing mixed effects models on single cell
data with complex experiment designs. The package is flexible and can
accomodate many experiment designs. It was developed for analysis of
multimodal single cell data from many individuals assayed pre and post
perturbation such as drug treatment, where each individual is nested
within one or more response groups. The methods herein allow one to
compare the difference in perturbation response effects between groups
while modeling variation in donor expression. It also has many wrappers
for downstream enrichment testing and visualization.</p>
<p>Please see vignettes.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/devtools/man/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span>repo <span class="op">=</span> <span class="st">"https://github.com/MattPM/scglmmr"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MattPM/scglmmr" class="external-link">scglmmr</a></span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/scglmmr.overview.png"></p>
<p><strong>With this type of experiment design, we can’t just color umap
plots and try to find the effects.</strong> We need statistical
models.</p>
</div>
<div class="section level2">
<h2 id="single-cell-within-cluster-perturbation-response-differential-expression">Single cell within cluster perturbation response differential
expression<a class="anchor" aria-label="anchor" href="#single-cell-within-cluster-perturbation-response-differential-expression"></a>
</h2>
<p>The purpose of this software is to analyze single cell genomics data
with pre and post perturbation measurements from the same individuals,
including complex designs wehre individuals with repeated measurements
are nested within in multiple response groups. The focus is on
implementing flexible generalized linear multilevel models to derive
group (i.e. good or poor clinical outcome, high or low rug response) and
treatment associated effects <em>within cell types</em> defined either
by protein (e.g. with CITE-seq data) or transcriptome based clustering
followed by downstream enrichment testing and visualization.</p>
<p>By default, the effect of treatment/perturbation across all subjects,
the baseline differences between outcome groups, and the difference in
the treatment effect between the outcome groups are tested. Any number
of model covariates can be specified and by default the package uses a
random intercept model to accomodate the non-independence of expression
within each subject.</p>
<p>An overview of methods provided:</p>
<div class="section level3">
<h3 id="id_1-pseudobulk-aggregated-models">1. pseudobulk aggregated models<a class="anchor" aria-label="anchor" href="#id_1-pseudobulk-aggregated-models"></a>
</h3>
<p>These functions implement wrappers around <code>limma</code> for
fitting fixed effect linear models and the <code>dream</code> method
from the <code>variancePartition</code> package. The dream method is the
only way to test differential expression while accomodating ‘random’ or
varying effects. This is statistically necessary to account for
non-independence when we have perturbation experiments with repeated
measurements from the same donors. To enable linear models
(e.g. modeling the mean with a normal distribution) to be fit to gene
counts, dream accounts for the mean variance trend via incorporating
<code>voom</code> observational weights.</p>
</div>
<div class="section level3">
<h3 id="id_2-single-cell-gene-and-gene-module-level-mixed-effects-models">2. single cell gene and gene module level mixed effects models<a class="anchor" aria-label="anchor" href="#id_2-single-cell-gene-and-gene-module-level-mixed-effects-models"></a>
</h3>
<p>Test perturbation effect using a gene level Poisson mixed model.</p>
</div>
<div class="section level3">
<h3 id="id_3-single-cell-module-level-mixed-effect-models">3. single cell module level mixed effect models<a class="anchor" aria-label="anchor" href="#id_3-single-cell-module-level-mixed-effect-models"></a>
</h3>
<p>Test perturbation effects and differences in perturbation responses
between groups at the gene module level.</p>
</div>
<div class="section level3">
<h3 id="id_4-downstream-enrichment-testing-and-visualization">4. Downstream enrichment testing and visualization<a class="anchor" aria-label="anchor" href="#id_4-downstream-enrichment-testing-and-visualization"></a>
</h3>
<p>Wrappers around methods from the <a href="https://www.biorxiv.org/content/10.1101/060012v2#disqus_thread" class="external-link">fast
set gene enrichment (fgsea)</a> and <a href="https://www.bioconductor.org/packages/release/bioc/html/clusterProfiler.html" class="external-link">clusterProfiler</a>
R packages.</p>
</div>
<div class="section level3">
<h3 id="philosophy">Philosophy<a class="anchor" aria-label="anchor" href="#philosophy"></a>
</h3>
<p>The scglmmr package considers each cluster/ cell type as a separate
‘experiment’ similar to fitting separate models to different FACS sorted
leukocyte subsets followed by RNAseq. Fitting models separately to each
subset provides maximum flexibility and avoids issues with e.g. modeling
mean variance trends or count distributions for cell type specific genes
in subsets that do not express the gene while still enabling comparison
of, for example, coherent perturbation effects for the same gene across
individuals between different cell clusters. This approach is
particularly well suited for CITE-seq data with cells clustered based on
normalized protein expression levels. Typically our workflow consists of
denoising ADT data using our method <a href="https://github.com/niaid/dsb" class="external-link">dsb</a> followed by modeling the
group level perturbation response effects using scglmmr.</p>
<p><strong>Experiment designs (within each cluster / celltype) supported
by scglmmr</strong> Below is a 2 group repeated measures experiment.
This data can be accomodated by scglmmr. More simple experiment designs
are also supported, for example data with 2 groups but not repeated
pre/post treatment measurements.</p>
<table class="table">
<colgroup>
<col width="20%">
<col width="20%">
<col width="20%">
<col width="20%">
<col width="20%">
</colgroup>
<thead><tr class="header">
<th align="left">sample</th>
<th align="center">sampleid</th>
<th align="right">timepoint</th>
<th align="left">Group</th>
<th align="center">sex</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">101_t0</td>
<td align="center">101</td>
<td align="right">d0</td>
<td align="left">good</td>
<td align="center">F</td>
</tr>
<tr class="even">
<td align="left">101_t1</td>
<td align="center">101</td>
<td align="right">d1</td>
<td align="left">poor</td>
<td align="center">F</td>
</tr>
<tr class="odd">
<td align="left">102_t0</td>
<td align="center">102</td>
<td align="right">d0</td>
<td align="left">good</td>
<td align="center">M</td>
</tr>
<tr class="even">
<td align="left">102_t1</td>
<td align="center">102</td>
<td align="right">d1</td>
<td align="left">poor</td>
<td align="center">M</td>
</tr>
<tr class="odd">
<td align="left">… (x n donors)</td>
<td align="center">… (x n donors)</td>
<td align="right">… (x n donors)</td>
<td align="left">… (x n donors)</td>
<td align="center">… (x n donors)</td>
</tr>
</tbody>
</table>
<!-- badges: start --><!-- badges: end --><p>Questions? Pls open an issue.</p>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/MattPM/scglmmr/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/MattPM/scglmmr/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>NA</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing scglmmr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Matthew Mulè <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-8457-2716" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Matthew Mulè.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
