<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="group and sample level comparisons for single cell data with mixed effects models ">
<title>Sample-level Single-cell Generalized Linear Multilevel Models in R  • scglmmr</title>
<script src="deps/jquery-3.5.1/jquery-3.5.1.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Sample-level Single-cell Generalized Linear Multilevel Models in R ">
<meta property="og:description" content="group and sample level comparisons for single cell data with mixed effects models ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">scglmmr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/pseudobulk_mixed_effects.html">Pseudobulk differential expression with nested group repeated measures single cell experiment designs</a>
    <a class="dropdown-item" href="articles/scglmmr_singlecell.html">Single cell gene and module differential expression with nested group repeated measures experiment designs</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MattPM/scglmmr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="sample-level-single-cell-glmms-in-r">Sample-level Single Cell GLMMs in R<a class="anchor" aria-label="anchor" href="#sample-level-single-cell-glmms-in-r"></a>
</h1></div>
<!-- README.md is generated from README.Rmd. Please edit that file -->
<p><a href="https://zenodo.org/badge/latestdoi/330215343" class="external-link"><img src="https://zenodo.org/badge/330215343.svg" alt="DOI"></a></p>
<p><strong>This package is under active development</strong></p>
<p>An R package for analysis of single cell data collected on many individuals. The package implements both fixed and mixed effects models and can accomodate many different complex experiment designs.</p>
<p>This package is built around <a href="https://www.jstatsoft.org/article/view/v067i01" class="external-link">lme4</a> and was originally made for analysis of single cell data from multiple subjects each assayed pre- and post- perturbation such as drug treatment, nested within one or more response groups. Other simpler designs are also supported, such as mutliple individuals in different groups, e.g. healthy vs disease.</p>
<p>Mixed effects models enable contrasting the difference in perturbation response effects between groups while modeling variation in donor expression. This package also has many wrappers for downstream enrichment testing and visualization.</p>
<p>Please see vignettes</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/devtools/man/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span>repo <span class="op">=</span> <span class="st">"https://github.com/MattPM/scglmmr"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MattPM/scglmmr" class="external-link">scglmmr</a></span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/scglmmr.overview.png"></p>
</div>
<div class="section level2">
<h2 id="single-cell-within-cluster-perturbation-response-differential-expression">Single cell within cluster perturbation response differential expression<a class="anchor" aria-label="anchor" href="#single-cell-within-cluster-perturbation-response-differential-expression"></a>
</h2>
<p>The focus is on implementing flexible generalized linear multilevel models to derive group (i.e. good or poor clinical outcome, high or low rug response) and treatment associated effects <em>within cell types</em> defined either by protein (e.g. with CITE-seq data) or transcriptome based clustering followed by downstream enrichment testing and visualization.</p>
<p>Any number of model covariates can be specified. The vignettes provide methods where a random intercept term for the donor ID of each cell or aggregated library is incluided in the model. These methods model variation around the average baseline expression across individuals, accomodating non-independence of expression for repeated timepoints from each subject.</p>
<p>An overview of methods provided:</p>
<div class="section level3">
<h3 id="id_1-pseudobulk-aggregated-models">1. pseudobulk aggregated models<a class="anchor" aria-label="anchor" href="#id_1-pseudobulk-aggregated-models"></a>
</h3>
<p>These functions implement wrappers around the following methods:</p>
<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29" class="external-link"><strong>Limma + voom (Law et al <em>Genome Biology</em> 2014)</strong></a> - fitting fixed effect linear models.</p>
<p><a href="https://doi.org/10.1093/bioinformatics/btaa687" class="external-link"><strong>‘dream’ – lme4 + voom”, described in Hoffman et al <em>Bioinformatics</em> 2020</strong></a> for experiments with non-independence due to replicated measurements (e.g. multiple timepoints) from the same donor.</p>
</div>
<div class="section level3">
<h3 id="id_2-single-cell-gene-and-gene-module-level-mixed-effects-models">2. single cell gene and gene module level mixed effects models<a class="anchor" aria-label="anchor" href="#id_2-single-cell-gene-and-gene-module-level-mixed-effects-models"></a>
</h3>
<p>Model variation in gene counts at the single cell and sample / donor level.</p>
<p>Rather than aggregating data at the sample level in to average or summed counts, this uses a Piosson generalized linear mixed effects model with a log link function to test the effect of perturbation at the single cell level across a cohort of individuals, incorporating random effects for replicated sampels from the same donor.</p>
</div>
<div class="section level3">
<h3 id="id_3-single-cell-module-level-mixed-effect-models">3. single cell module level mixed effect models<a class="anchor" aria-label="anchor" href="#id_3-single-cell-module-level-mixed-effect-models"></a>
</h3>
<p>Model variation of gene module activity at the single cell and sample / donor level</p>
<p>This set of functions scores the activity of modules in single cells, then tests the effect perturbation across donors using a linear mixed modethe single cell level. Any model can be specified. In addition, a custom function is provided for a experiment with a single perturbation applied to donors nested in 2 response groups, due to the commonality of this experiment design in biological studies.</p>
</div>
<div class="section level3">
<h3 id="id_4-downstream-enrichment-testing-and-visualization">4. Downstream enrichment testing and visualization<a class="anchor" aria-label="anchor" href="#id_4-downstream-enrichment-testing-and-visualization"></a>
</h3>
<p>There are wrapper functions around multiple gene set enrichment methods, with emphasis on the <a href="https://www.biorxiv.org/content/10.1101/060012v2#disqus_thread" class="external-link">fast set gene enrichment (fgsea) package</a>. The results from fgsea can then be further interrogated by methods for contrasting information content in genes driving enrichments within and between celltypes. Multiple visualization wrappers are also provided.</p>
</div>
<div class="section level3">
<h3 id="philosophy">Philosophy<a class="anchor" aria-label="anchor" href="#philosophy"></a>
</h3>
<p>This package models expression within each cluster/ cell type independently in order to capture perturbation effects of cell type specific genes as well as genes that are expressed by multiple cell types. Using a normal distribution on count data requires first modeling the mean variance trend <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29" class="external-link">(see Law et al)</a> this requires filtering features (genes) that are not expressed by each cell type. These cell type specific transcripts are therefore tested for perturbation effects within only in the cell types that express the genes, instead of across all cell types. Genes that are shared across cell types can be conpared for coherent perturbation effects across all subjects or between different groups of subjects using contrast coding.</p>
<p>This approach is particularly well suited for multimodal single cell data where cells are clustered based a independent information from the perturbation effects. For example, we have utilized methods in this package for CITE-seq data where we first denoise ADT data using <a href="https://github.com/niaid/dsb" class="external-link">dsb</a>, cluster the cells based on the normalized /denoised protein levels, then use this package to model transcriptome within each subset for effects of pertuebation, group status or both using mixed effects models at the single cell or pseudobulk level.</p>
<!-- badges: start -->
<!-- badges: end -->
<p>Questions? Please open an issue.</p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/MattPM/scglmmr/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/MattPM/scglmmr/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>NA</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing scglmmr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Matthew Mulè <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-8457-2716" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Matthew Mulè.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
