<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Sample-level Single-cell Generalized Linear Multilevel Models in R  • scglmmr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Sample-level Single-cell Generalized Linear Multilevel Models in R ">
<meta property="og:description" content="group and sample level comparisons for single cell data with mixed effects models ">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">scglmmr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/pseudobulk_mixed_effects.html">Pseudobulk differential expression with nested group repeated measures single cell experiment designs</a>
    </li>
    <li>
      <a href="articles/scglmmr_singlecell.html">Single cell gene and module differential expression with nested group repeated measures experiment designs</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MattPM/scglmmr">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="sample-level-single-cell-glmms-in-r" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#sample-level-single-cell-glmms-in-r" class="anchor"></a>Sample-level Single Cell GLMMs in R</h1></div>
<!-- README.md is generated from README.Rmd. Please edit that file -->
<p><a href="https://zenodo.org/badge/latestdoi/330215343"><img src="https://zenodo.org/badge/330215343.svg" alt="DOI"></a></p>
<p><strong>This package is under active development</strong></p>
<p>An R package for analysis of single cell data collected on many
individuals. The package implements both fixed and mixed effects models
and can accomodate many different complex experiment designs.</p>
<p>This package is built around <a href="https://www.jstatsoft.org/article/view/v067i01">lme4</a> and was
originally made for analysis of single cell data from multiple subjects
each assayed pre- and post- perturbation such as drug treatment, nested
within one or more response groups. Other simpler designs are also
supported, such as mutliple individuals in different groups,
e.g. healthy vs disease.</p>
<p>Mixed effects models enable contrasting the difference in
perturbation response effects between groups while modeling variation in
donor expression. This package also has many wrappers for downstream
enrichment testing and visualization.</p>
<p>Please see vignettes</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="at">repo =</span> <span class="st">"https://github.com/MattPM/scglmmr"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scglmmr)</span></code></pre></div>
<p><img src="reference/figures/scglmmr.overview.png"></p>
</div>
<div id="single-cell-within-cluster-perturbation-response-differential-expression" class="section level2">
<h2 class="hasAnchor">
<a href="#single-cell-within-cluster-perturbation-response-differential-expression" class="anchor"></a>Single cell within cluster perturbation response differential
expression</h2>
<p>The focus is on implementing flexible generalized linear multilevel
models to derive group (i.e. good or poor clinical outcome, high or low
rug response) and treatment associated effects <em>within cell
types</em> defined either by protein (e.g. with CITE-seq data) or
transcriptome based clustering followed by downstream enrichment testing
and visualization.</p>
<p>Any number of model covariates can be specified. The vignettes
provide methods where a random intercept term for the donor ID of each
cell or aggregated library is incluided in the model. These methods
model variation around the average baseline expression across
individuals, accomodating non-independence of expression for repeated
timepoints from each subject.</p>
<p>An overview of methods provided:</p>
<div id="id_1-pseudobulk-aggregated-models" class="section level3">
<h3 class="hasAnchor">
<a href="#id_1-pseudobulk-aggregated-models" class="anchor"></a>1. pseudobulk aggregated models</h3>
<p>These functions implement wrappers around the following methods:</p>
<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29"><strong>Limma
+ voom (Law et al <em>Genome Biology</em> 2014)</strong></a> - fitting
fixed effect linear models.</p>
<p><a href="https://doi.org/10.1093/bioinformatics/btaa687"><strong>‘dream’ –
lme4 + voom”, described in Hoffman et al <em>Bioinformatics</em>
2020</strong></a> for experiments with non-independence due to
replicated measurements (e.g. multiple timepoints) from the same
donor.</p>
</div>
<div id="id_2-single-cell-gene-and-gene-module-level-mixed-effects-models" class="section level3">
<h3 class="hasAnchor">
<a href="#id_2-single-cell-gene-and-gene-module-level-mixed-effects-models" class="anchor"></a>2. single cell gene and gene module level mixed effects models</h3>
<p>Model variation in gene counts at the single cell and sample / donor
level.</p>
<p>Rather than aggregating data at the sample level in to average or
summed counts, this uses a Piosson generalized linear mixed effects
model with a log link function to test the effect of perturbation at the
single cell level across a cohort of individuals, incorporating random
effects for replicated sampels from the same donor.</p>
</div>
<div id="id_3-single-cell-module-level-mixed-effect-models" class="section level3">
<h3 class="hasAnchor">
<a href="#id_3-single-cell-module-level-mixed-effect-models" class="anchor"></a>3. single cell module level mixed effect models</h3>
<p>Model variation of gene module activity at the single cell and sample
/ donor level</p>
<p>This set of functions scores the activity of modules in single cells,
then tests the effect perturbation across donors using a linear mixed
modethe single cell level. Any model can be specified. In addition, a
custom function is provided for a experiment with a single perturbation
applied to donors nested in 2 response groups, due to the commonality of
this experiment design in biological studies.</p>
</div>
<div id="id_4-downstream-enrichment-testing-and-visualization" class="section level3">
<h3 class="hasAnchor">
<a href="#id_4-downstream-enrichment-testing-and-visualization" class="anchor"></a>4. Downstream enrichment testing and visualization</h3>
<p>There are wrapper functions around multiple gene set enrichment
methods, with emphasis on the <a href="https://www.biorxiv.org/content/10.1101/060012v2#disqus_thread">fast
set gene enrichment (fgsea) package</a>. The results from fgsea can then
be further interrogated by methods for contrasting information content
in genes driving enrichments within and between celltypes. Multiple
visualization wrappers are also provided.</p>
</div>
<div id="philosophy" class="section level3">
<h3 class="hasAnchor">
<a href="#philosophy" class="anchor"></a>Philosophy</h3>
<p>This package models expression within each cluster/ cell type
independently in order to capture perturbation effects of cell type
specific genes as well as genes that are expressed by multiple cell
types. Using a normal distribution on count data requires first modeling
the mean variance trend <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29">(see
Law et al)</a> this requires filtering features (genes) that are not
expressed by each cell type. These cell type specific transcripts are
therefore tested for perturbation effects within only in the cell types
that express the genes, instead of across all cell types. Genes that are
shared across cell types can be conpared for coherent perturbation
effects across all subjects or between different groups of subjects
using contrast coding.</p>
<p>This approach is particularly well suited for multimodal single cell
data where cells are clustered based a independent information from the
perturbation effects. For example, we have utilized methods in this
package for CITE-seq data where we first denoise ADT data using <a href="https://github.com/niaid/dsb">dsb</a>, cluster the cells based on
the normalized /denoised protein levels, then use this package to model
transcriptome within each subset for effects of pertuebation, group
status or both using mixed effects models at the single cell or
pseudobulk level.</p>
<!-- badges: start -->
<!-- badges: end -->
<p>Questions? Please open an issue.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/MattPM/scglmmr">https://​github.com/​MattPM/​scglmmr</a>
</li>
<li>Report a bug at <br><a href="https://github.com/MattPM/scglmmr/issues">https://​github.com/​MattPM/​scglmmr/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>NA</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Matthew Mulè <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-8457-2716" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID"></a> </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Matthew Mulè.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
