<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Pseudobulk differential expression with nested group repeated measures single cell experiment designs • scglmmr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Pseudobulk differential expression with nested group repeated measures single cell experiment designs">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scglmmr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/pseudobulk_mixed_effects.html">Pseudobulk differential expression with nested group repeated measures single cell experiment designs</a>
    </li>
    <li>
      <a href="../articles/scglmmr_singlecell.html">Single cell gene and module differential expression with nested group repeated measures experiment designs</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MattPM/scglmmr">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Pseudobulk differential expression with nested group repeated
measures single cell experiment designs</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MattPM/scglmmr/blob/master/vignettes/pseudobulk_mixed_effects.rmd"><code>vignettes/pseudobulk_mixed_effects.rmd</code></a></small>
      <div class="hidden name"><code>pseudobulk_mixed_effects.rmd</code></div>

    </div>

    
    
<div id="pseudobulk-mixed-effects-models" class="section level3">
<h3 class="hasAnchor">
<a href="#pseudobulk-mixed-effects-models" class="anchor"></a>1. pseudobulk mixed effects models</h3>
<p>These functions implement wrappers around limma for fitting fixed
effect linear models, and the <code>dream</code> method from the
<code>variancePartition</code> package forfitting mixed (i.e. varying)
effects models. Mixed models are necessary for experiments that have the
design of repeated measurements from the same donors in perturbation
studies in multi sample ‘random’ (varying) effects. This is necessary to
account for non-independence when we have perturbation experiments with
repeated measurements from the same donors. To enable linear models
(e.g. modeling the mean with a normal distribution) to be fit to gene
counts, <code><a href="https://rdrr.io/pkg/variancePartition/man/dream-method.html">variancePartition::dream</a></code> accounts for the mean
variance trend via incorporating voom observational weights. The fits
are then shrunken toward the genome wide trend using an empirical bayes
step adjusting for the per gene degrees of freedom estimated in the
mixed model fits. This approach is described in <a href="doi.org/10.1093/bioinformatics/btaa687">Hoffman et al
Bininformatics 2020</a> <strong>this paper should be cited if using the
wrapper below <code>FitDream</code>.</strong></p>
<p>In this workflow, mixed effect model fits and gene set enrichment
steps are parallelized with the BiocParallel package.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#devtools::install_github(repo = "https://github.com/MattPM/scglmmr")</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(scglmmr))</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(magrittr))</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BiocParallel)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co">#the mixed model fits and gsea are parallelized </span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>BiocParallel<span class="sc">::</span><span class="fu">register</span>(BiocParallel<span class="sc">::</span><span class="fu">SnowParam</span>(<span class="dv">4</span>))</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>pparam <span class="ot">=</span> BiocParallel<span class="sc">::</span><span class="fu">SnowParam</span>(<span class="at">workers =</span> <span class="dv">4</span>, <span class="at">type =</span> <span class="st">"SOCK"</span>, <span class="at">progressbar =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>Below analysis of a 2 group repeated measures experiment design is
shown. Pre-treatment baseline effect differences, treatment effets
across all donors and the difference in treatment effects are all
compared while modeling variation in baseline expression using a random
intercept term.</p>
<table class="table">
<colgroup>
<col width="21%">
<col width="18%">
<col width="18%">
<col width="21%">
<col width="18%">
</colgroup>
<thead><tr class="header">
<th align="left">sample</th>
<th align="center">sampleid</th>
<th align="right">time</th>
<th align="left">group</th>
<th align="center">sex</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">101_t0</td>
<td align="center">101</td>
<td align="right">d0</td>
<td align="left">low</td>
<td align="center">F</td>
</tr>
<tr class="even">
<td align="left">101_t1</td>
<td align="center">101</td>
<td align="right">d1</td>
<td align="left">high</td>
<td align="center">F</td>
</tr>
<tr class="odd">
<td align="left">102_t0</td>
<td align="center">102</td>
<td align="right">d0</td>
<td align="left">low</td>
<td align="center">M</td>
</tr>
<tr class="even">
<td align="left">102_t1</td>
<td align="center">102</td>
<td align="right">d1</td>
<td align="left">high</td>
<td align="center">M</td>
</tr>
<tr class="odd">
<td align="left">… (x n donors)</td>
<td align="center">… (x n donors)</td>
<td align="right">… (x n donors)</td>
<td align="left">… (x n donors)</td>
<td align="center">… (x n donors)</td>
</tr>
</tbody>
</table>
</div>
<div id="load-single-cell-data-aggregate-and-quaity-control" class="section level3">
<h3 class="hasAnchor">
<a href="#load-single-cell-data-aggregate-and-quaity-control" class="anchor"></a>Load single cell data, aggregate and quaity control</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>datapath <span class="ot">=</span> <span class="st">"mypath/"</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># load seurat or sce object etc. </span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>s <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">"path/seuratobject.rds"</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co"># define counts and metadata and subset to cells above rm seurat object from workspace </span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">=</span> s<span class="sc">@</span>meta.data</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>umi <span class="ot">=</span> s<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>counts</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(s); <span class="fu">gc</span>()</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="co"># QC contingency of cells by subject for each celltype </span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>tab <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">SubjectCelltypeTable</span>(<span class="at">metadata =</span> meta, <span class="at">celltype_column =</span> <span class="st">"celltype"</span>, <span class="at">sample_column =</span> <span class="st">"sample"</span>)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>tab<span class="sc">$</span>celltypes_remove; tab<span class="sc">$</span><span class="st">`</span><span class="at">low representation celltypes</span><span class="st">`</span>; tab<span class="sc">$</span>table</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="co"># remove cells prior to pseudobulk analysis </span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">=</span> meta[<span class="sc">!</span>meta<span class="sc">$</span>celltype <span class="sc">%in%</span> tab<span class="sc">$</span>celltypes_remove, ]</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a><span class="co"># subset data </span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>umi <span class="ot">=</span> umi[ ,<span class="fu">rownames</span>(meta)]</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Create aggregated pseudobulk data</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>pb <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">PseudobulkList</span>(</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">rawcounts =</span> umi,</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">metadata =</span> meta, </span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample_col =</span> <span class="st">"sample"</span>,</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>  <span class="at">celltype_col =</span> <span class="st">"celltype"</span>,</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>  <span class="at">avg_or_sum =</span> <span class="st">"sum"</span></span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Create aggretated sample level metadata</span></span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>met <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">AggregateCellMetadata</span>(</span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>  <span class="at">cell.metadata =</span> meta, </span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample_column =</span> <span class="st">'sample'</span>, </span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a>  <span class="at">variable_columns =</span> <span class="fu">c</span>(<span class="st">'subjectid'</span>, <span class="st">'timepoint'</span>, <span class="st">'response'</span>, <span class="st">'age'</span>, <span class="st">'sex'</span>),</span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a>  <span class="at">pseudobulk.List =</span> pb</span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-40"><a href="#cb2-40" aria-hidden="true" tabindex="-1"></a><span class="co"># creation of a combined grouping factor indicating both timepoint and group</span></span>
<span id="cb2-41"><a href="#cb2-41" aria-hidden="true" tabindex="-1"></a><span class="co"># (e.g. t0_Group1, t0_group2) will make it simple to set up contrasts.</span></span>
<span id="cb2-42"><a href="#cb2-42" aria-hidden="true" tabindex="-1"></a><span class="co"># here group means 'response group' i.e. high vs low responder. </span></span>
<span id="cb2-43"><a href="#cb2-43" aria-hidden="true" tabindex="-1"></a><span class="co"># also convert other variables to factors and do some standard transformation</span></span>
<span id="cb2-44"><a href="#cb2-44" aria-hidden="true" tabindex="-1"></a><span class="co"># of metadata.</span></span>
<span id="cb2-45"><a href="#cb2-45" aria-hidden="true" tabindex="-1"></a>met<span class="sc">$</span>group.time <span class="ot">=</span> <span class="fu">paste</span>(met<span class="sc">$</span>group, met<span class="sc">$</span>timepoint, <span class="at">sep =</span> <span class="st">'_'</span>)</span>
<span id="cb2-46"><a href="#cb2-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-47"><a href="#cb2-47" aria-hidden="true" tabindex="-1"></a><span class="co"># make sure this is a factor and the levels are in a coherent order for the </span></span>
<span id="cb2-48"><a href="#cb2-48" aria-hidden="true" tabindex="-1"></a><span class="co"># contrast matrix -- see below. here: </span></span>
<span id="cb2-49"><a href="#cb2-49" aria-hidden="true" tabindex="-1"></a><span class="co"># time 0 = 0, time 1 = 1. </span></span>
<span id="cb2-50"><a href="#cb2-50" aria-hidden="true" tabindex="-1"></a><span class="co"># low response = 0 high response = 1. </span></span>
<span id="cb2-51"><a href="#cb2-51" aria-hidden="true" tabindex="-1"></a>met<span class="sc">$</span>group.time <span class="ot">=</span> <span class="fu">factor</span>(</span>
<span id="cb2-52"><a href="#cb2-52" aria-hidden="true" tabindex="-1"></a>  met<span class="sc">$</span>group.time,</span>
<span id="cb2-53"><a href="#cb2-53" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"1_0"</span>, <span class="st">"1_1"</span>, <span class="st">"0_0"</span>, <span class="st">"0_1"</span>)</span>
<span id="cb2-54"><a href="#cb2-54" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-55"><a href="#cb2-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-56"><a href="#cb2-56" aria-hidden="true" tabindex="-1"></a><span class="co"># now filter genes within each cell type that are reasonably expressed. </span></span>
<span id="cb2-57"><a href="#cb2-57" aria-hidden="true" tabindex="-1"></a>design <span class="ot">=</span> <span class="fu">model.matrix</span>( <span class="sc">~</span> <span class="dv">0</span> <span class="sc">+</span> met<span class="sc">$</span>group.time)</span>
<span id="cb2-58"><a href="#cb2-58" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">=</span> <span class="fu">Normalize</span>(<span class="at">pseudobulk.list =</span> pb, <span class="at">design =</span> design, <span class="at">minimum.gene.count =</span> <span class="dv">5</span>)</span></code></pre></div>
</div>
<div id="fit-models-and-specify-a-priori-contrasts-corresponding-to-the-desired-effect-comparisons" class="section level3">
<h3 class="hasAnchor">
<a href="#fit-models-and-specify-a-priori-contrasts-corresponding-to-the-desired-effect-comparisons" class="anchor"></a>Fit models and specify a priori contrasts corresponding to the
desired effect comparisons</h3>
<p>Below shows 2 steps. 1: Specify a varying intercept model using lme4
symbolic model formula. <a href="https://arxiv.org/pdf/1911.08628.pdfs">More information on
symbolic formula representations of models</a>. Most experiment designs
will have some covariate of interest such as time relative to
perturbation, other covariates which should be adjusted for (such as
batch or sex), and individuals measured with repeated timepoints will
have have a “subjectID” or “individual” variable specifying which person
the measurement came from. This variable will often modeled using a
varying intercept modeled with a normal distribution which is specified
with e.g. (1| subjectID). Other formula accepted by lme4 are also
possible.</p>
<p>In a simple experiment, e.g. a one way anova design with baseline and
post treatment and no different outcome groups, if the time relative to
treatment is the target we want estimates for, we don’t need a design
matrix, we can simply extract the effect of “time” from the lme4
fits.</p>
<p>In the example below, we want estimate baseline differences,
treatment effects across all individuals and the difference in fold
changes between groups adjusting estimates for age and sex (and modeling
the individual variation with a varying effect). To do this we specify a
custom a. priori contrast matrix.</p>
<p>2: Using the factor variable combining group and time, we create
contrasts over the levels to define effects of interest. The function
<code>makeContrastsDream</code> is used for this purpose which works the
same way as the limma function makeContrasts. More simple contrasts or
more complex contrasts are also possible. This step is critical for
deriving the correct effec size estimates. More information on
specifying contrast matrices is available here: <a href="https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html">A
guide to creating design matrices for gene expression
experiments</a></p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Now we  specify model </span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>f1 <span class="ot">&lt;-</span> <span class="er">~</span> <span class="dv">0</span> <span class="sc">+</span> age <span class="sc">+</span> sex <span class="sc">+</span> group.time <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>subjectid) </span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># make contrast matrix </span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>L2 <span class="ot">=</span> <span class="fu">makeContrastsDream</span>(</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">formula =</span> f1, </span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> samplemd,</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">contrasts =</span> <span class="fu">c</span>(</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">baseline =</span> <span class="st">"group.time1_0 - group.time0_0"</span>, <span class="co"># note fixef model also fit for this single time point contrast </span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">treatment_delta =</span> <span class="st">"( group.time1_1 - group.time1_0 ) - ( group.time0_1 - group.time0_0 )"</span>,</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">treatment =</span> <span class="st">"( group.time1_1 + group.time0_1 ) / 2 - ( group.time1_0 + group.time0_0 ) / 2 "</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="co"># visualize the contrast matrix, compare to levels of group.time variable</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a><span class="co"># to check for correctly specified effects. </span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="fu">plotContrasts</span>(L2) </span></code></pre></div>
<p>Fit the models with <code><a href="https://rdrr.io/pkg/variancePartition/man/dream-method.html">variancePartition::dream</a></code> which uses
voom weights in lme4 mixed effects models with an empirical bayes
shrinkage toward genome wide trend accounting for per gene degrees of
freedom.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>fit1 <span class="ot">=</span> <span class="fu">FitDream</span>(<span class="at">pb.list =</span> dge, </span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">sample.metadata =</span> met, </span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">lme4.formula =</span> f1,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">dream.contrast.matrix =</span> L2,</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">ncores =</span> <span class="dv">4</span>)</span></code></pre></div>
<p>Note, <code><a href="https://rdrr.io/pkg/variancePartition/man/dream-method.html">variancePartition::dream</a></code> now incorporates an
emperical Bayes step derived for mixed effects models accounting for per
gene degrees of freedom, please see: <a href="https://github.com/GabrielHoffman/variancePartition/issues/54" class="uri">https://github.com/GabrielHoffman/variancePartition/issues/54</a>.</p>
</div>
<div id="downstream-gene-set-enrichment-analysis-within-celltypes-for-different-effects" class="section level3">
<h3 class="hasAnchor">
<a href="#downstream-gene-set-enrichment-analysis-within-celltypes-for-different-effects" class="anchor"></a>Downstream gene set enrichment analysis within celltypes for
different effects</h3>
<p>Run gene set enrichment analysis within each cell type fbased on
genes ranked by effect size for each of the effects defined above.</p>
<p>Here within each cell type and for each contrast we run gene set
enrichment analysis using 2 functions <code>ExtractResult</code> and
<code>FgseaList</code>. <code>ExtractResult</code> is used to extract a
list of gene ranks (it can be used as a wrapper around dream::topTable
and limma::topTable to return a full list of results). Since we had
multiple covariates in a mixed model and we used a custom contrast
matrix, we specify the covariate of interest from the contrast using
arguments <code>coefficient.number</code> and <code>coef.name</code>
which are output in results based on the names of the contrasts.</p>
<p>With the list of ranks we then run gene set enrichment with the <a href="https://www.biorxiv.org/content/10.1101/060012v3">fgsea method by
Korotkevich et al</a> using the function <code>FseaList</code>.
Conveniently this also is parallelized using the same biocparallel</p>
<p>Based on our contrast matrix (the object <code>L2</code> we created
above with <code>makeContrastsDream</code>),
<code>coefficient.number = 1</code> corresponds to the baseline
difference between groups, <code>coefficient.number = 2</code> is the
difference in fold changes between the groups and
<code>coefficient.number = 3</code> is the pre vs post perturbation
difference across subjects in both groups. Estimates for the covariates
are also availabe.</p>
<div id="examples-of-extracting-gene-ranks-based-on-contrasts-and-running-gene-set-enrichment" class="section level4">
<h4 class="hasAnchor">
<a href="#examples-of-extracting-gene-ranks-based-on-contrasts-and-running-gene-set-enrichment" class="anchor"></a>Examples of extracting gene ranks based on contrasts and running
gene set enrichment</h4>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co"># msigDB hallmark pathways are included in the scglmmr package</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>hlmk <span class="ot">=</span> scglmmr<span class="sc">::</span>hallmark</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>scglmmr<span class="sc">::</span><span class="fu">ExtractResult</span>()</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co"># extract genes ranked by treatment effect (across all donors) </span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>rtreat <span class="ot">=</span> <span class="fu">ExtractResult</span>(<span class="at">model.fit.list =</span> fit1,</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>                       <span class="at">what =</span> <span class="st">'lmer.z.ranks'</span>,</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>                       <span class="at">coefficient.number =</span> <span class="dv">3</span>,</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>                       <span class="at">coef.name =</span> <span class="st">'treatment'</span>)</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="co"># run fgsea </span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>hlmk.treat <span class="ot">=</span> <span class="fu">FgseaList</span>(<span class="at">rank.list.celltype =</span> rtreat,</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>                       <span class="at">pathways =</span> hlmk,</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>                       <span class="at">BPPARAM =</span> pparam)</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a><span class="co"># extract gene ranks of treatment effect (across all donors) </span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>rdelta <span class="ot">=</span> <span class="fu">ExtractResult</span>(<span class="at">model.fit.list =</span> fit1,</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>                       <span class="at">what =</span> <span class="st">'lmer.z.ranks'</span>,</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>                       <span class="at">coefficient.number =</span> <span class="dv">2</span>,</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>                       <span class="at">coef.name =</span> <span class="st">'treatment_delta'</span>)</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>hlmk.delta <span class="ot">=</span> <span class="fu">RunFgseaOnRankList</span>(<span class="at">rank.list.celltype =</span> rdelta,</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>                                <span class="at">pathways =</span> hlmk,</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>                                <span class="at">BPPARAM =</span> pparam)</span></code></pre></div>
<p>On a technical note, it’s also possible to fit a simple model for the
baseline contrast not estimating the variation across subjects with a
random effect as this contrast is more typically estimated with a
standard group 1 vs 2 least squares approach. The function
<code>RunVoomLimma</code> is provided to fit these models. If you
compare the effect size estimates for an individual cell type for the
<code>coefficient.number = 1</code> from the mixed model fits (object
<code>fit1</code> above) to the models below they will be very highly
corrleated along the diagonal with some differences arising to the
different degrees of freedom, sample size and variance coming from the
additional layer of variation estimated by the mixed model. The choice
of model to use is up to the user.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># fit simple linear model for the baseline group level contrast </span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>design<span class="fl">.2</span> <span class="ot">=</span> <span class="fu">model.matrix</span>(<span class="sc">~</span><span class="dv">0</span> <span class="sc">+</span> age <span class="sc">+</span> sex <span class="sc">+</span> group.time, <span class="at">data =</span> met)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>fit0 <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">RunVoomLimma</span>(<span class="at">dgelists =</span> dge, </span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>                           <span class="at">design_matrix =</span> design<span class="fl">.2</span>, </span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>                           <span class="at">do_contrast_fit =</span> T,</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>                           <span class="co"># we use only the first row of the contrast matrix L2</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>                           <span class="at">my_contrast_matrix =</span> L2[ ,<span class="dv">1</span>])</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="co"># extract fixed efect model estimate of baseline contrast (pre treatment differences between groups)</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>r0 <span class="ot">=</span> <span class="fu">ExtractResult</span>(<span class="at">model.fit.list =</span> fit0,</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>                   <span class="at">what =</span> <span class="st">'gene.t.ranks'</span>,</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>                   <span class="at">coefficient.number =</span> <span class="dv">1</span>,</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>                   <span class="at">coef.name =</span> <span class="st">'baseline'</span>)</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a><span class="do">## run gene set enrichment </span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>hlmk<span class="fl">.0</span> <span class="ot">=</span> <span class="fu">FgseaList</span>(<span class="at">rank.list.celltype =</span> r0,</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>                   <span class="at">pathways =</span> hlmk,</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>                   <span class="at">BPPARAM =</span> pparam)</span></code></pre></div>
</div>
</div>
<div id="further-analysis-and-curation-of-enrichment-results" class="section level3">
<h3 class="hasAnchor">
<a href="#further-analysis-and-curation-of-enrichment-results" class="anchor"></a>Further analysis and curation of enrichment results</h3>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># full set of leading edge genes indexed by celltype x effect x module </span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>lefull <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">GetLeadingEdgeFull</span>(<span class="at">gsea.list =</span> hlmk.treat,</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">padj.filter =</span> <span class="fl">0.02</span>, </span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">NES.filter =</span> <span class="sc">-</span><span class="cn">Inf</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co"># extract model fit results instead of ranks</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co"># automatically this sets argument `result` to 'statistics'</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>fit1.res <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">ExtractResult</span>(<span class="at">model.fit.list =</span> fit1 , </span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">coefficient.number =</span> <span class="dv">3</span>, </span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">coef.name =</span> <span class="st">'treatment'</span>)</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="co"># combine GSEA results with model coefficient for each gene in leading edge </span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a><span class="co"># include all leading edge genes irrespective of individual gene p value </span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>cr <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">CombineResults</span>(<span class="at">gsealist =</span> hlmk.treat, </span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>                             <span class="at">contrastlist =</span> fit1.res, </span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>                             <span class="at">gseafdr =</span> <span class="fl">0.02</span>, </span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>                             <span class="at">genefdr =</span> <span class="dv">1</span>)</span></code></pre></div>
</div>
<div id="extract-all-leading-edge-genes-indexed-by-cell-type" class="section level3">
<h3 class="hasAnchor">
<a href="#extract-all-leading-edge-genes-indexed-by-cell-type" class="anchor"></a>Extract all leading edge genes indexed by cell type</h3>
<p>This outputs a nested list of leading edge genes from each enrichment
across cell types. list level 1 is cell type, level 2 is enrichment
signal.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>li <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">LeadingEdgeIndexed</span>(<span class="at">gsea.result.list =</span> hlmk.treat, <span class="at">padj.threshold =</span> <span class="fl">0.02</span>)</span></code></pre></div>
</div>
<div id="calculate-the-jaccard-similarity-of-the-leading-edge-genes-for-enrichments-within-a-given-cell-type-and-effect" class="section level3">
<h3 class="hasAnchor">
<a href="#calculate-the-jaccard-similarity-of-the-leading-edge-genes-for-enrichments-within-a-given-cell-type-and-effect" class="anchor"></a>Calculate the Jaccard similarity of the leading edge genes for
enrichments within a given cell type and effect</h3>
<p>This is useful for understanding which enrichment signals may come
from the same vs distinct genes.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># figpath.temp = here('figures')</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>treat.JI <span class="ot">=</span> <span class="fu">EnrichmentJaccard</span>(<span class="at">gsealist =</span> hlmk.treat, </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">indexedgenes =</span> li, </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>                          <span class="co">#saveplot = TRUE, </span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>                          <span class="co">#figpath = figpath.temp,</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>                          <span class="at">returnJaccardMtx =</span> <span class="cn">TRUE</span>)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co"># curate results </span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>results.sorted <span class="ot">=</span> treat.JI<span class="sc">$</span>sortedgsea <span class="sc">%&gt;%</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">signal =</span> <span class="fu">paste</span>(celltype, pathway, <span class="at">sep =</span> <span class="st">'~'</span>))</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="co"># data.table::fwrite(results.sorted, file = paste0(datapath, 'g0.result.sort.txt'),sep = "\t")</span></span></code></pre></div>
</div>
<div id="visualization-of-enrichment-results" class="section level3">
<h3 class="hasAnchor">
<a href="#visualization-of-enrichment-results" class="anchor"></a>Visualization of enrichment results</h3>
<p><strong>Create a bubble plot heatmap of enrichment results within
clusters</strong></p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># NES_filter to -Inf to plot positive and negative enrichment</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">=</span> <span class="fu">PlotFgsea</span>(<span class="at">gsea_result_list =</span> hlmk.treat, <span class="at">NES_filter =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">padj_filter =</span> <span class="fl">0.05</span>)</span></code></pre></div>
<p><strong>Create heatmap of gene perturbation fold changes across cell
subsets based on model fit coefficients</strong></p>
<p>Extract and make a heatmap of log fold changes across celltypes.
HeatmapDiag uses the package <a href="https://CRAN.R-project.org/package=slanter">slanter</a> which
accepts the same arguments as <code>pheatmap</code>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>gene.mat <span class="ot">=</span> <span class="fu">GetGeneMatrix</span>(<span class="at">result.list =</span> fit1, </span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>                         <span class="at">pvalfilter =</span> <span class="fl">0.05</span>, </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>                         <span class="at">stat_for_matrix =</span> <span class="st">'logFC'</span>,</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>                         <span class="at">logfcfilter =</span> <span class="fl">0.1</span>)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="co">#make heatmap  e.g. </span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>pheatmap<span class="sc">::</span><span class="fu">pheatmap</span>(gene.mat, <span class="at">fontsize_row =</span> <span class="dv">5</span>)</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a><span class="co"># diagnoalize the heatmap so that the most correlated signals are grouped </span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a><span class="co"># this uses the package slanter </span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="fu">HeatmapDiag</span>(<span class="at">matrix =</span> gene.mat, <span class="at">fontsize_row =</span> <span class="dv">5</span>)</span></code></pre></div>
<p><strong>Create heatmap of gene perturbation fold changes from
enrichments across individuals within a cell subset</strong></p>
<p>Visualize the log counts per million at the sample level of the gene
set enrichment results.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="co"># make tidy average data for visualization of weighted pb results </span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>lcmp <span class="ot">=</span> <span class="fu">lapply</span>(pb, edgeR<span class="sc">::</span>cpm, <span class="at">log =</span> <span class="cn">TRUE</span> )</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>le_expr <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">LeadEdgeTidySampleExprs</span>(<span class="at">av.exprs.list =</span> lcmp,</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">gsea.list =</span> hlmk.treat, </span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">padj.filter =</span> <span class="fl">0.1</span>,</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">NES.filter =</span> <span class="sc">-</span><span class="cn">Inf</span>)</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a><span class="co"># example plot of sample level average leading edge genes annotated </span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>scglmmr<span class="sc">::</span><span class="fu">LeadEdgeSampleHeatmap</span>(<span class="at">tidy.exprs.list =</span> le_expr,</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>                               <span class="at">modulename =</span> <span class="st">"HALLMARK_HYPOXIA"</span>,</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>                               <span class="at">elltype_plot =</span> <span class="st">"CD4_NaiveTcell"</span>,</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>                               <span class="at">metadata =</span> meta, </span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>                               <span class="at">metadata_annotate =</span> <span class="fu">c</span>(<span class="st">'group'</span>, <span class="st">'timepoint'</span>, <span class="st">'age'</span>, <span class="st">'sex'</span>),</span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>                               <span class="at">sample_column =</span> <span class="st">'sample'</span>,</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>                               <span class="at">returnmat =</span> <span class="cn">FALSE</span>, </span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>                               <span class="at">savepath =</span> figpath, </span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>                               <span class="at">savename =</span> <span class="st">"filename"</span>)</span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a><span class="co"># see also: </span></span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a><span class="co"># scglmmr::TopGenesTidySampleExprs() - same as aobve for 'top' de genes estimated by model coeffcient / p value. </span></span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a><span class="co"># scglmmr::GetTidySummary() - for custom plotting</span></span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a><span class="co"># sample level vsualization of average data above </span></span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a><span class="co"># repeat this for all enriched pathways </span></span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true" tabindex="-1"></a>heatpath <span class="ot">=</span> <span class="fu">here</span>(<span class="st">"sime/path"</span>); <span class="fu">dir.create</span>(heatpath)</span>
<span id="cb12-31"><a href="#cb12-31" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(le_expr)) {</span>
<span id="cb12-32"><a href="#cb12-32" aria-hidden="true" tabindex="-1"></a>  cdat <span class="ot">=</span> le_expr[[i]]</span>
<span id="cb12-33"><a href="#cb12-33" aria-hidden="true" tabindex="-1"></a>  ctype <span class="ot">=</span> <span class="fu">names</span>(le_expr[i])</span>
<span id="cb12-34"><a href="#cb12-34" aria-hidden="true" tabindex="-1"></a>  umod <span class="ot">=</span> <span class="fu">unique</span>(cdat<span class="sc">$</span>module)</span>
<span id="cb12-35"><a href="#cb12-35" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (u <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(umod)) {</span>
<span id="cb12-36"><a href="#cb12-36" aria-hidden="true" tabindex="-1"></a>    scglmmr<span class="sc">::</span><span class="fu">LeadEdgeSampleHeatmap</span>(<span class="at">tidy.exprs.list =</span> le_expr, </span>
<span id="cb12-37"><a href="#cb12-37" aria-hidden="true" tabindex="-1"></a>                          <span class="at">modulename =</span> umod[u], </span>
<span id="cb12-38"><a href="#cb12-38" aria-hidden="true" tabindex="-1"></a>                          <span class="at">celltype_plot =</span> ctype,</span>
<span id="cb12-39"><a href="#cb12-39" aria-hidden="true" tabindex="-1"></a>                          <span class="at">metadata =</span> meta, <span class="at">metadata_annotate =</span> <span class="fu">c</span>(<span class="st">'group'</span>, <span class="st">'timepoint'</span>, <span class="st">'age'</span>, <span class="st">'gender'</span>),</span>
<span id="cb12-40"><a href="#cb12-40" aria-hidden="true" tabindex="-1"></a>                          <span class="at">sample_column =</span> <span class="st">'sample'</span>,</span>
<span id="cb12-41"><a href="#cb12-41" aria-hidden="true" tabindex="-1"></a>                          <span class="at">returnmat =</span> F, </span>
<span id="cb12-42"><a href="#cb12-42" aria-hidden="true" tabindex="-1"></a>                          <span class="at">savepath =</span> heatpath,</span>
<span id="cb12-43"><a href="#cb12-43" aria-hidden="true" tabindex="-1"></a>                          <span class="at">savename =</span> <span class="fu">paste0</span>(ctype, <span class="st">" "</span>,umod[u],<span class="st">'.pdf'</span>))</span>
<span id="cb12-44"><a href="#cb12-44" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb12-45"><a href="#cb12-45" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>You can also create a customized map by returning the matrix from
<code>LeadEdgeSampleHeatmap</code>.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co"># scglmmr function to extract leading edge genes </span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>lexp1 <span class="ot">=</span> <span class="fu">LeadEdgeTidySampleExprs</span>(<span class="at">av.exprs.list =</span> lcpm, <span class="at">gsea.list =</span> g1c, <span class="at">padj.filter =</span> <span class="fl">0.05</span>, <span class="at">NES.filter =</span> <span class="dv">0</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="co"># annotate time and batch on the heatmap</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>heatmap_anno <span class="ot">=</span> meta[, <span class="fu">c</span>(<span class="st">'batch'</span>, <span class="st">'timepoint'</span>)]</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>anno_color <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">timepoint =</span> <span class="fu">c</span>(<span class="st">'1'</span> <span class="ot">=</span> <span class="st">"orange"</span>, <span class="st">'2'</span> <span class="ot">=</span> <span class="st">'red'</span>,  <span class="st">"0"</span> <span class="ot">=</span> <span class="st">"white"</span>),</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">batch =</span> <span class="fu">c</span>(<span class="st">'1'</span> <span class="ot">=</span> <span class="st">"black"</span>, <span class="st">'2'</span> <span class="ot">=</span> <span class="st">"white"</span>)</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a><span class="co"># define your own custom color vector for the log fold change values</span></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>cu <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"#053061"</span>, <span class="st">"#1E61A5"</span>, <span class="st">"#3C8ABE"</span>, <span class="st">"#7CB7D6"</span>, <span class="st">"#BAD9E9"</span>, <span class="st">"#E5EEF3"</span>, </span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>       <span class="st">"#F9EAE1"</span>, <span class="st">"#F9C7AD"</span>, <span class="st">"#EB9273"</span>, <span class="st">"#CF5246"</span>, <span class="st">"#AB1529"</span>, <span class="st">"#67001F"</span>)</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a><span class="co"># scglmmr function for leading edge gene matrix across donors</span></span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>mat2 <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">LeadEdgeSampleHeatmap</span>(<span class="at">tidy.exprs.list =</span> le_expr, </span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">modulename =</span> <span class="st">"reactome interferon signaling"</span>,</span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">celltype_plot =</span> <span class="st">'CD14_Mono'</span>,</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>                                 <span class="co"># metadata = meta, # unused  </span></span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>                                 <span class="co"># metadata_annotate = c('batch'), # unused </span></span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">sample_column =</span> <span class="st">'sample'</span>,</span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">returnmat =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a><span class="co"># draw heatmap </span></span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a>pheatmap<span class="sc">::</span><span class="fu">pheatmap</span>(mat2, </span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a>                   <span class="at">border_color =</span> <span class="cn">NA</span>,</span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a>                   <span class="at">treeheight_row =</span> <span class="dv">0</span>, <span class="at">treeheight_col =</span> <span class="dv">10</span>,</span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a>                   <span class="at">annotation =</span> heatmap_anno,</span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a>                   <span class="at">annotation_colors =</span> anno_color,</span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a>                   <span class="at">color =</span> cu,</span>
<span id="cb13-31"><a href="#cb13-31" aria-hidden="true" tabindex="-1"></a>                   <span class="at">width =</span> <span class="dv">5</span>,  <span class="at">height =</span> <span class="fl">7.6</span>,</span>
<span id="cb13-32"><a href="#cb13-32" aria-hidden="true" tabindex="-1"></a>                   <span class="co"># this can be set to false to look at raw expression e.g. </span></span>
<span id="cb13-33"><a href="#cb13-33" aria-hidden="true" tabindex="-1"></a>                   <span class="at">scale =</span> <span class="st">"row"</span>,</span>
<span id="cb13-34"><a href="#cb13-34" aria-hidden="true" tabindex="-1"></a>                   <span class="at">filename =</span> <span class="fu">paste0</span>(figpath, <span class="st">"mono_d1_ReactomeIFN.pdf"</span>)</span>
<span id="cb13-35"><a href="#cb13-35" aria-hidden="true" tabindex="-1"></a>                   )</span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Matthew Mulè.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
