<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scglmmr">
<title>Analysis of cohort perturbation experiments including nested group designs - statistical models to multicell correlation networks • scglmmr</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of cohort perturbation experiments including nested group designs - statistical models to multicell correlation networks">
<meta property="og:description" content="scglmmr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scglmmr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/networks.html">Relationships between cell subsets with latent information correlation network analysis</a>
    <a class="dropdown-item" href="../articles/pseudobulk_mixed_effects.html">Analysis of cohort perturbation experiments including nested group designs - statistical models to multicell correlation networks</a>
    <a class="dropdown-item" href="../articles/scglmmr_singlecell.html">Single cell gene and module differential expression with nested group repeated measures experiment designs</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MattPM/scglmmr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Analysis of cohort perturbation experiments including nested group designs - statistical models to multicell correlation networks</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MattPM/scglmmr/blob/HEAD/vignettes/pseudobulk_mixed_effects.rmd" class="external-link"><code>vignettes/pseudobulk_mixed_effects.rmd</code></a></small>
      <div class="d-none name"><code>pseudobulk_mixed_effects.rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="pseudobulk-mixed-effects-model-workflow">Pseudobulk mixed effects model workflow<a class="anchor" aria-label="anchor" href="#pseudobulk-mixed-effects-model-workflow"></a>
</h3>
<p>These functions implement wrappers <code>dream</code> from the
<code>variancePartition</code> package to enable mixed (i.e. varying)
effects models of gene expression. This approach is described in <a href="doi.org/10.1093/bioinformatics/btaa687">Hoffman et al
Bininformatics 2020</a> <strong>this paper should be cited if using the
wrapper below <code>FitDream</code>.</strong></p>
<p>We then run enrichment testing with <code>fgsea</code> to define and
use network methods to understand relationships between different</p>
<p>These models: 1. Allow estimation of treatment effects and treatment
differences across groups in multi-subject perturbation experiments 1.
Adjust for non-independence of repeated measures from the same donors
with random effects<br>
2. Utilize observational level weights to enable normal methods. 3.
Shrink estimates toward genome wide trends with empirical Bayes methods
tailored to lme4 based models.</p>
<p>The experiment design shown below includes multiple subjects
(subjectid), each with multiple measurements (sample) corresponding to a
pre and post treatment timepoint. Additionally subjects are nested into
different pre-defined response groups (e.g. high and low responder).
Below the workflow will describe defining pre-treatment baseline effect
differences, treatment effects across all subjects and the difference in
treatment effects between the groups including covariate adjustment.</p>
<table class="table">
<colgroup>
<col width="21%">
<col width="18%">
<col width="18%">
<col width="21%">
<col width="18%">
</colgroup>
<thead><tr class="header">
<th align="left">sample</th>
<th align="center">subjectid</th>
<th align="right">time</th>
<th align="left">group</th>
<th align="center">sex</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">101_t0</td>
<td align="center">101</td>
<td align="right">d0</td>
<td align="left">low</td>
<td align="center">F</td>
</tr>
<tr class="even">
<td align="left">101_t1</td>
<td align="center">101</td>
<td align="right">d1</td>
<td align="left">high</td>
<td align="center">F</td>
</tr>
<tr class="odd">
<td align="left">102_t0</td>
<td align="center">102</td>
<td align="right">d0</td>
<td align="left">low</td>
<td align="center">M</td>
</tr>
<tr class="even">
<td align="left">102_t1</td>
<td align="center">102</td>
<td align="right">d1</td>
<td align="left">high</td>
<td align="center">M</td>
</tr>
<tr class="odd">
<td align="left">… (x n donors)</td>
<td align="center">… (x n donors)</td>
<td align="right">… (x n donors)</td>
<td align="left">… (x n donors)</td>
<td align="center">… (x n donors)</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="getting-started---setup-data-loading-and-quaity-control">Getting started - setup, data loading, and quaity control<a class="anchor" aria-label="anchor" href="#getting-started---setup-data-loading-and-quaity-control"></a>
</h3>
<p>In the workflow below, mixed effect model fits and gene set
enrichment steps are parallelized with the BiocParallel package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#devtools::install_github(repo = "https://github.com/MattPM/scglmmr")</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MattPM/scglmmr" class="external-link">scglmmr</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#the mixed model fits and gsea are parallelized </span></span>
<span><span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html" class="external-link">register</a></span><span class="op">(</span><span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pparam</span> <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span>, type <span class="op">=</span> <span class="st">"SOCK"</span>, progressbar <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Load single cell data and quality control</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">datapath</span> <span class="op">=</span> <span class="st">"mypath/"</span></span>
<span></span>
<span><span class="co"># load seurat or sce object etc. </span></span>
<span><span class="va">s</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"path/seuratobject.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># define counts and metadata and subset to cells above rm seurat object from workspace </span></span>
<span><span class="va">meta</span> <span class="op">=</span> <span class="va">s</span><span class="op">@</span><span class="va">meta.data</span></span>
<span><span class="va">umi</span> <span class="op">=</span> <span class="va">s</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">s</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/gc.html" class="external-link">gc</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># QC contingency of cells by subject for each celltype </span></span>
<span><span class="va">tab</span> <span class="op">=</span> <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="../reference/SubjectCelltypeTable.html">SubjectCelltypeTable</a></span><span class="op">(</span>metadata <span class="op">=</span> <span class="va">meta</span>, celltype_column <span class="op">=</span> <span class="st">"celltype"</span>, sample_column <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></span>
<span><span class="va">tab</span><span class="op">$</span><span class="va">celltypes_remove</span>; <span class="va">tab</span><span class="op">$</span><span class="va">`low representation celltypes`</span>; <span class="va">tab</span><span class="op">$</span><span class="va">table</span></span>
<span></span>
<span><span class="co"># remove cells prior to pseudobulk analysis with 0 representation in some donors. </span></span>
<span><span class="va">meta</span> <span class="op">=</span> <span class="va">meta</span><span class="op">[</span><span class="op">!</span><span class="va">meta</span><span class="op">$</span><span class="va">celltype</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">tab</span><span class="op">$</span><span class="va">celltypes_remove</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># subset data </span></span>
<span><span class="va">umi</span> <span class="op">=</span> <span class="va">umi</span><span class="op">[</span> ,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Create aggregated pseudobulk data</span></span>
<span><span class="va">pb</span> <span class="op">=</span> <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="../reference/PseudobulkList.html">PseudobulkList</a></span><span class="op">(</span></span>
<span>  rawcounts <span class="op">=</span> <span class="va">umi</span>,</span>
<span>  metadata <span class="op">=</span> <span class="va">meta</span>, </span>
<span>  sample_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>  celltype_col <span class="op">=</span> <span class="st">"celltype"</span>,</span>
<span>  avg_or_sum <span class="op">=</span> <span class="st">"sum"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Create aggretated sample level metadata</span></span>
<span><span class="va">met</span> <span class="op">=</span> <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="../reference/AggregateCellMetadata.html">AggregateCellMetadata</a></span><span class="op">(</span></span>
<span>  cell.metadata <span class="op">=</span> <span class="va">meta</span>, </span>
<span>  sample_column <span class="op">=</span> <span class="st">'sample'</span>, </span>
<span>  variable_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'subjectid'</span>, <span class="st">'timepoint'</span>, <span class="st">'response'</span>, <span class="st">'age'</span>, <span class="st">'sex'</span><span class="op">)</span>,</span>
<span>  pseudobulk.List <span class="op">=</span> <span class="va">pb</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>Now we create a combined grouping factor indicating both timepoint
and response group for each sample which will help define any contrast
of effects with respect to time and group, e.g. differences in treatment
effects.</p>
<p>Note in a one way design where the target estimand is for example the
treatment effect only, this step is not needed as we can simply extract
the effect of “time” from the lme4 fits.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">met</span><span class="op">$</span><span class="va">group.time</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">met</span><span class="op">$</span><span class="va">group</span>, <span class="va">met</span><span class="op">$</span><span class="va">timepoint</span>, sep <span class="op">=</span> <span class="st">'_'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># make sure this is a factor and the levels are in a coherent order for the </span></span>
<span><span class="co"># contrast matrix -- see below. here: </span></span>
<span><span class="co"># time 0 = 0, time 1 = 1. low group = 0 high group = 1. </span></span>
<span><span class="va">met</span><span class="op">$</span><span class="va">group.time</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span></span>
<span>  <span class="va">met</span><span class="op">$</span><span class="va">group.time</span>,</span>
<span>  levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1_0"</span>, <span class="st">"1_1"</span>, <span class="st">"0_0"</span>, <span class="st">"0_1"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># now filter genes within each cell type that are reasonably expressed. </span></span>
<span><span class="va">design</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">met</span><span class="op">$</span><span class="va">group.time</span><span class="op">)</span></span>
<span><span class="va">dge</span> <span class="op">=</span> <span class="fu"><a href="../reference/Normalize.html">Normalize</a></span><span class="op">(</span>pseudobulk.list <span class="op">=</span> <span class="va">pb</span>, design <span class="op">=</span> <span class="va">design</span>, minimum.gene.count <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fit-models-and-specify-a-priori-contrasts-corresponding-to-the-desired-effect-comparisons">Fit models and specify a priori contrasts corresponding to the
desired effect comparisons<a class="anchor" aria-label="anchor" href="#fit-models-and-specify-a-priori-contrasts-corresponding-to-the-desired-effect-comparisons"></a>
</h3>
<p>Below shows 2 steps. 1: Specify a varying intercept model using <a href="https://arxiv.org/pdf/1911.08628.pdfs" class="external-link">lme4 symbolic model
formula</a>. The <code>1|subjectID</code> term treats individual as a
random effect. This partially pools the estimate for baseline expression
toward the global trend for each cell type by fitting a normal
distribution to baseline expression (for each gene in the genome
separately).</p>
<p>In the example below, we want estimate baseline differences,
treatment effects across all individuals and the difference in fold
changes between groups adjusting estimates for age and sex (and modeling
the individual variation with a varying effect). To do this we specify a
custom a. priori contrast matrix.</p>
<p>2: Using the factor variable combining group and time, we create
contrasts over the levels to define effects of interest. The function
<code>makeContrastsDream</code> is used for this purpose which works the
same way as the limma function makeContrasts. More simple contrasts or
more complex contrasts are also possible. This step is critical for
deriving the correct effec size estimates. More information on
specifying contrast matrices is available here: <a href="https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html" class="external-link">A
guide to creating design matrices for gene expression
experiments</a></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specify a random intercept model with lme4 syntax.</span></span>
<span><span class="va">f1</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">age</span> <span class="op">+</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">group.time</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span><span class="op">|</span><span class="va">subjectid</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># make contrast matrix </span></span>
<span><span class="va">L2</span> <span class="op">=</span> <span class="fu">makeContrastsDream</span><span class="op">(</span></span>
<span>  formula <span class="op">=</span> <span class="va">f1</span>, </span>
<span>  data <span class="op">=</span> <span class="va">samplemd</span>,</span>
<span>  contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    baseline <span class="op">=</span> <span class="st">"group.time1_0 - group.time0_0"</span>, <span class="co"># note fixef model also fit for this single time point contrast </span></span>
<span>    treatment_delta <span class="op">=</span> <span class="st">"( group.time1_1 - group.time1_0 ) - ( group.time0_1 - group.time0_0 )"</span>,</span>
<span>    treatment <span class="op">=</span> <span class="st">"( group.time1_1 + group.time0_1 ) / 2 - ( group.time1_0 + group.time0_0 ) / 2 "</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># visualize the contrast matrix, compare to levels of group.time variable</span></span>
<span><span class="co"># to check for correctly specified effects. </span></span>
<span><span class="fu">plotContrasts</span><span class="op">(</span><span class="va">L2</span><span class="op">)</span> </span></code></pre></div>
<p>Fit the models with <code><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/dream-method.html" class="external-link">variancePartition::dream</a></code> which uses
voom weights in lme4 mixed effects models with an empirical bayes
shrinkage toward genome wide trend accounting for per gene degrees of
freedom.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">fit1</span> <span class="op">=</span> <span class="fu"><a href="../reference/FitDream.html">FitDream</a></span><span class="op">(</span>pb.list <span class="op">=</span> <span class="va">dge</span>, </span>
<span>                sample.metadata <span class="op">=</span> <span class="va">met</span>, </span>
<span>                lme4.formula <span class="op">=</span> <span class="va">f1</span>,</span>
<span>                dream.contrast.matrix <span class="op">=</span> <span class="va">L2</span>,</span>
<span>                ncores <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p>Note, <code><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/dream-method.html" class="external-link">variancePartition::dream</a></code> now incorporates an
emperical Bayes step derived for mixed effects models accounting for per
gene degrees of freedom, please see: <a href="https://github.com/GabrielHoffman/variancePartition/issues/54" class="external-link uri">https://github.com/GabrielHoffman/variancePartition/issues/54</a>.</p>
</div>
<div class="section level3">
<h3 id="downstream-gene-set-enrichment-analysis-within-celltypes-for-different-effects">Downstream gene set enrichment analysis within celltypes for
different effects<a class="anchor" aria-label="anchor" href="#downstream-gene-set-enrichment-analysis-within-celltypes-for-different-effects"></a>
</h3>
<p>Run gene set enrichment analysis within each cell type fbased on
genes ranked by effect size for each of the effects defined above.</p>
<p>Here within each cell type and for each contrast we run gene set
enrichment analysis using 2 functions <code>ExtractResult</code> and
<code>FgseaList</code>. <code>ExtractResult</code> is used to extract a
list of gene ranks (it can be used as a wrapper around dream::topTable
and limma::topTable to return a full list of results). Since we had
multiple covariates in a mixed model and we used a custom contrast
matrix, we specify the covariate of interest from the contrast using
arguments <code>coefficient.number</code> and <code>coef.name</code>
which are output in results based on the names of the contrasts.</p>
<p>Based on our contrast matrix (the object <code>L2</code> we created
above with <code>makeContrastsDream</code>),
<code>coefficient.number = 1</code> corresponds to the baseline
difference between groups, <code>coefficient.number = 2</code> is the
difference in fold changes between the groups and
<code>coefficient.number = 3</code> is the pre vs post perturbation
difference across subjects in both groups. Estimates for the covariates
are also availabe.</p>
<div class="section level4">
<h4 id="examples-of-extracting-gene-ranks-based-on-contrasts-and-running-gene-set-enrichment">Examples of extracting gene ranks based on contrasts and running
gene set enrichment<a class="anchor" aria-label="anchor" href="#examples-of-extracting-gene-ranks-based-on-contrasts-and-running-gene-set-enrichment"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># msigDB hallmark pathways are included in the scglmmr package</span></span>
<span><span class="va">hlmk</span> <span class="op">=</span> <span class="fu">scglmmr</span><span class="fu">::</span><span class="va">hallmark</span></span>
<span><span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="../reference/ExtractResult.html">ExtractResult</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># extract genes ranked by treatment effect (across all donors) </span></span>
<span><span class="va">rtreat</span> <span class="op">=</span> <span class="fu"><a href="../reference/ExtractResult.html">ExtractResult</a></span><span class="op">(</span>model.fit.list <span class="op">=</span> <span class="va">fit1</span>,</span>
<span>                       what <span class="op">=</span> <span class="st">'lmer.z.ranks'</span>,</span>
<span>                       coefficient.number <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                       coef.name <span class="op">=</span> <span class="st">'treatment'</span><span class="op">)</span></span>
<span><span class="co"># run fgsea </span></span>
<span><span class="va">hlmk.treat</span> <span class="op">=</span> <span class="fu"><a href="../reference/FgseaList.html">FgseaList</a></span><span class="op">(</span>rank.list.celltype <span class="op">=</span> <span class="va">rtreat</span>,</span>
<span>                       pathways <span class="op">=</span> <span class="va">hlmk</span>,</span>
<span>                       BPPARAM <span class="op">=</span> <span class="va">pparam</span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co"># extract gene ranks of treatment effect (across all donors) </span></span>
<span><span class="va">rdelta</span> <span class="op">=</span> <span class="fu"><a href="../reference/ExtractResult.html">ExtractResult</a></span><span class="op">(</span>model.fit.list <span class="op">=</span> <span class="va">fit1</span>,</span>
<span>                       what <span class="op">=</span> <span class="st">'lmer.z.ranks'</span>,</span>
<span>                       coefficient.number <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                       coef.name <span class="op">=</span> <span class="st">'treatment_delta'</span><span class="op">)</span></span>
<span><span class="va">hlmk.delta</span> <span class="op">=</span> <span class="fu"><a href="../reference/RunFgseaOnRankList.html">RunFgseaOnRankList</a></span><span class="op">(</span>rank.list.celltype <span class="op">=</span> <span class="va">rdelta</span>,</span>
<span>                                pathways <span class="op">=</span> <span class="va">hlmk</span>,</span>
<span>                                BPPARAM <span class="op">=</span> <span class="va">pparam</span><span class="op">)</span></span></code></pre></div>
<p>On a technical note, it’s also possible to fit a simple model for the
baseline contrast not estimating the variation across subjects with a
random effect as this contrast is more typically estimated with a
standard group 1 vs 2 least squares approach. The function
<code>RunVoomLimma</code> is provided to fit these models. If you
compare the effect size estimates for an individual cell type for the
<code>coefficient.number = 1</code> from the mixed model fits (object
<code>fit1</code> above) to the models below they will be very highly
corrleated along the diagonal with some differences arising to the
different degrees of freedom, sample size and variance coming from the
additional layer of variation estimated by the mixed model. The choice
of model to use is up to the user.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># fit simple linear model for the baseline group level contrast </span></span>
<span><span class="va">design.2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fl">0</span> <span class="op">+</span> <span class="va">age</span> <span class="op">+</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">group.time</span>, data <span class="op">=</span> <span class="va">met</span><span class="op">)</span></span>
<span><span class="va">fit0</span> <span class="op">=</span> <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="../reference/RunVoomLimma.html">RunVoomLimma</a></span><span class="op">(</span>dgelists <span class="op">=</span> <span class="va">dge</span>, </span>
<span>                           design_matrix <span class="op">=</span> <span class="va">design.2</span>, </span>
<span>                           do_contrast_fit <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                           <span class="co"># we use only the first row of the contrast matrix L2</span></span>
<span>                           my_contrast_matrix <span class="op">=</span> <span class="va">L2</span><span class="op">[</span> ,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># extract fixed efect model estimate of baseline contrast (pre treatment differences between groups)</span></span>
<span><span class="va">r0</span> <span class="op">=</span> <span class="fu"><a href="../reference/ExtractResult.html">ExtractResult</a></span><span class="op">(</span>model.fit.list <span class="op">=</span> <span class="va">fit0</span>,</span>
<span>                   what <span class="op">=</span> <span class="st">'gene.t.ranks'</span>,</span>
<span>                   coefficient.number <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                   coef.name <span class="op">=</span> <span class="st">'baseline'</span><span class="op">)</span></span>
<span><span class="co">## run gene set enrichment </span></span>
<span><span class="va">hlmk.0</span> <span class="op">=</span> <span class="fu"><a href="../reference/FgseaList.html">FgseaList</a></span><span class="op">(</span>rank.list.celltype <span class="op">=</span> <span class="va">r0</span>,</span>
<span>                   pathways <span class="op">=</span> <span class="va">hlmk</span>,</span>
<span>                   BPPARAM <span class="op">=</span> <span class="va">pparam</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="further-analysis-and-curation-of-enrichment-results">Further analysis and curation of enrichment results<a class="anchor" aria-label="anchor" href="#further-analysis-and-curation-of-enrichment-results"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># full set of leading edge genes indexed by celltype x effect x module </span></span>
<span><span class="va">lefull</span> <span class="op">=</span> <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="../reference/GetLeadingEdgeFull.html">GetLeadingEdgeFull</a></span><span class="op">(</span>gsea.list <span class="op">=</span> <span class="va">hlmk.treat</span>,</span>
<span>                                     padj.filter <span class="op">=</span> <span class="fl">0.02</span>, </span>
<span>                                     NES.filter <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract model fit results instead of ranks</span></span>
<span><span class="co"># automatically this sets argument `result` to 'statistics'</span></span>
<span><span class="va">fit1.res</span> <span class="op">=</span> <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="../reference/ExtractResult.html">ExtractResult</a></span><span class="op">(</span>model.fit.list <span class="op">=</span> <span class="va">fit1</span> , </span>
<span>                                  coefficient.number <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                                  coef.name <span class="op">=</span> <span class="st">'treatment'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># combine GSEA results with model coefficient for each gene in leading edge </span></span>
<span><span class="co"># include all leading edge genes irrespective of individual gene p value </span></span>
<span><span class="va">cr</span> <span class="op">=</span> <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="../reference/CombineResults.html">CombineResults</a></span><span class="op">(</span>gsealist <span class="op">=</span> <span class="va">hlmk.treat</span>, </span>
<span>                             contrastlist <span class="op">=</span> <span class="va">fit1.res</span>, </span>
<span>                             gseafdr <span class="op">=</span> <span class="fl">0.02</span>, </span>
<span>                             genefdr <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="extract-all-leading-edge-genes-indexed-by-cell-type">Extract all leading edge genes indexed by cell type<a class="anchor" aria-label="anchor" href="#extract-all-leading-edge-genes-indexed-by-cell-type"></a>
</h3>
<p>This outputs a nested list of leading edge genes from each enrichment
across cell types. list level 1 is cell type, level 2 is enrichment
signal.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">li</span> <span class="op">=</span> <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="../reference/LeadingEdgeIndexed.html">LeadingEdgeIndexed</a></span><span class="op">(</span>gsea.result.list <span class="op">=</span> <span class="va">hlmk.treat</span>, padj.threshold <span class="op">=</span> <span class="fl">0.02</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="calculate-the-jaccard-similarity-of-the-leading-edge-genes-for-enrichments-within-a-given-cell-type-and-effect">Calculate the Jaccard similarity of the leading edge genes for
enrichments within a given cell type and effect<a class="anchor" aria-label="anchor" href="#calculate-the-jaccard-similarity-of-the-leading-edge-genes-for-enrichments-within-a-given-cell-type-and-effect"></a>
</h3>
<p>This is useful for understanding which enrichment signals may come
from the same vs distinct genes.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># figpath.temp = here('figures')</span></span>
<span><span class="va">treat.JI</span> <span class="op">=</span> <span class="fu"><a href="../reference/EnrichmentJaccard.html">EnrichmentJaccard</a></span><span class="op">(</span>gsealist <span class="op">=</span> <span class="va">hlmk.treat</span>, </span>
<span>                          indexedgenes <span class="op">=</span> <span class="va">li</span>, </span>
<span>                          <span class="co">#saveplot = TRUE, </span></span>
<span>                          <span class="co">#figpath = figpath.temp,</span></span>
<span>                          returnJaccardMtx <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># curate results </span></span>
<span><span class="va">results.sorted</span> <span class="op">=</span> <span class="va">treat.JI</span><span class="op">$</span><span class="va">sortedgsea</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>signal <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">celltype</span>, <span class="va">pathway</span>, sep <span class="op">=</span> <span class="st">'~'</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span><span class="co"># data.table::fwrite(results.sorted, file = paste0(datapath, 'g0.result.sort.txt'),sep = "\t")</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualization-of-enrichment-results">Visualization of enrichment results<a class="anchor" aria-label="anchor" href="#visualization-of-enrichment-results"></a>
</h3>
<p><strong>Create a bubble plot heatmap of enrichment results within
clusters</strong></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># NES_filter to -Inf to plot positive and negative enrichment</span></span>
<span><span class="va">p</span> <span class="op">=</span> <span class="fu"><a href="../reference/PlotFgsea.html">PlotFgsea</a></span><span class="op">(</span>gsea_result_list <span class="op">=</span> <span class="va">hlmk.treat</span>, NES_filter <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span>,padj_filter <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p><strong>Create heatmap of gene perturbation fold changes across cell
subsets based on model fit coefficients</strong></p>
<p>Extract and make a heatmap of log fold changes across celltypes.
HeatmapDiag uses the package <a href="https://CRAN.R-project.org/package=slanter" class="external-link">slanter</a> which
accepts the same arguments as <code>pheatmap</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene.mat</span> <span class="op">=</span> <span class="fu"><a href="../reference/GetGeneMatrix.html">GetGeneMatrix</a></span><span class="op">(</span>result.list <span class="op">=</span> <span class="va">fit1</span>, </span>
<span>                         pvalfilter <span class="op">=</span> <span class="fl">0.05</span>, </span>
<span>                         stat_for_matrix <span class="op">=</span> <span class="st">'logFC'</span>,</span>
<span>                         logfcfilter <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#make heatmap  e.g. </span></span>
<span><span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">gene.mat</span>, fontsize_row <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># diagnoalize the heatmap so that the most correlated signals are grouped </span></span>
<span><span class="co"># this uses the package slanter </span></span>
<span><span class="fu"><a href="../reference/HeatmapDiag.html">HeatmapDiag</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="va">gene.mat</span>, fontsize_row <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p><strong>Create heatmap of gene perturbation fold changes from
enrichments across individuals within a cell subset</strong></p>
<p>Visualize the log counts per million at the sample level of the gene
set enrichment results.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># make tidy average data for visualization of weighted pb results </span></span>
<span><span class="va">lcmp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">pb</span>, <span class="fu">edgeR</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html" class="external-link">cpm</a></span>, log <span class="op">=</span> <span class="cn">TRUE</span> <span class="op">)</span></span>
<span></span>
<span><span class="co"># </span></span>
<span><span class="va">le_expr</span> <span class="op">=</span> <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="../reference/LeadEdgeTidySampleExprs.html">LeadEdgeTidySampleExprs</a></span><span class="op">(</span>av.exprs.list <span class="op">=</span> <span class="va">lcmp</span>,</span>
<span>                                           gsea.list <span class="op">=</span> <span class="va">hlmk.treat</span>, </span>
<span>                                           padj.filter <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                                           NES.filter <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># example plot of sample level average leading edge genes annotated </span></span>
<span><span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="../reference/LeadEdgeSampleHeatmap.html">LeadEdgeSampleHeatmap</a></span><span class="op">(</span>tidy.exprs.list <span class="op">=</span> <span class="va">le_expr</span>,</span>
<span>                               modulename <span class="op">=</span> <span class="st">"HALLMARK_HYPOXIA"</span>,</span>
<span>                               elltype_plot <span class="op">=</span> <span class="st">"CD4_NaiveTcell"</span>,</span>
<span>                               metadata <span class="op">=</span> <span class="va">meta</span>, </span>
<span>                               metadata_annotate <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'group'</span>, <span class="st">'timepoint'</span>, <span class="st">'age'</span>, <span class="st">'sex'</span><span class="op">)</span>,</span>
<span>                               sample_column <span class="op">=</span> <span class="st">'sample'</span>,</span>
<span>                               returnmat <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                               savepath <span class="op">=</span> <span class="va">figpath</span>, </span>
<span>                               savename <span class="op">=</span> <span class="st">"filename"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># see also: </span></span>
<span><span class="co"># scglmmr::TopGenesTidySampleExprs() - same as aobve for 'top' de genes estimated by model coeffcient / p value. </span></span>
<span><span class="co"># scglmmr::GetTidySummary() - for custom plotting</span></span>
<span></span>
<span></span>
<span><span class="co"># sample level vsualization of average data above </span></span>
<span><span class="co"># repeat this for all enriched pathways </span></span>
<span><span class="va">heatpath</span> <span class="op">=</span> <span class="fu">here</span><span class="op">(</span><span class="st">"sime/path"</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">heatpath</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">le_expr</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">cdat</span> <span class="op">=</span> <span class="va">le_expr</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">ctype</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">le_expr</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">umod</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">cdat</span><span class="op">$</span><span class="va">module</span><span class="op">)</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">u</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">umod</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="../reference/LeadEdgeSampleHeatmap.html">LeadEdgeSampleHeatmap</a></span><span class="op">(</span>tidy.exprs.list <span class="op">=</span> <span class="va">le_expr</span>, </span>
<span>                          modulename <span class="op">=</span> <span class="va">umod</span><span class="op">[</span><span class="va">u</span><span class="op">]</span>, </span>
<span>                          celltype_plot <span class="op">=</span> <span class="va">ctype</span>,</span>
<span>                          metadata <span class="op">=</span> <span class="va">meta</span>, metadata_annotate <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'group'</span>, <span class="st">'timepoint'</span>, <span class="st">'age'</span>, <span class="st">'gender'</span><span class="op">)</span>,</span>
<span>                          sample_column <span class="op">=</span> <span class="st">'sample'</span>,</span>
<span>                          returnmat <span class="op">=</span> <span class="cn">F</span>, </span>
<span>                          savepath <span class="op">=</span> <span class="va">heatpath</span>,</span>
<span>                          savename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">ctype</span>, <span class="st">" "</span>,<span class="va">umod</span><span class="op">[</span><span class="va">u</span><span class="op">]</span>,<span class="st">'.pdf'</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>You can also create a customized map by returning the matrix from
<code>LeadEdgeSampleHeatmap</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># scglmmr function to extract leading edge genes </span></span>
<span><span class="va">lexp1</span> <span class="op">=</span> <span class="fu"><a href="../reference/LeadEdgeTidySampleExprs.html">LeadEdgeTidySampleExprs</a></span><span class="op">(</span>av.exprs.list <span class="op">=</span> <span class="va">lcpm</span>, gsea.list <span class="op">=</span> <span class="va">g1c</span>, padj.filter <span class="op">=</span> <span class="fl">0.05</span>, NES.filter <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># annotate time and batch on the heatmap</span></span>
<span><span class="va">heatmap_anno</span> <span class="op">=</span> <span class="va">meta</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'batch'</span>, <span class="st">'timepoint'</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">anno_color</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  timepoint <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'1'</span> <span class="op">=</span> <span class="st">"orange"</span>, <span class="st">'2'</span> <span class="op">=</span> <span class="st">'red'</span>,  <span class="st">"0"</span> <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span>,</span>
<span>  batch <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'1'</span> <span class="op">=</span> <span class="st">"black"</span>, <span class="st">'2'</span> <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># define your own custom color vector for the log fold change values</span></span>
<span><span class="va">cu</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#053061"</span>, <span class="st">"#1E61A5"</span>, <span class="st">"#3C8ABE"</span>, <span class="st">"#7CB7D6"</span>, <span class="st">"#BAD9E9"</span>, <span class="st">"#E5EEF3"</span>, </span>
<span>       <span class="st">"#F9EAE1"</span>, <span class="st">"#F9C7AD"</span>, <span class="st">"#EB9273"</span>, <span class="st">"#CF5246"</span>, <span class="st">"#AB1529"</span>, <span class="st">"#67001F"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># scglmmr function for leading edge gene matrix across donors</span></span>
<span><span class="va">mat2</span> <span class="op">=</span> <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="../reference/LeadEdgeSampleHeatmap.html">LeadEdgeSampleHeatmap</a></span><span class="op">(</span>tidy.exprs.list <span class="op">=</span> <span class="va">le_expr</span>, </span>
<span>                                 modulename <span class="op">=</span> <span class="st">"reactome interferon signaling"</span>,</span>
<span>                                 celltype_plot <span class="op">=</span> <span class="st">'CD14_Mono'</span>,</span>
<span>                                 <span class="co"># metadata = meta, # unused  </span></span>
<span>                                 <span class="co"># metadata_annotate = c('batch'), # unused </span></span>
<span>                                 sample_column <span class="op">=</span> <span class="st">'sample'</span>,</span>
<span>                                 returnmat <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># draw heatmap </span></span>
<span><span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">mat2</span>, </span>
<span>                   border_color <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>                   treeheight_row <span class="op">=</span> <span class="fl">0</span>, treeheight_col <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                   annotation <span class="op">=</span> <span class="va">heatmap_anno</span>,</span>
<span>                   annotation_colors <span class="op">=</span> <span class="va">anno_color</span>,</span>
<span>                   color <span class="op">=</span> <span class="va">cu</span>,</span>
<span>                   width <span class="op">=</span> <span class="fl">5</span>,  height <span class="op">=</span> <span class="fl">7.6</span>,</span>
<span>                   <span class="co"># this can be set to false to look at raw expression e.g. </span></span>
<span>                   scale <span class="op">=</span> <span class="st">"row"</span>,</span>
<span>                   filename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">figpath</span>, <span class="st">"mono_d1_ReactomeIFN.pdf"</span><span class="op">)</span></span>
<span>                   <span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Matthew Mulè.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
