<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>nested group repeated measures pseudobulk differential expression testing • scglmmr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="nested group repeated measures pseudobulk differential expression testing">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scglmmr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/pseudobulk_mixed_effects.html">nested group repeated measures pseudobulk differential expression testing</a>
    </li>
    <li>
      <a href="../articles/scglmmr_pseudobulk.html">nested group repeated measures pseudobulk differential expression testing</a>
    </li>
    <li>
      <a href="../articles/scglmmr_singlecell.html">Single cell gene and module level differential expression with nested group repeated measures experiment designs using scglmmr</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MattPM/scglmmr">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>nested group repeated measures pseudobulk differential expression testing</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MattPM/scglmmr/blob/master/vignettes/scglmmr_pseudobulk.Rmd"><code>vignettes/scglmmr_pseudobulk.Rmd</code></a></small>
      <div class="hidden name"><code>scglmmr_pseudobulk.Rmd</code></div>

    </div>

    
    
<div id="pseudobulk-mixed-effects-models" class="section level3">
<h3 class="hasAnchor">
<a href="#pseudobulk-mixed-effects-models" class="anchor"></a>1. pseudobulk mixed effects models</h3>
<p>These functions implement convenience wrappers around both <code>limma</code> for fitting traditional linear models (for e.g. baseline differenced between groups) and the <code>dream</code> method from the <code>variancePartition</code> package. This is the only pseudobulk differential expression method that can accomodate both fixed and random effects (which are statistically necessary to account for non-independence of multiple measurements from the same donor) via use of <code>lme4</code> for mixed effects modeling while accounting for measurement uncertainty via pooling unqeual library sizes by incorporating <code>voom</code> observational weights. This method performed well in related simulation studies by <a href="https://www.nature.com/articles/s41467-020-19894-4">Crowell et. al. 2020</a>. scglmmr performs statistical contrasts between groups comparing the least squares means after accounting for model covariates is empowered by the <a href="https://cran.r-project.org/web/packages/emmeans/index.html">emmeans</a> package.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(scglmmr)</a></code></pre></div>
<p>Below is a 2 group repeated measures experiment. A random intercept model is required to account for autocorrelation in each donors post perturbation response with their baseline level of expression. This data can be accomodated by scglmmr. More simple experiment designs are also supported, for example data with 2 groups but not repeated pre/post treatment measurements, see the <code><a href="../reference/RunVoomLimma.html">RunVoomLimma()</a></code> function.</p>
<table class="table">
<thead><tr class="header">
<th align="left">sample</th>
<th align="center">sampleid</th>
<th align="right">time</th>
<th align="left">group</th>
<th align="center">sex</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">101_t0</td>
<td align="center">101</td>
<td align="right">d0</td>
<td align="left">low</td>
<td align="center">F</td>
</tr>
<tr class="even">
<td align="left">101_t1</td>
<td align="center">101</td>
<td align="right">d1</td>
<td align="left">high</td>
<td align="center">F</td>
</tr>
<tr class="odd">
<td align="left">102_t0</td>
<td align="center">102</td>
<td align="right">d0</td>
<td align="left">low</td>
<td align="center">M</td>
</tr>
<tr class="even">
<td align="left">102_t1</td>
<td align="center">102</td>
<td align="right">d1</td>
<td align="left">high</td>
<td align="center">M</td>
</tr>
<tr class="odd">
<td align="left">… (x n donors)</td>
<td align="center">… (x n donors)</td>
<td align="right">… (x n donors)</td>
<td align="left">… (x n donors)</td>
<td align="center">… (x n donors)</td>
</tr>
</tbody>
</table>
</div>
<div id="pseudobulk-mixed-effects-model" class="section level3">
<h3 class="hasAnchor">
<a href="#pseudobulk-mixed-effects-model" class="anchor"></a>1. “Pseudobulk” mixed effects model</h3>
<p>Fit a weighted mixed effects model to test group and time effects using lme4 model formula syntax and using dream observational weights using <code>dream</code> from the <code>variancePartition</code> package.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co">#devtools::install_github(repo = "https://github.com/MattPM/scglmmr")</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(scglmmr)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">datapath =<span class="st"> "mypath/"</span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="co"># load seurat or sce object etc. </span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6">s =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span>(<span class="st">"path/seuratobject.rds"</span>)</a>
<a class="sourceLine" id="cb2-7" data-line-number="7"></a>
<a class="sourceLine" id="cb2-8" data-line-number="8"><span class="co"># define counts and metadata and subset to cells above rm seurat object from workspace </span></a>
<a class="sourceLine" id="cb2-9" data-line-number="9">meta =<span class="st"> </span>s<span class="op">@</span>meta.data</a>
<a class="sourceLine" id="cb2-10" data-line-number="10">umi =<span class="st"> </span>s<span class="op">@</span>assays<span class="op">$</span>RNA<span class="op">@</span>counts</a>
<a class="sourceLine" id="cb2-11" data-line-number="11"><span class="kw"><a href="https://rdrr.io/r/base/rm.html">rm</a></span>(s); <span class="kw"><a href="https://rdrr.io/r/base/gc.html">gc</a></span>()</a>
<a class="sourceLine" id="cb2-12" data-line-number="12"></a>
<a class="sourceLine" id="cb2-13" data-line-number="13"><span class="co"># QC contingency of cells by subject for each celltype </span></a>
<a class="sourceLine" id="cb2-14" data-line-number="14">tab =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/SubjectCelltypeTable.html">SubjectCelltypeTable</a></span>(<span class="dt">metadata =</span> meta, <span class="dt">celltype_column =</span> <span class="st">"lineage"</span>, <span class="dt">sample_column =</span> <span class="st">"sample"</span>)</a>
<a class="sourceLine" id="cb2-15" data-line-number="15">tab<span class="op">$</span>celltypes_remove; tab<span class="op">$</span><span class="st">`</span><span class="dt">low representation celltypes</span><span class="st">`</span>; tab<span class="op">$</span>table</a>
<a class="sourceLine" id="cb2-16" data-line-number="16"></a>
<a class="sourceLine" id="cb2-17" data-line-number="17"><span class="co"># remove cells prior to pseudobulk analysis </span></a>
<a class="sourceLine" id="cb2-18" data-line-number="18">meta =<span class="st"> </span>meta[<span class="op">!</span>meta<span class="op">$</span>celltype_label_<span class="dv">3</span> <span class="op">%in%</span><span class="st"> </span>tab<span class="op">$</span>celltypes_remove, ]</a>
<a class="sourceLine" id="cb2-19" data-line-number="19"></a>
<a class="sourceLine" id="cb2-20" data-line-number="20"><span class="co"># subset data </span></a>
<a class="sourceLine" id="cb2-21" data-line-number="21">umi =<span class="st"> </span>umi[ ,<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(meta)]</a>
<a class="sourceLine" id="cb2-22" data-line-number="22"></a>
<a class="sourceLine" id="cb2-23" data-line-number="23"><span class="co"># pseudobulk workflow </span></a>
<a class="sourceLine" id="cb2-24" data-line-number="24">pb =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/PseudobulkList.html">PseudobulkList</a></span>(<span class="dt">rawcounts =</span> umi, <span class="dt">metadata =</span> meta, <span class="dt">sample_col =</span> <span class="st">"sample"</span>,</a>
<a class="sourceLine" id="cb2-25" data-line-number="25">                             <span class="dt">celltype_col =</span> <span class="st">"lineage"</span>, <span class="dt">avg_or_sum =</span> <span class="st">"sum"</span>)</a>
<a class="sourceLine" id="cb2-26" data-line-number="26">designmat =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/BulkDesignMatrix.html">BulkDesignMatrix</a></span>(<span class="dt">metadata =</span> meta, <span class="dt">sample_column =</span> <span class="st">"sample"</span>,</a>
<a class="sourceLine" id="cb2-27" data-line-number="27">                                      <span class="dt">variable_column =</span> <span class="st">"cohort_timepoint"</span>, <span class="dt">pseudobulklist =</span> pb)</a>
<a class="sourceLine" id="cb2-28" data-line-number="28">dge =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/NormalizePseudobulk.html">NormalizePseudobulk</a></span>(<span class="dt">pseudobulklist =</span> pb, <span class="dt">design_matrix =</span> designmat, <span class="dt">minimum.gene.count =</span> <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb2-29" data-line-number="29"></a>
<a class="sourceLine" id="cb2-30" data-line-number="30"><span class="co"># custom a priori contrasts over a combined factor of the group and timepoint variables </span></a>
<a class="sourceLine" id="cb2-31" data-line-number="31">c_mat =<span class="st"> </span><span class="kw">makeContrasts</span>(</a>
<a class="sourceLine" id="cb2-32" data-line-number="32">  <span class="dt">foldchange_difference =</span> (group_timepoint1_<span class="dv">1</span> <span class="op">-</span><span class="st"> </span>group_timepoint1_<span class="dv">0</span>) <span class="op">-</span><span class="st"> </span>(group_timepoint0_<span class="dv">1</span> <span class="op">-</span><span class="st"> </span>group_timepoint0_<span class="dv">0</span>),</a>
<a class="sourceLine" id="cb2-33" data-line-number="33">  <span class="dt">time1_foldchange =</span> (group_timepoint1_<span class="dv">1</span> <span class="op">+</span><span class="st"> </span>group_timepoint0_<span class="dv">1</span>) <span class="op">/</span><span class="st"> </span><span class="dv">2</span>  <span class="op">-</span><span class="st"> </span>(group_timepoint1_<span class="dv">0</span> <span class="op">+</span><span class="st"> </span>group_timepoint0_<span class="dv">0</span>) <span class="op">/</span><span class="st"> </span><span class="dv">2</span>,</a>
<a class="sourceLine" id="cb2-34" data-line-number="34">  <span class="dt">baseline_groups =</span> (group_timepoint1_<span class="dv">0</span> <span class="op">-</span><span class="st"> </span>group_timepoint0_<span class="dv">0</span>),</a>
<a class="sourceLine" id="cb2-35" data-line-number="35">  <span class="dt">levels =</span> <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(designmat)</a>
<a class="sourceLine" id="cb2-36" data-line-number="36">)</a>
<a class="sourceLine" id="cb2-37" data-line-number="37">  </a>
<a class="sourceLine" id="cb2-38" data-line-number="38"><span class="co"># fit mixed model for the multi timepoint contrasts </span></a>
<a class="sourceLine" id="cb2-39" data-line-number="39">fit =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/dreamMixedModel.html">dreamMixedModel</a></span>(<span class="dt">dge_lists =</span> dge, </a>
<a class="sourceLine" id="cb2-40" data-line-number="40">                               <span class="dt">apriori_contrasts =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb2-41" data-line-number="41">                               <span class="dt">sample_column =</span> <span class="st">'sample'</span>,</a>
<a class="sourceLine" id="cb2-42" data-line-number="42">                               <span class="dt">cell_metadata =</span> meta, </a>
<a class="sourceLine" id="cb2-43" data-line-number="43">                               <span class="dt">contrast_matrix =</span> c_mat, </a>
<a class="sourceLine" id="cb2-44" data-line-number="44">                               <span class="dt">design_matrix =</span> designmat, </a>
<a class="sourceLine" id="cb2-45" data-line-number="45">                               <span class="dt">lme4_formula =</span>  <span class="st">'~ 0 + age + gender + cohort_timepoint + (1|sampleid)'</span>, </a>
<a class="sourceLine" id="cb2-46" data-line-number="46">                               <span class="dt">fixed_effects =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'age'</span>, <span class="st">'gender'</span>, <span class="st">'cohort_timepoint'</span>), </a>
<a class="sourceLine" id="cb2-47" data-line-number="47">                               <span class="dt">plotsavepath =</span> figpath, </a>
<a class="sourceLine" id="cb2-48" data-line-number="48">                               <span class="dt">ncores =</span> <span class="dv">4</span>)</a>
<a class="sourceLine" id="cb2-49" data-line-number="49"></a>
<a class="sourceLine" id="cb2-50" data-line-number="50"><span class="co"># fit simple linear model for the baseline group level contrast </span></a>
<a class="sourceLine" id="cb2-51" data-line-number="51">bl =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/RunVoomLimma.html">RunVoomLimma</a></span>(<span class="dt">dgelists =</span> dge, </a>
<a class="sourceLine" id="cb2-52" data-line-number="52">                           <span class="dt">design_matrix =</span> designmat, </a>
<a class="sourceLine" id="cb2-53" data-line-number="53">                           <span class="dt">do_contrast_fit =</span> T,</a>
<a class="sourceLine" id="cb2-54" data-line-number="54">                           <span class="dt">my_contrast_matrix =</span> c_mat[ ,<span class="dv">3</span>])</a>
<a class="sourceLine" id="cb2-55" data-line-number="55"></a>
<a class="sourceLine" id="cb2-56" data-line-number="56"><span class="co"># save </span></a>
<a class="sourceLine" id="cb2-57" data-line-number="57"><span class="kw"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span>(bl, <span class="dt">file =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(datapath, <span class="st">"blfit_object.rds"</span>))</a>
<a class="sourceLine" id="cb2-58" data-line-number="58"><span class="kw"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span>(fit, <span class="dt">file =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(datapath, <span class="st">"fit_object.rds"</span>))</a></code></pre></div>
</div>
<div id="downstream-analysis-on-pseudobulk-results" class="section level3">
<h3 class="hasAnchor">
<a href="#downstream-analysis-on-pseudobulk-results" class="anchor"></a>Downstream analysis on pseudobulk results</h3>
<p>Here we provide examples of downstream analysis on results of the model fit above focusing on the first contrast (the first coefficient in the results), corresponding to:<br>
foldchange_difference = (group_timepoint1_1 - group_timepoint1_0) - (group_timepoint0_1 - group_timepoint0_0)</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># hlmk = readRDS(file = here("signature_curation/hallmark.rds"))</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">figpath =<span class="st"> "your/path"</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3">test =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/GetRankResults.html">GetRankResults</a></span>(<span class="dt">limma.fit.object.list =</span> bl, <span class="dt">coefficient.number =</span> <span class="dv">1</span>, <span class="st">"test"</span>)</a>
<a class="sourceLine" id="cb3-4" data-line-number="4">res =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/GetContrastResults.html">GetContrastResults</a></span>(<span class="dt">limma.fit.object.list =</span> bl, <span class="dt">coefficient.number =</span> <span class="dv">1</span>, <span class="dt">contrast.name =</span> <span class="st">"test"</span>)</a>
<a class="sourceLine" id="cb3-5" data-line-number="5"></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="co"># get the mixed effects model results for the difference in fold changes </span></a>
<a class="sourceLine" id="cb3-7" data-line-number="7">fit_res =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/GetContrastResultsRaw.html">GetContrastResultsRaw</a></span>(<span class="dt">limma.fit.object.list =</span> fit, </a>
<a class="sourceLine" id="cb3-8" data-line-number="8">                                         <span class="dt">coefficient.number =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb3-9" data-line-number="9">                                         <span class="dt">contrast.name =</span> <span class="st">"foldchangedifference"</span>)</a>
<a class="sourceLine" id="cb3-10" data-line-number="10">fit_rank =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/GetRankResultsRaw.html">GetRankResultsRaw</a></span>(<span class="dt">contrast.result.raw.list =</span> fit_res)</a>
<a class="sourceLine" id="cb3-11" data-line-number="11"></a>
<a class="sourceLine" id="cb3-12" data-line-number="12"></a>
<a class="sourceLine" id="cb3-13" data-line-number="13"><span class="co"># Gene Set Enrichment Analysis </span></a>
<a class="sourceLine" id="cb3-14" data-line-number="14">gsea1 =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/RunFgseaOnRankList.html">RunFgseaOnRankList</a></span>(<span class="dt">rank.list.celltype =</span> test, <span class="dt">pathways =</span> hlmk)</a>
<a class="sourceLine" id="cb3-15" data-line-number="15">d =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/RbindGseaResultList.html">RbindGseaResultList</a></span>(<span class="dt">gsea_result_list =</span> gsea1,<span class="dt">NES_filter =</span> <span class="op">-</span><span class="ot">Inf</span>,<span class="dt">padj_filter =</span> <span class="fl">0.2</span>)</a>
<a class="sourceLine" id="cb3-16" data-line-number="16">scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/GSEABubblePlot.html">GSEABubblePlot</a></span>(d, <span class="dt">save_path =</span> figpath, <span class="dt">save_name =</span> <span class="st">"plot.pdf"</span>)</a>
<a class="sourceLine" id="cb3-17" data-line-number="17"><span class="co"># also see scglmmr::GSEABarPlot()</span></a>
<a class="sourceLine" id="cb3-18" data-line-number="18"></a>
<a class="sourceLine" id="cb3-19" data-line-number="19"><span class="co"># full set of leading edge genes vs celltypes </span></a>
<a class="sourceLine" id="cb3-20" data-line-number="20">lefull =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/GetLeadingEdgeFull.html">GetLeadingEdgeFull</a></span>(<span class="dt">gsea.list =</span> gsea1, <span class="dt">padj.filter =</span> <span class="fl">0.1</span>,<span class="dt">NES.filter =</span> <span class="op">-</span><span class="ot">Inf</span>)</a>
<a class="sourceLine" id="cb3-21" data-line-number="21"></a>
<a class="sourceLine" id="cb3-22" data-line-number="22"><span class="co"># combine GSEA results with model coefficient for each gene in leading edge </span></a>
<a class="sourceLine" id="cb3-23" data-line-number="23">cr =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/CombineResults.html">CombineResults</a></span>(<span class="dt">gsealist =</span> gsea1, <span class="dt">contrastlist =</span> res, <span class="dt">gseafdr =</span> <span class="fl">0.1</span>, <span class="dt">genefdr =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb3-24" data-line-number="24"></a>
<a class="sourceLine" id="cb3-25" data-line-number="25"></a>
<a class="sourceLine" id="cb3-26" data-line-number="26"><span class="co"># make tidy average data for visualization of weighted pb results </span></a>
<a class="sourceLine" id="cb3-27" data-line-number="27">av =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/PseudobulkList.html">PseudobulkList</a></span>(<span class="dt">rawcounts =</span> umi, </a>
<a class="sourceLine" id="cb3-28" data-line-number="28">                             <span class="dt">metadata =</span> meta, </a>
<a class="sourceLine" id="cb3-29" data-line-number="29">                             <span class="dt">sample_col =</span> <span class="st">"sample"</span>, </a>
<a class="sourceLine" id="cb3-30" data-line-number="30">                             <span class="dt">celltype_col =</span> <span class="st">"celltype"</span>,</a>
<a class="sourceLine" id="cb3-31" data-line-number="31">                             <span class="dt">avg_or_sum =</span> <span class="st">'average'</span>)</a>
<a class="sourceLine" id="cb3-32" data-line-number="32">le_expr =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/LeadEdgeTidySampleExprs.html">LeadEdgeTidySampleExprs</a></span>(<span class="dt">av.exprs.list =</span> av,</a>
<a class="sourceLine" id="cb3-33" data-line-number="33">                                           <span class="dt">gsea.list =</span> hlmk_ctm0, </a>
<a class="sourceLine" id="cb3-34" data-line-number="34">                                           <span class="dt">padj.filter =</span> <span class="fl">0.1</span>,</a>
<a class="sourceLine" id="cb3-35" data-line-number="35">                                           <span class="dt">NES.filter =</span> <span class="op">-</span><span class="ot">Inf</span>)</a>
<a class="sourceLine" id="cb3-36" data-line-number="36"></a>
<a class="sourceLine" id="cb3-37" data-line-number="37"></a>
<a class="sourceLine" id="cb3-38" data-line-number="38"><span class="co"># example plot of sample level average leading edge genes annotated </span></a>
<a class="sourceLine" id="cb3-39" data-line-number="39">scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/LeadEdgeSampleHeatmap.html">LeadEdgeSampleHeatmap</a></span>(<span class="dt">tidy.exprs.list =</span> le_expr,</a>
<a class="sourceLine" id="cb3-40" data-line-number="40">                               <span class="dt">modulename =</span> <span class="st">"MODULENAME"</span>,</a>
<a class="sourceLine" id="cb3-41" data-line-number="41">                               <span class="dt">elltype_plot =</span> <span class="st">"TCELL"</span>,</a>
<a class="sourceLine" id="cb3-42" data-line-number="42">                               <span class="dt">metadata =</span> meta, </a>
<a class="sourceLine" id="cb3-43" data-line-number="43">                               <span class="dt">metadata_annotate =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'group'</span>, <span class="st">'timepoint'</span>, <span class="st">'age'</span>, <span class="st">'gender'</span>),</a>
<a class="sourceLine" id="cb3-44" data-line-number="44">                               <span class="dt">sample_column =</span> <span class="st">'sample'</span>,</a>
<a class="sourceLine" id="cb3-45" data-line-number="45">                               <span class="dt">returnmat =</span> F, </a>
<a class="sourceLine" id="cb3-46" data-line-number="46">                               <span class="dt">savepath =</span> figpath, </a>
<a class="sourceLine" id="cb3-47" data-line-number="47">                               <span class="dt">savename =</span> <span class="st">"filename"</span>)</a>
<a class="sourceLine" id="cb3-48" data-line-number="48"></a>
<a class="sourceLine" id="cb3-49" data-line-number="49"><span class="co"># see also: </span></a>
<a class="sourceLine" id="cb3-50" data-line-number="50"><span class="co"># scglmmr::TopGenesTidySampleExprs() - same as aobve for 'top' de genes estimated by model coeffcient / p value. </span></a>
<a class="sourceLine" id="cb3-51" data-line-number="51"><span class="co"># scglmmr::GetTidySummary() - for custom plotting</span></a>
<a class="sourceLine" id="cb3-52" data-line-number="52"></a>
<a class="sourceLine" id="cb3-53" data-line-number="53"></a>
<a class="sourceLine" id="cb3-54" data-line-number="54"><span class="co"># repeat this for all enriched pathways </span></a>
<a class="sourceLine" id="cb3-55" data-line-number="55">heatpath =<span class="st"> </span><span class="kw">here</span>(<span class="st">"sime/path"</span>); <span class="kw"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span>(heatpath)</a>
<a class="sourceLine" id="cb3-56" data-line-number="56"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(le_expr)) {</a>
<a class="sourceLine" id="cb3-57" data-line-number="57">  cdat =<span class="st"> </span>le_expr[[i]]</a>
<a class="sourceLine" id="cb3-58" data-line-number="58">  ctype =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(le_expr[i])</a>
<a class="sourceLine" id="cb3-59" data-line-number="59">  umod =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(cdat<span class="op">$</span>module)</a>
<a class="sourceLine" id="cb3-60" data-line-number="60">  <span class="cf">for</span> (u <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(umod)) {</a>
<a class="sourceLine" id="cb3-61" data-line-number="61">    scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/LeadEdgeSampleHeatmap.html">LeadEdgeSampleHeatmap</a></span>(<span class="dt">tidy.exprs.list =</span> le_expr, </a>
<a class="sourceLine" id="cb3-62" data-line-number="62">                          <span class="dt">modulename =</span> umod[u], </a>
<a class="sourceLine" id="cb3-63" data-line-number="63">                          <span class="dt">celltype_plot =</span> ctype,</a>
<a class="sourceLine" id="cb3-64" data-line-number="64">                          <span class="dt">metadata =</span> meta, <span class="dt">metadata_annotate =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'group'</span>, <span class="st">'timepoint'</span>, <span class="st">'age'</span>, <span class="st">'gender'</span>),</a>
<a class="sourceLine" id="cb3-65" data-line-number="65">                          <span class="dt">sample_column =</span> <span class="st">'sample'</span>,</a>
<a class="sourceLine" id="cb3-66" data-line-number="66">                          <span class="dt">returnmat =</span> F, </a>
<a class="sourceLine" id="cb3-67" data-line-number="67">                          <span class="dt">savepath =</span> heatpath,</a>
<a class="sourceLine" id="cb3-68" data-line-number="68">                          <span class="dt">savename =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(ctype, <span class="st">" "</span>,umod[u],<span class="st">'.pdf'</span>))</a>
<a class="sourceLine" id="cb3-69" data-line-number="69">  }</a>
<a class="sourceLine" id="cb3-70" data-line-number="70">}</a>
<a class="sourceLine" id="cb3-71" data-line-number="71"></a>
<a class="sourceLine" id="cb3-72" data-line-number="72"><span class="co"># plot only the leading edge genes from enrichment in a heatmap of log Fold Change estimated for a model contrast </span></a>
<a class="sourceLine" id="cb3-73" data-line-number="73">le =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/GetLeadingEdgeFull.html">GetLeadingEdgeFull</a></span>(<span class="dt">gsea.list =</span> gsea1, <span class="dt">padj.filter =</span> <span class="fl">0.1</span>, <span class="dt">NES.filter =</span> <span class="op">-</span><span class="ot">Inf</span>)</a>
<a class="sourceLine" id="cb3-74" data-line-number="74">genesub =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(rbind, le) <span class="op">%$%</span><span class="st"> </span>gene <span class="op">%&gt;%</span><span class="st"> </span>unique </a>
<a class="sourceLine" id="cb3-75" data-line-number="75">mtx2 =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/GetGeneMatrix.html">GetGeneMatrix</a></span>(<span class="dt">result.list =</span> res, </a>
<a class="sourceLine" id="cb3-76" data-line-number="76">                    <span class="dt">stat_for_matrix =</span> <span class="st">"logFC"</span>,</a>
<a class="sourceLine" id="cb3-77" data-line-number="77">                    <span class="dt">gene_subset =</span> genesub, </a>
<a class="sourceLine" id="cb3-78" data-line-number="78">                    <span class="dt">pvalfilter =</span> <span class="op">-</span><span class="ot">Inf</span>, </a>
<a class="sourceLine" id="cb3-79" data-line-number="79">                    <span class="dt">logfcfilter =</span> <span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb3-80" data-line-number="80"></a>
<a class="sourceLine" id="cb3-81" data-line-number="81">pheatmap<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span>(mtx2,</a>
<a class="sourceLine" id="cb3-82" data-line-number="82">                   <span class="dt">breaks =</span><span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="dt">from =</span> <span class="dv">0</span>, <span class="dt">to =</span> <span class="dv">2</span>,<span class="dt">length.out =</span> <span class="dv">99</span>),</a>
<a class="sourceLine" id="cb3-83" data-line-number="83">                   <span class="dt">filename =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(figpath,<span class="st">"LEgenes_heatmap.pdf"</span>))</a>
<a class="sourceLine" id="cb3-84" data-line-number="84"></a>
<a class="sourceLine" id="cb3-85" data-line-number="85"></a>
<a class="sourceLine" id="cb3-86" data-line-number="86"></a>
<a class="sourceLine" id="cb3-87" data-line-number="87"><span class="co">### Hypergeometric erichment </span></a>
<a class="sourceLine" id="cb3-88" data-line-number="88"><span class="kw"><a href="https://rdrr.io/r/base/load.html">load</a></span>(termdf) <span class="co"># this term2gene dataframe is included in the package see clusterProfiler</span></a>
<a class="sourceLine" id="cb3-89" data-line-number="89">hyp =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/RunHypergeometricTest.html">RunHypergeometricTest</a></span>(<span class="dt">result_list =</span> fit_res,</a>
<a class="sourceLine" id="cb3-90" data-line-number="90">                                     <span class="dt">TERM2GENE_dataframe =</span> termdf,</a>
<a class="sourceLine" id="cb3-91" data-line-number="91">                                     <span class="dt">pval_threshold =</span> <span class="fl">0.1</span>,</a>
<a class="sourceLine" id="cb3-92" data-line-number="92">                                     <span class="dt">logFC_threshold =</span> <span class="dv">0</span>,</a>
<a class="sourceLine" id="cb3-93" data-line-number="93">                                     <span class="dt">usefdr_threshold =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb3-94" data-line-number="94"><span class="co"># plot results </span></a>
<a class="sourceLine" id="cb3-95" data-line-number="95">scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/PlotHypergeometric.html">PlotHypergeometric</a></span>(<span class="dt">hyperg_result =</span> hyp, </a>
<a class="sourceLine" id="cb3-96" data-line-number="96">                            <span class="dt">p.adjust.filter =</span> <span class="fl">0.1</span>,</a>
<a class="sourceLine" id="cb3-97" data-line-number="97">                            <span class="dt">genenumber_filter =</span> <span class="dv">2</span>,</a>
<a class="sourceLine" id="cb3-98" data-line-number="98">                            <span class="dt">savepath =</span> figpath,</a>
<a class="sourceLine" id="cb3-99" data-line-number="99">                            <span class="dt">savename =</span> <span class="st">"name"</span>, </a>
<a class="sourceLine" id="cb3-100" data-line-number="100">                            <span class="dt">title =</span> <span class="st">"title"</span>)</a>
<a class="sourceLine" id="cb3-101" data-line-number="101"></a>
<a class="sourceLine" id="cb3-102" data-line-number="102"><span class="co"># calculate average module z score across samples (z score across samples of gene z scores)</span></a>
<a class="sourceLine" id="cb3-103" data-line-number="103"><span class="co"># baseline_samples = c('') # user defined vector </span></a>
<a class="sourceLine" id="cb3-104" data-line-number="104"><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(av, <span class="cf">function</span>(x), x[ ,baseline_samples])</a>
<a class="sourceLine" id="cb3-105" data-line-number="105">mz =<span class="st"> </span>scglmmr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scglmmr/man/AverageSampleModuleZscore.html">AverageSampleModuleZscore</a></span>(<span class="dt">average.metacell.list =</span> av,</a>
<a class="sourceLine" id="cb3-106" data-line-number="106">                                        <span class="dt">module.list =</span> hlmk,<span class="dt">use.module.subset =</span> F)</a></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Matthew Mulè.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
