<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Single cell gene and module level differential expression with nested group repeated measures experiment designs using scglmmr • scglmmr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Single cell gene and module level differential expression with nested group repeated measures experiment designs using scglmmr">
<meta property="og:description" content="scglmmr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scglmmr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/scglmmr_pseudobulk.html">Pseudobulk differential expression with nested group repeated measures single cell experiment designs using scglmmr</a>
    </li>
    <li>
      <a href="../articles/scglmmr_singlecell.html">Single cell gene and module level differential expression with nested group repeated measures experiment designs using scglmmr</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MattPM/scglmmr/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Single cell gene and module level differential expression with nested group repeated measures experiment designs using scglmmr</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MattPM/scglmmr/blob/master/vignettes/scglmmr_singlecell.Rmd"><code>vignettes/scglmmr_singlecell.Rmd</code></a></small>
      <div class="hidden name"><code>scglmmr_singlecell.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">scglmmr</span>)</pre></body></html></div>
<div id="single-cell-gene-level-de-testing" class="section level3">
<h3 class="hasAnchor">
<a href="#single-cell-gene-level-de-testing" class="anchor"></a>2. Single cell gene level DE testing</h3>
<p>This single cell level models of differential expression pre and post treatment testing of genes by fitting a Poisson mixed glm. You specify a reduced and full model formula to extract p values and effect sizes.</p>
<p>Recommended: The function accepts an argument ‘celltype_genes_test’ this is a named list indexed by celltypes.</p>
<p>First we load a single cell object an define highly variable genes within each celltype, this can be done many different ways.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="co"># Single cell mixed model. </span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressMessages</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tidyverse</span>))
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">scglmmr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1990</span>)

<span class="no">datapath</span> <span class="kw">=</span> <span class="st">'your_path_to_save_results'</span>
<span class="no">s</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span>(<span class="st">'path_to_seurat_or_SingleCellExperimentObject'</span>)

<span class="co"># define highly variable genes within each cell type </span>
<span class="no">gene_list</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>()
<span class="kw">for</span> (<span class="no">i</span> <span class="kw">in</span> <span class="fl">1</span>:<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">s</span>@<span class="kw">meta.data</span>$<span class="no">celltype</span>))) {
  <span class="no">gene_list</span><span class="kw">[[</span><span class="no">i</span>]] <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(<span class="no">s</span>, <span class="no">celltype_joint</span> <span class="kw">==</span> <span class="no">celltype_unique</span>[<span class="no">i</span>]) <span class="kw">%&gt;%</span>
    <span class="kw pkg">Seurat</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindVariableFeatures.html">FindVariableFeatures</a></span>(<span class="no">x</span><span class="kw">[[</span><span class="no">i</span>]],<span class="kw">top.genes</span> <span class="kw">=</span> <span class="fl">10000</span>)
  <span class="no">gene_list</span><span class="kw">[[</span><span class="no">i</span>]] <span class="kw">=</span> <span class="kw pkg">Seurat</span><span class="kw ns">::</span><span class="fu">VariableGenes</span>(<span class="no">gene_list</span><span class="kw">[[</span><span class="no">i</span>]])
}

<span class="co"># one can also use scran trendVar etc. </span></pre></body></html></div>
<p>Now we run the single cell Poisson model by specifying a model formula, the formula must have the LHS equal to <code>gene ~</code> and include a random effect term <code>(1|subjectid)</code>. the variable that corresponds to the perturbation variable as a 2 level factor with pre and post perturbation measurements (repeated measurements for the same subject denoted by <code>subjectid</code> must be identified, for example, if the metadata variable coding for the time post perturbation is called treatment.time, then the model formula could be <code>f1 = 'gene ~ offset(log(nUMI)) + treatment.time + (1|subjectid)'</code> with this model one would set the argument <code>test_variable = treatment.time</code> which should have 2 ordered levels corresponding to pre and post treatment. Within the function, the <code>emmeans</code> package automatically tests the difference in the marginal means of the post vs pre perturbation effect.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="co"># load subset of genes to test for each celltype </span>
<span class="no">gene_union</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="no">genes_test</span>))

<span class="co"># create day 1 metadata dataframe from a single cell object (seurat, SingleCellExperiment etc.)</span>
<span class="no">s</span>@<span class="kw">meta.data</span>$<span class="no">subjectid</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">s</span>@<span class="kw">meta.data</span>$<span class="no">sampleid</span>))

<span class="co"># create metadata for lme4</span>
<span class="no">md</span> <span class="kw">=</span> <span class="no">s</span>@<span class="kw">meta.data</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/droplevels.html">droplevels</a></span>() <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="no">barcode_check</span>, <span class="no">batch</span>, <span class="no">subjectid</span>, <span class="no">timepoint</span>, <span class="no">celltype_joint</span>, <span class="no">nUMI</span>) <span class="kw">%&gt;%</span>
  <span class="fu">mutate</span>(<span class="kw">timepoint</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">timepoint</span>, <span class="kw">levels</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"d0"</span>, <span class="st">"d1"</span>))) <span class="kw">%&gt;%</span>
  <span class="fu">rename</span>(<span class="kw">celltype</span> <span class="kw">=</span> <span class="no">celltype_joint</span>) <span class="kw">%&gt;%</span>
  <span class="fu">column_to_rownames</span>(<span class="st">'barcode_check'</span>)

<span class="co"># get gene data </span>
<span class="no">gene_data</span> <span class="kw">=</span> <span class="kw pkg">Matrix</span><span class="kw ns">::</span><span class="fu">t</span>(<span class="no">s</span>@<span class="kw">raw.data</span>[<span class="no">gene_union</span>, ])

<span class="co"># we can now remove whatever single cell object was loaded in the environment. </span>
<span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span>(<span class="no">s</span>); <span class="fu"><a href="https://rdrr.io/r/base/gc.html">gc</a></span>()

<span class="co">## specify model formula </span>
<span class="no">f1</span> <span class="kw">=</span> <span class="st">'gene ~ offset(log(nUMI)) + timepoint + (1|subjectid)'</span>

<span class="co"># method 1</span>
<span class="no">results</span> <span class="kw">=</span> <span class="kw pkg">scglmmr</span><span class="kw ns">::</span><span class="fu"><a href="../reference/SCMixedPoisson.html">SCMixedPoisson</a></span>(<span class="kw">gene_data</span> <span class="kw">=</span> <span class="no">gene_data</span>,
                                  <span class="kw">metadata</span> <span class="kw">=</span> <span class="no">md</span>,
                                  <span class="kw">model_formula</span> <span class="kw">=</span> <span class="no">f1</span>,
                                  <span class="kw">test_variable</span> <span class="kw">=</span> <span class="st">'timepoint'</span>,
                                  <span class="kw">celltype_genes_test</span> <span class="kw">=</span> <span class="no">genes_test</span>,
                                  <span class="kw">save_path</span> <span class="kw">=</span> <span class="no">datapath</span>)
<span class="co"># write results </span>
<span class="co"># data.table::fwrite(x = results,file = paste0(datapath, "/h1_sc_time_merged_v2.txt"), sep = '\t')</span></pre></body></html></div>
<p>Another way to specify the model automatically:</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">results</span> <span class="kw">=</span> <span class="kw pkg">scglmmr</span><span class="kw ns">::</span><span class="fu"><a href="../reference/SCMixedPoisson.html">SCMixedPoisson</a></span>(<span class="kw">gene_data</span> <span class="kw">=</span> <span class="no">gene_data</span>,
                                  <span class="kw">metadata</span> <span class="kw">=</span> <span class="no">md</span>,
                                  <span class="kw">test_variable</span> <span class="kw">=</span> <span class="st">'timepoint'</span>,
                                  <span class="kw">covariate_variables</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'BMI'</span>, <span class="st">'SEX'</span>),
                                  <span class="kw">celltype_genes_test</span> <span class="kw">=</span> <span class="no">genes_test</span>,
                                  <span class="kw">save_path</span> <span class="kw">=</span> <span class="no">datapath</span>)</pre></body></html></div>
<p>This would automatically test the delta for pre and post perturbation calculating the marginal means over the coavariates BMI and SEX with the generalized linear mixed effects Poisson model: <code>'gene ~ BMI + SEX + timepoint + (1|subjectid)'</code>.</p>
</div>
<div id="enrichment-of-single-cell-poisson-model-of-timepoint-effect-using-clusterprofiler" class="section level3">
<h3 class="hasAnchor">
<a href="#enrichment-of-single-cell-poisson-model-of-timepoint-effect-using-clusterprofiler" class="anchor"></a>enrichment of single cell poisson model of timepoint effect using clusterprofiler</h3>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="co">## enrichment on these results </span>

<span class="co">### plot enrichment with a hypergeometric test against the li BTM </span>
<span class="no">scd1</span> <span class="kw">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span>( <span class="no">results</span> , <span class="kw">f</span> <span class="kw">=</span> <span class="no">mmres</span>$<span class="no">celltype</span> )
<span class="no">scd1</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">scd1</span>, <span class="kw">function</span>(<span class="no">x</span>){<span class="no">x</span> <span class="kw">%&gt;%</span>
    <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="no">gene</span>, <span class="no">celltype</span>, <span class="no">estimate</span>, <span class="no">padj</span>)  <span class="kw">%&gt;%</span>
    <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span>(<span class="st">'logFC'</span><span class="kw">=</span> <span class="no">fixed_effect</span>,  <span class="st">'P.Value'</span> <span class="kw">=</span> <span class="no">padj</span>) })


<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span>(<span class="no">termdf</span>)
<span class="no">hypsc</span> <span class="kw">=</span> <span class="kw pkg">scglmmr</span><span class="kw ns">::</span><span class="fu"><a href="../reference/RunHypergeometricTest.html">RunHypergeometricTest</a></span>(<span class="kw">result_list</span> <span class="kw">=</span> <span class="no">scd1</span>,
                                       <span class="kw">TERM2GENE_dataframe</span> <span class="kw">=</span> <span class="no">term_df</span>,
                                       <span class="kw">pval_threshold</span> <span class="kw">=</span> <span class="fl">0.05</span>,
                                       <span class="kw">logFC_threshold</span> <span class="kw">=</span> <span class="fl">0.1</span>)
<span class="kw pkg">scglmmr</span><span class="kw ns">::</span><span class="fu"><a href="../reference/PlotHypergeometric.html">PlotHypergeometric</a></span>(<span class="kw">hyperg_result</span> <span class="kw">=</span> <span class="no">hypsc</span>,
                            <span class="kw">p.adjust.filter</span> <span class="kw">=</span> <span class="fl">0.05</span>,
                            <span class="kw">genenumber_filter</span> <span class="kw">=</span> <span class="fl">2</span>,
                            <span class="kw">savepath</span> <span class="kw">=</span> <span class="no">figpath</span>,
                            <span class="kw">savename</span> <span class="kw">=</span> <span class="st">"d1singlecell_hypergeometric_padj0.05.pdf"</span>,
                            <span class="kw">title</span> <span class="kw">=</span> <span class="st">"single cell 24h vs baseline gene_exprs ~ offset(nUMI) + timepoint + (1|subjectid) \n 
                            hypergeometric test LI BTM, FDR 0.05, gene filter n &gt; 2"</span>,
                            <span class="kw">height</span> <span class="kw">=</span> <span class="fl">5.5</span>, <span class="kw">width</span> <span class="kw">=</span> <span class="fl">8</span>)</pre></body></html></div>
</div>
<div id="single-cell-module-level-testing-group-level-fold-change-comparisons" class="section level3">
<h3 class="hasAnchor">
<a href="#single-cell-module-level-testing-group-level-fold-change-comparisons" class="anchor"></a>2. Single cell module level testing + group level fold change comparisons</h3>
<p>This is designed for testing the difference in single cell module activity scores between groups at baseline, between groups, post treatment across groups and the difference in treatment effect between groups using the same <em>a priori</em> contrast approach as in the pseudobulk method described above.</p>
<div id="part-i" class="section level4">
<h4 class="hasAnchor">
<a href="#part-i" class="anchor"></a>Part I</h4>
<p>You first fit a module score to each single cell. Options: <em>threshold</em> what percent of genes in the module must a cell express to get a score (i.e. of 0.1, if the cell has less than 10% of genes &gt;0 in the module don’t score) <em>return_weighted</em> whether to return the weighted module score which is the average * weight where weight = number of genes in the cell in that module with non-zeroexpression</p>
<p><em>cellwise_scaling</em> should the scores be scaled across cells? Not recommended unless you are testing only a single celltype. <em>Seurat_version</em> if this is set to “2” it will accomodate seurat v2, if set to “3” will accomodate v3</p>
</div>
<div id="part-ii-fitting-models" class="section level4">
<h4 class="hasAnchor">
<a href="#part-ii-fitting-models" class="anchor"></a>Part II fitting models:</h4>
<p><em>The SCGroupContrastGLMM Function</em> To simultaneously test the difference in the fold changes due to treatment (for example) the levels of the combined group + time factor "group_id have to be ordered as follows (can be any names, these are just the order): group1 timepoint 0, group1 timepoint 1, group2 timepoint 0, group2 timepoint 1</p>
<p>It is also possible to use custom contrasts by changing the argument to contrast_list from contrast_2 (automatically loaded) to something else.</p>
<p>Specify any lme4 model formula in the argument to f1, with the randomeffect term for subjectid set to <code>(1|sampleid)</code> which should have a pre and post perturbation measurement. <code>f1 = 'modulescore ~ age + gender +  group_id + (1|sampleid)'</code></p>
<p>This function will also automatically save a helpful plot of the least squares means (the residual means of each group after accounting for covariates in the model) of the module over the levels of each combined group_timepoint factor.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r">
<span class="co"># load data</span>
<span class="no">Seurat</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span>(<span class="st">"my_seurat_object.rds"</span>)

<span class="co"># add cellwise module score for each signature </span>
<span class="no">mod_scores</span> <span class="kw">=</span> <span class="fu"><a href="../reference/WeightedCellModuleScore.html">WeightedCellModuleScore</a></span>(<span class="kw">seurat_object</span> <span class="kw">=</span> <span class="no">Seurat</span>,
                                     <span class="kw">module_list</span> <span class="kw">=</span> <span class="no">btm</span>,
                                     <span class="kw">threshold</span> <span class="kw">=</span> <span class="fl">0.1</span>,
                                     <span class="kw">return_weighted</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">cellwise_scaling</span> <span class="kw">=</span> <span class="fl">FALSE</span>,
                                     <span class="kw">Seurat_version</span> <span class="kw">=</span> <span class="st">"2"</span>)
<span class="no">Seurat</span> <span class="kw">=</span> <span class="fu">AddMetaData</span>(<span class="no">Seurat</span>,<span class="kw">metadata</span> <span class="kw">=</span> <span class="no">mod_scores</span>)
<span class="no">module_n</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">sig_test</span>)

<span class="co"># set up module data frame </span>
<span class="no">module_df</span> <span class="kw">=</span> <span class="no">Seurat</span>@<span class="kw">meta.data</span> <span class="kw">%&gt;%</span> <span class="fu">select</span>(<span class="no">barcode_check</span>, <span class="no">celltype_joint</span>, <span class="no">module_n</span>)

<span class="co"># format metadata as factors group_id is order leveled for contrast_fit = contrast(emm1, method = list( (c21 - c20) - (c11 - c10) ))</span>
<span class="no">md</span> <span class="kw">=</span> <span class="no">Seurat</span>@<span class="kw">meta.data</span> <span class="kw">%&gt;%</span>
  <span class="fu">mutate</span>(<span class="kw">group_id</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">adjmfc.time</span>,  <span class="kw">levels</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'d0 low'</span>, <span class="st">'d1 low'</span>, <span class="st">'d0 high'</span>, <span class="st">'d1 high'</span>))) <span class="kw">%&gt;%</span>
  <span class="fu">mutate</span>(<span class="kw">sampleid</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">sampleid</span>)) <span class="kw">%&gt;%</span>
  <span class="fu">select</span>(<span class="no">barcode_check</span>, <span class="no">celltype_joint</span>, <span class="no">sampleid</span>,  <span class="no">age</span>, <span class="no">group_id</span>)

<span class="co"># Fit mixed model </span>
<span class="no">plot_savepath</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">my_figure_save_path</span>, <span class="st">"/marginalmeans/"</span>); <span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span>(<span class="no">plot_savepath</span>)

<span class="co"># specify any random intercept model e.g.</span>
<span class="no">f1</span> <span class="kw">=</span> <span class="st">'modulescore ~ age + group_id + (1|sampleid)'</span>

<span class="co"># fit sc mod mixed model on ewighted module scores. </span>
<span class="no">mm_res</span> <span class="kw">=</span> <span class="fu">FitModuleMixModel</span>(<span class="kw">module_data_frame</span> <span class="kw">=</span> <span class="no">module_df</span>,
                           <span class="kw">celltype_column</span> <span class="kw">=</span> <span class="st">'celltype'</span>,
                           <span class="kw">metadata</span> <span class="kw">=</span> <span class="no">md</span>,
                           <span class="kw">fixed_effects</span> <span class="kw">=</span> <span class="kw">NULL</span>,
                           <span class="kw">lmer_formula</span> <span class="kw">=</span> <span class="no">f1</span>,
                           <span class="kw">plotdatqc</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
                           <span class="kw">figpath</span> <span class="kw">=</span> <span class="st">'your/file/path'</span>)

<span class="co">### This function returns p values, effect sizes for baseline and difference in fld changes between the groups </span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">mmres</span>)
<span class="co"># "celltype"                    "modulename"                  "contrasttime1vs0_group2vs1" </span>
<span class="co"># "estimatetime1vs0_group2vs1"  "std.errortime1vs0_group2vs1" "dftime1vs0_group2vs1"       </span>
<span class="co"># "z.ratiotime1vs0_group2vs1"   "p.valuetime1vs0_group2vs1"   "contrasttime0_group2vs1"    </span>
<span class="co"># "estimatetime0_group2vs1"     "std.errortime0_group2vs1"    "dftime0_group2vs1"          </span>
<span class="co"># "z.ratiotime0_group2vs1"      "p.valuetime0_group2vs1"      "d0 low_marginal_mean"       </span>
<span class="co"># "d1 low_marginal_mean"        "d0 high_marginal_mean"       "d1 high_marginal_mean"      </span>
<span class="co"># "formula"                     "statistictime1vs0_group2vs1" "statistictime0_group2vs1"   </span></pre></body></html></div>
<p>A plot like this will be generated for each module for each cell type with the estimated marginal means in the right margin and the actual single cell module score distribution across each level of group_id:</p>
<div class="figure">
<img src="https://user-images.githubusercontent.com/15280712/89591805-1f5a1200-d819-11ea-8b49-5b84477dd178.png" alt="image"><p class="caption">image</p>
</div>
<p><em>The WilcoxWithinCluster function</em> To run a simple test of baseline differences between groups, use the <em>WilcoxWithinCluster</em> function as below if you only have one timepoint but 2 outcome groups.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r">
<span class="co">######## baseline differences with a simple wilcox test </span>
<span class="no">weighted_score</span> <span class="kw">=</span> <span class="fu"><a href="../reference/WeightedCellModuleScore.html">WeightedCellModuleScore</a></span>(<span class="kw">seurat_object</span> <span class="kw">=</span> <span class="no">s</span>, <span class="kw">module_list</span> <span class="kw">=</span> <span class="no">sig</span>,
                                         <span class="kw">threshold</span> <span class="kw">=</span> <span class="fl">0.1</span>, <span class="kw">cellwise_scaling</span> <span class="kw">=</span> <span class="fl">FALSE</span>,
                                         <span class="kw">return_weighted</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
                                         <span class="kw">Seurat_version</span> <span class="kw">=</span> <span class="st">"3"</span>)

<span class="co"># set module dataframe</span>
<span class="no">module_df</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="kw">celltype_joint</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">s</span>@<span class="kw">meta.data</span>$<span class="no">seurat_clusters</span>),
                  <span class="kw">barcode_check</span> <span class="kw">=</span> <span class="no">s</span>@<span class="kw">meta.data</span>$<span class="no">barcode_full</span>,
                  <span class="no">weighted_score</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/droplevels.html">droplevels</a></span>()

<span class="co"># format lme4 metadata (note here just wilcox so group IDS do not have to be lme4 factors</span>
<span class="no">meta_data</span> <span class="kw">=</span> <span class="no">s</span>@<span class="kw">meta.data</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/droplevels.html">droplevels</a></span>() <span class="kw">%&gt;%</span>
  <span class="fu">select</span>(<span class="kw">celltype_joint</span> <span class="kw">=</span> <span class="no">seurat_clusters</span>,  <span class="kw">barcode_check</span> <span class="kw">=</span> <span class="no">barcode_full</span>, <span class="no">sample</span>, <span class="kw">group_id</span> <span class="kw">=</span> <span class="no">IRAE</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="no">group_id</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"poor_outcome"</span>, <span class="st">"good_outcome"</span>)) <span class="kw">%&gt;%</span>
  <span class="fu">mutate</span>(<span class="kw">group_id</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">group_id</span>, <span class="kw">levels</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"poor_outcome"</span>, <span class="st">"good_outcome"</span>)))

<span class="co"># Fit wilcox test and do bonferonni correction </span>
<span class="no">plot_savepath2</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">figpath</span>, <span class="st">"module_distribution/"</span>); <span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span>(<span class="no">plot_savepath2</span>)
<span class="no">m2</span> <span class="kw">=</span> <span class="no">module_df</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="no">barcode_check</span> <span class="kw">%in%</span> <span class="no">meta_data</span>$<span class="no">barcode_check</span> )
<span class="no">cvnc</span> <span class="kw">=</span> <span class="fu"><a href="../reference/WilcoxWithinCluster.html">WilcoxWithinCluster</a></span>(<span class="kw">module_data_frame</span> <span class="kw">=</span> <span class="no">m2</span>, <span class="kw">lme4metadata</span> <span class="kw">=</span> <span class="no">meta_data</span>, <span class="kw">plot_savepath</span> <span class="kw">=</span> <span class="no">plot_savepath2</span>)
<span class="no">cvnc</span>$<span class="no">padj</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span>(<span class="kw">p</span> <span class="kw">=</span> <span class="no">cvnc</span>$<span class="no">p.value.t0.wilcox</span>, <span class="kw">method</span> <span class="kw">=</span> <span class="st">"BH"</span>)</pre></body></html></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Matthew Mulè.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
