<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scglmmr">
<title>nested group repeated measures pseudobulk differential expression testing • scglmmr</title>
<script src="../deps/jquery-3.5.1/jquery-3.5.1.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="nested group repeated measures pseudobulk differential expression testing">
<meta property="og:description" content="scglmmr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scglmmr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/pseudobulk_mixed_effects.html">nested group repeated measures pseudobulk differential expression testing</a>
    <a class="dropdown-item" href="../articles/scglmmr_pseudobulk.html">nested group repeated measures pseudobulk differential expression testing</a>
    <a class="dropdown-item" href="../articles/scglmmr_singlecell.html">Single cell gene and module level differential expression with nested group repeated measures experiment designs using scglmmr</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MattPM/scglmmr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>nested group repeated measures pseudobulk differential expression testing</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MattPM/scglmmr/blob/HEAD/vignettes/scglmmr_pseudobulk.Rmd" class="external-link"><code>vignettes/scglmmr_pseudobulk.Rmd</code></a></small>
      <div class="d-none name"><code>scglmmr_pseudobulk.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="pseudobulk-mixed-effects-models">1. pseudobulk mixed effects models<a class="anchor" aria-label="anchor" href="#pseudobulk-mixed-effects-models"></a>
</h3>
<p>These functions implement convenience wrappers around both
<code>limma</code> for fitting traditional linear models (for
e.g. baseline differenced between groups) and the <code>dream</code>
method from the <code>variancePartition</code> package. This is the only
pseudobulk differential expression method that can accomodate both fixed
and random effects (which are statistically necessary to account for
non-independence of multiple measurements from the same donor) via use
of <code>lme4</code> for mixed effects modeling while accounting for
measurement uncertainty via pooling unqeual library sizes by
incorporating <code>voom</code> observational weights. This method
performed well in related simulation studies by <a href="https://www.nature.com/articles/s41467-020-19894-4" class="external-link">Crowell et.
al. 2020</a>. scglmmr performs statistical contrasts between groups
comparing the least squares means after accounting for model covariates
is empowered by the <a href="https://cran.r-project.org/web/packages/emmeans/index.html" class="external-link">emmeans</a>
package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MattPM/scglmmr" class="external-link">scglmmr</a></span><span class="op">)</span></code></pre></div>
<p>Below is a 2 group repeated measures experiment. A random intercept
model is required to account for autocorrelation in each donors post
perturbation response with their baseline level of expression. This data
can be accomodated by scglmmr. More simple experiment designs are also
supported, for example data with 2 groups but not repeated pre/post
treatment measurements, see the <code><a href="../reference/RunVoomLimma.html">RunVoomLimma()</a></code>
function.</p>
<table class="table">
<colgroup>
<col width="21%">
<col width="18%">
<col width="18%">
<col width="21%">
<col width="18%">
</colgroup>
<thead><tr class="header">
<th align="left">sample</th>
<th align="center">sampleid</th>
<th align="right">time</th>
<th align="left">group</th>
<th align="center">sex</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">101_t0</td>
<td align="center">101</td>
<td align="right">d0</td>
<td align="left">low</td>
<td align="center">F</td>
</tr>
<tr class="even">
<td align="left">101_t1</td>
<td align="center">101</td>
<td align="right">d1</td>
<td align="left">high</td>
<td align="center">F</td>
</tr>
<tr class="odd">
<td align="left">102_t0</td>
<td align="center">102</td>
<td align="right">d0</td>
<td align="left">low</td>
<td align="center">M</td>
</tr>
<tr class="even">
<td align="left">102_t1</td>
<td align="center">102</td>
<td align="right">d1</td>
<td align="left">high</td>
<td align="center">M</td>
</tr>
<tr class="odd">
<td align="left">… (x n donors)</td>
<td align="center">… (x n donors)</td>
<td align="right">… (x n donors)</td>
<td align="left">… (x n donors)</td>
<td align="center">… (x n donors)</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="pseudobulk-mixed-effects-model">1. “Pseudobulk” mixed effects model<a class="anchor" aria-label="anchor" href="#pseudobulk-mixed-effects-model"></a>
</h3>
<p>Fit a weighted mixed effects model to test group and time effects
using lme4 model formula syntax and using dream observational weights
using <code>dream</code> from the <code>variancePartition</code>
package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#devtools::install_github(repo = "https://github.com/MattPM/scglmmr")</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MattPM/scglmmr" class="external-link">scglmmr</a></span><span class="op">)</span>
<span class="va">datapath</span> <span class="op">=</span> <span class="st">"mypath/"</span>

<span class="co"># load seurat or sce object etc. </span>
<span class="va">s</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"path/seuratobject.rds"</span><span class="op">)</span>

<span class="co"># define counts and metadata and subset to cells above rm seurat object from workspace </span>
<span class="va">meta</span> <span class="op">=</span> <span class="va">s</span><span class="op">@</span><span class="va">meta.data</span>
<span class="va">umi</span> <span class="op">=</span> <span class="va">s</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span>
<span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">s</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/gc.html" class="external-link">gc</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># QC contingency of cells by subject for each celltype </span>
<span class="va">tab</span> <span class="op">=</span> <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="../reference/SubjectCelltypeTable.html">SubjectCelltypeTable</a></span><span class="op">(</span>metadata <span class="op">=</span> <span class="va">meta</span>, celltype_column <span class="op">=</span> <span class="st">"lineage"</span>, sample_column <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span>
<span class="va">tab</span><span class="op">$</span><span class="va">celltypes_remove</span>; <span class="va">tab</span><span class="op">$</span><span class="va">`low representation celltypes`</span>; <span class="va">tab</span><span class="op">$</span><span class="va">table</span>

<span class="co"># remove cells prior to pseudobulk analysis </span>
<span class="va">meta</span> <span class="op">=</span> <span class="va">meta</span><span class="op">[</span><span class="op">!</span><span class="va">meta</span><span class="op">$</span><span class="va">celltype_label_3</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">tab</span><span class="op">$</span><span class="va">celltypes_remove</span>, <span class="op">]</span>

<span class="co"># subset data </span>
<span class="va">umi</span> <span class="op">=</span> <span class="va">umi</span><span class="op">[</span> ,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span><span class="op">]</span>

<span class="co"># pseudobulk workflow </span>
<span class="va">pb</span> <span class="op">=</span> <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="../reference/PseudobulkList.html">PseudobulkList</a></span><span class="op">(</span>rawcounts <span class="op">=</span> <span class="va">umi</span>, metadata <span class="op">=</span> <span class="va">meta</span>, sample_col <span class="op">=</span> <span class="st">"sample"</span>,
                             celltype_col <span class="op">=</span> <span class="st">"lineage"</span>, avg_or_sum <span class="op">=</span> <span class="st">"sum"</span><span class="op">)</span>
<span class="va">designmat</span> <span class="op">=</span> <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="../reference/BulkDesignMatrix.html">BulkDesignMatrix</a></span><span class="op">(</span>metadata <span class="op">=</span> <span class="va">meta</span>, sample_column <span class="op">=</span> <span class="st">"sample"</span>,
                                      variable_column <span class="op">=</span> <span class="st">"cohort_timepoint"</span>, pseudobulklist <span class="op">=</span> <span class="va">pb</span><span class="op">)</span>
<span class="va">dge</span> <span class="op">=</span> <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="../reference/NormalizePseudobulk.html">NormalizePseudobulk</a></span><span class="op">(</span>pseudobulklist <span class="op">=</span> <span class="va">pb</span>, design_matrix <span class="op">=</span> <span class="va">designmat</span>, minimum.gene.count <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>

<span class="co"># custom a priori contrasts over a combined factor of the group and timepoint variables </span>
<span class="va">c_mat</span> <span class="op">=</span> <span class="fu">makeContrasts</span><span class="op">(</span>
  foldchange_difference <span class="op">=</span> <span class="op">(</span><span class="va">group_timepoint1_1</span> <span class="op">-</span> <span class="va">group_timepoint1_0</span><span class="op">)</span> <span class="op">-</span> <span class="op">(</span><span class="va">group_timepoint0_1</span> <span class="op">-</span> <span class="va">group_timepoint0_0</span><span class="op">)</span>,
  time1_foldchange <span class="op">=</span> <span class="op">(</span><span class="va">group_timepoint1_1</span> <span class="op">+</span> <span class="va">group_timepoint0_1</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span>  <span class="op">-</span> <span class="op">(</span><span class="va">group_timepoint1_0</span> <span class="op">+</span> <span class="va">group_timepoint0_0</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span>,
  baseline_groups <span class="op">=</span> <span class="op">(</span><span class="va">group_timepoint1_0</span> <span class="op">-</span> <span class="va">group_timepoint0_0</span><span class="op">)</span>,
  levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">designmat</span><span class="op">)</span>
<span class="op">)</span>
  
<span class="co"># fit mixed model for the multi timepoint contrasts </span>
<span class="va">fit</span> <span class="op">=</span> <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu">dreamMixedModel</span><span class="op">(</span>dge_lists <span class="op">=</span> <span class="va">dge</span>, 
                               apriori_contrasts <span class="op">=</span> <span class="cn">TRUE</span>, 
                               sample_column <span class="op">=</span> <span class="st">'sample'</span>,
                               cell_metadata <span class="op">=</span> <span class="va">meta</span>, 
                               contrast_matrix <span class="op">=</span> <span class="va">c_mat</span>, 
                               design_matrix <span class="op">=</span> <span class="va">designmat</span>, 
                               lme4_formula <span class="op">=</span>  <span class="st">'~ 0 + age + gender + cohort_timepoint + (1|sampleid)'</span>, 
                               fixed_effects <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'age'</span>, <span class="st">'gender'</span>, <span class="st">'cohort_timepoint'</span><span class="op">)</span>, 
                               plotsavepath <span class="op">=</span> <span class="va">figpath</span>, 
                               ncores <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>

<span class="co"># fit simple linear model for the baseline group level contrast </span>
<span class="va">bl</span> <span class="op">=</span> <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="../reference/RunVoomLimma.html">RunVoomLimma</a></span><span class="op">(</span>dgelists <span class="op">=</span> <span class="va">dge</span>, 
                           design_matrix <span class="op">=</span> <span class="va">designmat</span>, 
                           do_contrast_fit <span class="op">=</span> <span class="cn">T</span>,
                           my_contrast_matrix <span class="op">=</span> <span class="va">c_mat</span><span class="op">[</span> ,<span class="fl">3</span><span class="op">]</span><span class="op">)</span>

<span class="co"># save </span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">bl</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">datapath</span>, <span class="st">"blfit_object.rds"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">fit</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">datapath</span>, <span class="st">"fit_object.rds"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="downstream-analysis-on-pseudobulk-results">Downstream analysis on pseudobulk results<a class="anchor" aria-label="anchor" href="#downstream-analysis-on-pseudobulk-results"></a>
</h3>
<p>Here we provide examples of downstream analysis on results of the
model fit above focusing on the first contrast (the first coefficient in
the results), corresponding to:<br>
foldchange_difference = (group_timepoint1_1 - group_timepoint1_0) -
(group_timepoint0_1 - group_timepoint0_0)</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># hlmk = readRDS(file = here("signature_curation/hallmark.rds"))</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>figpath <span class="ot">=</span> <span class="st">"your/path"</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>test <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">GetRankResults</span>(<span class="at">limma.fit.object.list =</span> bl, <span class="at">coefficient.number =</span> <span class="dv">1</span>, <span class="st">"test"</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>res <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">GetContrastResults</span>(<span class="at">limma.fit.object.list =</span> bl, <span class="at">coefficient.number =</span> <span class="dv">1</span>, <span class="at">contrast.name =</span> <span class="st">"test"</span>)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="co"># get the mixed effects model results for the difference in fold changes </span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>fit_res <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">GetContrastResultsRaw</span>(<span class="at">limma.fit.object.list =</span> fit, </span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">coefficient.number =</span> <span class="dv">1</span>,</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">contrast.name =</span> <span class="st">"foldchangedifference"</span>)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>fit_rank <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">GetRankResultsRaw</span>(<span class="at">contrast.result.raw.list =</span> fit_res)</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene Set Enrichment Analysis </span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>gsea1 <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">RunFgseaOnRankList</span>(<span class="at">rank.list.celltype =</span> test, <span class="at">pathways =</span> hlmk)</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>d <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">RbindGseaResultList</span>(<span class="at">gsea_result_list =</span> gsea1,<span class="at">NES_filter =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">padj_filter =</span> <span class="fl">0.2</span>)</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>scglmmr<span class="sc">::</span><span class="fu">GSEABubblePlot</span>(d, <span class="at">save_path =</span> figpath, <span class="at">save_name =</span> <span class="st">"plot.pdf"</span>)</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="co"># also see scglmmr::GSEABarPlot()</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="co"># full set of leading edge genes vs celltypes </span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>lefull <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">GetLeadingEdgeFull</span>(<span class="at">gsea.list =</span> gsea1, <span class="at">padj.filter =</span> <span class="fl">0.1</span>,<span class="at">NES.filter =</span> <span class="sc">-</span><span class="cn">Inf</span>)</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="co"># combine GSEA results with model coefficient for each gene in leading edge </span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>cr <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">CombineResults</span>(<span class="at">gsealist =</span> gsea1, <span class="at">contrastlist =</span> res, <span class="at">gseafdr =</span> <span class="fl">0.1</span>, <span class="at">genefdr =</span> <span class="dv">1</span>)</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a><span class="co"># make tidy average data for visualization of weighted pb results </span></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>av <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">PseudobulkList</span>(<span class="at">rawcounts =</span> umi, </span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>                             <span class="at">metadata =</span> meta, </span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a>                             <span class="at">sample_col =</span> <span class="st">"sample"</span>, </span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a>                             <span class="at">celltype_col =</span> <span class="st">"celltype"</span>,</span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>                             <span class="at">avg_or_sum =</span> <span class="st">'average'</span>)</span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a>le_expr <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">LeadEdgeTidySampleExprs</span>(<span class="at">av.exprs.list =</span> av,</span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">gsea.list =</span> hlmk_ctm0, </span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">padj.filter =</span> <span class="fl">0.1</span>,</span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">NES.filter =</span> <span class="sc">-</span><span class="cn">Inf</span>)</span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a><span class="co"># example plot of sample level average leading edge genes annotated </span></span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true" tabindex="-1"></a>scglmmr<span class="sc">::</span><span class="fu">LeadEdgeSampleHeatmap</span>(<span class="at">tidy.exprs.list =</span> le_expr,</span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true" tabindex="-1"></a>                               <span class="at">modulename =</span> <span class="st">"MODULENAME"</span>,</span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true" tabindex="-1"></a>                               <span class="at">elltype_plot =</span> <span class="st">"TCELL"</span>,</span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true" tabindex="-1"></a>                               <span class="at">metadata =</span> meta, </span>
<span id="cb3-43"><a href="#cb3-43" aria-hidden="true" tabindex="-1"></a>                               <span class="at">metadata_annotate =</span> <span class="fu">c</span>(<span class="st">'group'</span>, <span class="st">'timepoint'</span>, <span class="st">'age'</span>, <span class="st">'gender'</span>),</span>
<span id="cb3-44"><a href="#cb3-44" aria-hidden="true" tabindex="-1"></a>                               <span class="at">sample_column =</span> <span class="st">'sample'</span>,</span>
<span id="cb3-45"><a href="#cb3-45" aria-hidden="true" tabindex="-1"></a>                               <span class="at">returnmat =</span> F, </span>
<span id="cb3-46"><a href="#cb3-46" aria-hidden="true" tabindex="-1"></a>                               <span class="at">savepath =</span> figpath, </span>
<span id="cb3-47"><a href="#cb3-47" aria-hidden="true" tabindex="-1"></a>                               <span class="at">savename =</span> <span class="st">"filename"</span>)</span>
<span id="cb3-48"><a href="#cb3-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-49"><a href="#cb3-49" aria-hidden="true" tabindex="-1"></a><span class="co"># see also: </span></span>
<span id="cb3-50"><a href="#cb3-50" aria-hidden="true" tabindex="-1"></a><span class="co"># scglmmr::TopGenesTidySampleExprs() - same as aobve for 'top' de genes estimated by model coeffcient / p value. </span></span>
<span id="cb3-51"><a href="#cb3-51" aria-hidden="true" tabindex="-1"></a><span class="co"># scglmmr::GetTidySummary() - for custom plotting</span></span>
<span id="cb3-52"><a href="#cb3-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-53"><a href="#cb3-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-54"><a href="#cb3-54" aria-hidden="true" tabindex="-1"></a><span class="co"># repeat this for all enriched pathways </span></span>
<span id="cb3-55"><a href="#cb3-55" aria-hidden="true" tabindex="-1"></a>heatpath <span class="ot">=</span> <span class="fu">here</span>(<span class="st">"sime/path"</span>); <span class="fu">dir.create</span>(heatpath)</span>
<span id="cb3-56"><a href="#cb3-56" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(le_expr)) {</span>
<span id="cb3-57"><a href="#cb3-57" aria-hidden="true" tabindex="-1"></a>  cdat <span class="ot">=</span> le_expr[[i]]</span>
<span id="cb3-58"><a href="#cb3-58" aria-hidden="true" tabindex="-1"></a>  ctype <span class="ot">=</span> <span class="fu">names</span>(le_expr[i])</span>
<span id="cb3-59"><a href="#cb3-59" aria-hidden="true" tabindex="-1"></a>  umod <span class="ot">=</span> <span class="fu">unique</span>(cdat<span class="sc">$</span>module)</span>
<span id="cb3-60"><a href="#cb3-60" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (u <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(umod)) {</span>
<span id="cb3-61"><a href="#cb3-61" aria-hidden="true" tabindex="-1"></a>    scglmmr<span class="sc">::</span><span class="fu">LeadEdgeSampleHeatmap</span>(<span class="at">tidy.exprs.list =</span> le_expr, </span>
<span id="cb3-62"><a href="#cb3-62" aria-hidden="true" tabindex="-1"></a>                          <span class="at">modulename =</span> umod[u], </span>
<span id="cb3-63"><a href="#cb3-63" aria-hidden="true" tabindex="-1"></a>                          <span class="at">celltype_plot =</span> ctype,</span>
<span id="cb3-64"><a href="#cb3-64" aria-hidden="true" tabindex="-1"></a>                          <span class="at">metadata =</span> meta, <span class="at">metadata_annotate =</span> <span class="fu">c</span>(<span class="st">'group'</span>, <span class="st">'timepoint'</span>, <span class="st">'age'</span>, <span class="st">'gender'</span>),</span>
<span id="cb3-65"><a href="#cb3-65" aria-hidden="true" tabindex="-1"></a>                          <span class="at">sample_column =</span> <span class="st">'sample'</span>,</span>
<span id="cb3-66"><a href="#cb3-66" aria-hidden="true" tabindex="-1"></a>                          <span class="at">returnmat =</span> F, </span>
<span id="cb3-67"><a href="#cb3-67" aria-hidden="true" tabindex="-1"></a>                          <span class="at">savepath =</span> heatpath,</span>
<span id="cb3-68"><a href="#cb3-68" aria-hidden="true" tabindex="-1"></a>                          <span class="at">savename =</span> <span class="fu">paste0</span>(ctype, <span class="st">" "</span>,umod[u],<span class="st">'.pdf'</span>))</span>
<span id="cb3-69"><a href="#cb3-69" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb3-70"><a href="#cb3-70" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb3-71"><a href="#cb3-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-72"><a href="#cb3-72" aria-hidden="true" tabindex="-1"></a><span class="co"># plot only the leading edge genes from enrichment in a heatmap of log Fold Change estimated for a model contrast </span></span>
<span id="cb3-73"><a href="#cb3-73" aria-hidden="true" tabindex="-1"></a>le <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">GetLeadingEdgeFull</span>(<span class="at">gsea.list =</span> gsea1, <span class="at">padj.filter =</span> <span class="fl">0.1</span>, <span class="at">NES.filter =</span> <span class="sc">-</span><span class="cn">Inf</span>)</span>
<span id="cb3-74"><a href="#cb3-74" aria-hidden="true" tabindex="-1"></a>genesub <span class="ot">=</span> <span class="fu">do.call</span>(rbind, le) <span class="sc">%$%</span> gene <span class="sc">%&gt;%</span> unique </span>
<span id="cb3-75"><a href="#cb3-75" aria-hidden="true" tabindex="-1"></a>mtx2 <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">GetGeneMatrix</span>(<span class="at">result.list =</span> res, </span>
<span id="cb3-76"><a href="#cb3-76" aria-hidden="true" tabindex="-1"></a>                    <span class="at">stat_for_matrix =</span> <span class="st">"logFC"</span>,</span>
<span id="cb3-77"><a href="#cb3-77" aria-hidden="true" tabindex="-1"></a>                    <span class="at">gene_subset =</span> genesub, </span>
<span id="cb3-78"><a href="#cb3-78" aria-hidden="true" tabindex="-1"></a>                    <span class="at">pvalfilter =</span> <span class="sc">-</span><span class="cn">Inf</span>, </span>
<span id="cb3-79"><a href="#cb3-79" aria-hidden="true" tabindex="-1"></a>                    <span class="at">logfcfilter =</span> <span class="fl">0.1</span>)</span>
<span id="cb3-80"><a href="#cb3-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-81"><a href="#cb3-81" aria-hidden="true" tabindex="-1"></a>pheatmap<span class="sc">::</span><span class="fu">pheatmap</span>(mtx2,</span>
<span id="cb3-82"><a href="#cb3-82" aria-hidden="true" tabindex="-1"></a>                   <span class="at">breaks =</span><span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">0</span>, <span class="at">to =</span> <span class="dv">2</span>,<span class="at">length.out =</span> <span class="dv">99</span>),</span>
<span id="cb3-83"><a href="#cb3-83" aria-hidden="true" tabindex="-1"></a>                   <span class="at">filename =</span> <span class="fu">paste0</span>(figpath,<span class="st">"LEgenes_heatmap.pdf"</span>))</span>
<span id="cb3-84"><a href="#cb3-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-85"><a href="#cb3-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-86"><a href="#cb3-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-87"><a href="#cb3-87" aria-hidden="true" tabindex="-1"></a><span class="do">### Hypergeometric erichment </span></span>
<span id="cb3-88"><a href="#cb3-88" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(termdf) <span class="co"># this term2gene dataframe is included in the package see clusterProfiler</span></span>
<span id="cb3-89"><a href="#cb3-89" aria-hidden="true" tabindex="-1"></a>hyp <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">RunHypergeometricTest</span>(<span class="at">result_list =</span> fit_res,</span>
<span id="cb3-90"><a href="#cb3-90" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">TERM2GENE_dataframe =</span> termdf,</span>
<span id="cb3-91"><a href="#cb3-91" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">pval_threshold =</span> <span class="fl">0.1</span>,</span>
<span id="cb3-92"><a href="#cb3-92" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">logFC_threshold =</span> <span class="dv">0</span>,</span>
<span id="cb3-93"><a href="#cb3-93" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">usefdr_threshold =</span> <span class="cn">FALSE</span>)</span>
<span id="cb3-94"><a href="#cb3-94" aria-hidden="true" tabindex="-1"></a><span class="co"># plot results </span></span>
<span id="cb3-95"><a href="#cb3-95" aria-hidden="true" tabindex="-1"></a>scglmmr<span class="sc">::</span><span class="fu">PlotHypergeometric</span>(<span class="at">hyperg_result =</span> hyp, </span>
<span id="cb3-96"><a href="#cb3-96" aria-hidden="true" tabindex="-1"></a>                            <span class="at">p.adjust.filter =</span> <span class="fl">0.1</span>,</span>
<span id="cb3-97"><a href="#cb3-97" aria-hidden="true" tabindex="-1"></a>                            <span class="at">genenumber_filter =</span> <span class="dv">2</span>,</span>
<span id="cb3-98"><a href="#cb3-98" aria-hidden="true" tabindex="-1"></a>                            <span class="at">savepath =</span> figpath,</span>
<span id="cb3-99"><a href="#cb3-99" aria-hidden="true" tabindex="-1"></a>                            <span class="at">savename =</span> <span class="st">"name"</span>, </span>
<span id="cb3-100"><a href="#cb3-100" aria-hidden="true" tabindex="-1"></a>                            <span class="at">title =</span> <span class="st">"title"</span>)</span>
<span id="cb3-101"><a href="#cb3-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-102"><a href="#cb3-102" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate average module z score across samples (z score across samples of gene z scores)</span></span>
<span id="cb3-103"><a href="#cb3-103" aria-hidden="true" tabindex="-1"></a><span class="co"># baseline_samples = c('') # user defined vector </span></span>
<span id="cb3-104"><a href="#cb3-104" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(av, <span class="cf">function</span>(x), x[ ,baseline_samples])</span>
<span id="cb3-105"><a href="#cb3-105" aria-hidden="true" tabindex="-1"></a>mz <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">AverageSampleModuleZscore</span>(<span class="at">average.metacell.list =</span> av,</span>
<span id="cb3-106"><a href="#cb3-106" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">module.list =</span> hlmk,<span class="at">use.module.subset =</span> F)</span></code></pre></div>
</div>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Matthew Mulè.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
