<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Single cell gene and module differential expression with nested group repeated measures experiment designs • scglmmr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Single cell gene and module differential expression with nested group repeated measures experiment designs">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scglmmr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/pseudobulk_mixed_effects.html">Pseudobulk differential expression with nested group repeated measures single cell experiment designs</a>
    </li>
    <li>
      <a href="../articles/scglmmr_singlecell.html">Single cell gene and module differential expression with nested group repeated measures experiment designs</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MattPM/scglmmr">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Single cell gene and module differential expression with
nested group repeated measures experiment designs</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MattPM/scglmmr/blob/master/vignettes/scglmmr_singlecell.Rmd"><code>vignettes/scglmmr_singlecell.Rmd</code></a></small>
      <div class="hidden name"><code>scglmmr_singlecell.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scglmmr)</span></code></pre></div>
<div id="single-cell-gene-module-level-testing-" class="section level3">
<h3 class="hasAnchor">
<a href="#single-cell-gene-module-level-testing-" class="anchor"></a>Single cell gene module level testing.</h3>
<p>Compare single cell level module score responses across groups. As in
the pseudobulk vignette, fit models within each cell type adjusting for
covariates using a random intercept term for each donor. Functions used
in this workflow include:</p>
<p><code>WeightedCellModuleScore</code> - score each module activity for
each cell; any method can be used here. Below we discuss relative vs
absolute scaling across cells vs within cell types for effect size
interpretation.</p>
<p><code>FitLmer</code> - fits a linear model using lme4 notation to the
module scores. Returns a nested list of model results that can be
queried to extract effects of interest.</p>
<p><code>FitLmerContrast</code> - as above, but specifically designed
for 2 group 2 timepoint repeated measures perturbation designs.
Automatically calculates baseline differences, treatment effect across
subjects and difference in treatment effects between groups using the <a href="https://cran.r-project.org/web/packages/emmeans/index.html">emmeans</a>
package.</p>
<p>A plot as shown below is generated for each module x cell type.</p>
<p><img src="../man/figures/example_sc.png" heightt="300" width="300/"></p>
</div>
<div id="load-and-normalize-single-cell-data" class="section level3">
<h3 class="hasAnchor">
<a href="#load-and-normalize-single-cell-data" class="anchor"></a>Load and normalize single cell data</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(tidyverse))</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(Seurat))</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(here))</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(scglmmr))</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co"># demonstrate using Seurat object</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>s <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">here</span>(<span class="st">"singlecellobject.rds"</span>))</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="co"># create a vector of celltypes </span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>ct <span class="ot">=</span> <span class="fu">unique</span>(s<span class="sc">@</span>meta.data<span class="sc">$</span>celltype)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="co"># normalize with log normalization</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>s <span class="ot">=</span> <span class="fu">NormalizeData</span>(s,<span class="at">normalization.method =</span> <span class="st">'LogNormalize'</span>,<span class="at">assay =</span> <span class="st">'RNA'</span>)</span></code></pre></div>
</div>
<div id="define-modules-to-fit" class="section level3">
<h3 class="hasAnchor">
<a href="#define-modules-to-fit" class="anchor"></a>Define modules to fit</h3>
<p>These functions can be useful to further interrogate the pathways
identified in the pseudobulk pipeline or for unbiased analysis. Here for
demonstration purposes, a subset of the hallmark pathways are used.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>sigs_test <span class="ot">=</span> hallmark[<span class="fu">c</span>(<span class="dv">22</span>, <span class="dv">23</span>, <span class="dv">24</span>,<span class="dv">25</span>,<span class="dv">26</span>,<span class="dv">27</span>)]</span></code></pre></div>
</div>
<div id="format-cell-metadata" class="section level3">
<h3 class="hasAnchor">
<a href="#format-cell-metadata" class="anchor"></a>Format cell metadata</h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># format cell metadata as factors group_id is order leveled for: </span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>md <span class="ot">=</span> s<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> </span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(timepoint <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">group_id =</span> <span class="fu">paste</span>(group, timepoint, <span class="at">sep =</span> <span class="st">'_'</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">group_id =</span> <span class="fu">factor</span>(group_id,  <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">'0_0'</span>, <span class="st">'0_1'</span>, <span class="st">'1_0'</span>, <span class="st">'1_1'</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(celltype, subjectid, sex , age, timepoint, group_id) <span class="sc">%&gt;%</span> </span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">age =</span> <span class="fu">as.numeric</span>(age)) <span class="sc">%&gt;%</span> </span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">droplevels</span>()</span></code></pre></div>
</div>
<div id="quality-control-cell-types-to-be-fit" class="section level3">
<h3 class="hasAnchor">
<a href="#quality-control-cell-types-to-be-fit" class="anchor"></a>Quality control cell types to be fit</h3>
<p>If a cell type is very donor specific such that there are only cells
for a couple of subjects we cant interpret the group level effects so we
remove these cell types.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># qc data to remove celltypes with no cells for some subhects at both timepoints </span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co"># keeps MLE more stable for the estimates of random intercept </span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>ct.si <span class="ot">=</span> <span class="fu">apply</span>(<span class="fu">table</span>(md<span class="sc">$</span>celltype, md<span class="sc">$</span>subjectid) , <span class="dv">1</span>, min) </span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>c.keep <span class="ot">=</span> <span class="fu">names</span>(ct.si[ct.si <span class="sc">&gt;</span> <span class="dv">7</span>])</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>md <span class="ot">=</span> md[md<span class="sc">$</span>celltype <span class="sc">%in%</span> c.keep, ]</span></code></pre></div>
</div>
<div id="calculate-module-scores" class="section level3">
<h3 class="hasAnchor">
<a href="#calculate-module-scores" class="anchor"></a>Calculate module scores</h3>
<p>This can be done in multiple ways. The function below calculates the
simple average. Note that for interpretability of the efects within each
cluster, the score is being applied and scaled within each cell
type.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co"># add single cell weighted module scores</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co"># split to standardize within cell type </span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>ct.md <span class="ot">=</span> <span class="fu">split</span>(md, <span class="at">f =</span> md<span class="sc">$</span>celltype)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co"># get umi data </span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>umi <span class="ot">=</span> s<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>data[ ,<span class="fu">rownames</span>(md)]</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="co"># fit module scores </span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>mod_scores <span class="ot">=</span> <span class="fu">lapply</span>(ct.md, <span class="cf">function</span>(x){ </span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>  scglmmr<span class="sc">::</span><span class="fu">WeightedCellModuleScore</span>(<span class="at">gene_matrix =</span> umi[ ,<span class="fu">rownames</span>(x)], </span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">module_list =</span> sigs_test, </span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">threshold =</span> <span class="dv">0</span>,</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>                                   <span class="co"># standardize within protein celltype</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">cellwise_scaling =</span> <span class="cn">TRUE</span>, </span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">return_weighted =</span> <span class="cn">FALSE</span> )</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>ms <span class="ot">=</span> <span class="fu">bind_rows</span>(mod_scores)</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a><span class="co"># correctly order rows after the split. </span></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>ms <span class="ot">=</span> ms[<span class="fu">match</span>(<span class="at">x =</span> <span class="fu">rownames</span>(md), <span class="at">table =</span> <span class="fu">rownames</span>(ms)), ]</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">all.equal</span>(<span class="fu">rownames</span>(ms), <span class="fu">rownames</span>(md)))</span></code></pre></div>
</div>
<div id="fit-modules" class="section level3">
<h3 class="hasAnchor">
<a href="#fit-modules" class="anchor"></a>Fit modules</h3>
<p>Specify a model using lme4 notation and fit models. Note that the LHS
outcome “modulescore” should not be changed as this is adaptively
applied to each module internally.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co"># specify model </span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>f1 <span class="ot">=</span> <span class="st">'modulescore ~ 0 + timepoint + age + sex + (1|subjectid)'</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co"># specify a path to save plots </span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>plot_savepath <span class="ot">=</span> <span class="fu">here</span>(<span class="st">'/mypath/'</span>)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="co"># fit sc mod mixed model on ewighted module scores. </span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>mmres <span class="ot">=</span> <span class="fu">FitLmer</span>(<span class="at">module_data_frame =</span> ms,</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>                <span class="at">celltype_column =</span> <span class="st">'celltype'</span>,</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>                <span class="at">metadata =</span> md, </span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>                <span class="at">lme4.formula =</span> f1)</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>  </span></code></pre></div>
</div>
<div id="examine-results" class="section level3">
<h3 class="hasAnchor">
<a href="#examine-results" class="anchor"></a>Examine results</h3>
<p>Results are formatted as a list indexed by cell type. If we want to
find the effects of time on expression we can use the following code to
extract out this effect as shown below using the emmeans package. This
allows maximum flexibility since any effect can be extracted and custom
contrast matrices can be utilized – see emmeans documentation.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co"># write wrapper to extract out time effect.</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>extract.time.effect <span class="ot">=</span> <span class="cf">function</span>(fit.list.celltype){ </span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  c_res <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(u <span class="cf">in</span> <span class="fu">length</span>(fit.list.celltype)){ </span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    emm1 <span class="ot">=</span> <span class="fu">tryCatch</span>(</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>        emmeans<span class="sc">::</span><span class="fu">emmeans</span>(<span class="at">object =</span> m1, <span class="at">specs =</span> <span class="sc">~</span> timepoint , <span class="at">data =</span> res, <span class="at">lmer.df =</span> <span class="st">"asymptotic"</span>),</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">error =</span> <span class="cf">function</span>(e) <span class="fu">return</span>(<span class="cn">NA</span>)</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>    c_res[[u]] <span class="ot">=</span> emm1</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a><span class="co"># now we can apply this to each cell type. </span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>time.result <span class="ot">=</span> <span class="fu">lapply</span>(mmres , extract.time.effect)</span></code></pre></div>
</div>
<div id="a-custom-function-for-automatically-extracting-effects-from-a-2-group-repeated-measures-treatment-design" class="section level3">
<h3 class="hasAnchor">
<a href="#a-custom-function-for-automatically-extracting-effects-from-a-2-group-repeated-measures-treatment-design" class="anchor"></a>a custom function for automatically extracting effects from a 2
group repeated measures treatment design</h3>
<p>This function was written for a 2 group design (as shown in the main
readme) where subjects each have a pre and post perturbation
measurement. WE wantto compare the pertuebation effects across subjects,
and between groups and test the baseline differences. All of these
effects are extracted automatically into a simple output, and plots are
created as above.<br>
For this function, on the RHS, + (1|subjectid) should also not be
changed. One can also specify fixed effects covariates as a vector.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>f1 <span class="ot">=</span> <span class="st">'modulescore ~ 0 + groupid + age + sex + (1|subjectid)'</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># fit sc mod mixed model on ewighted module scores. </span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>mm_res.m1 <span class="ot">=</span> <span class="fu">FitLmerContrast</span>(<span class="at">module_data_frame =</span> ms, </span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>                            <span class="at">celltype_column =</span> <span class="st">'celltype'</span>, </span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>                            <span class="at">metadata =</span> md, </span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>                            <span class="at">lmer_formula =</span> f1, </span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>                            <span class="at">plotdatqc =</span> <span class="cn">TRUE</span>, </span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>                            <span class="at">fixed_effects =</span> <span class="cn">NULL</span>,</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>                            <span class="at">figpath =</span> plot_savepath)</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="co"># saveRDS(mm_res.m1,file = paste0(datapath, "mm_res.m1.rds"))</span></span></code></pre></div>
<p>Note that the effect size of specific module scores may be compared
using the code below to determine which celltype has the greatest
<em>absolute</em> effect size (epression) of a given module. Both the
between and within cell type expression group level differences are
likely of interest for interpretation.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>ms.full <span class="ot">=</span> scglmmr<span class="sc">::</span><span class="fu">WeightedCellModuleScore</span>(</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_matrix =</span> umi, </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">module_list =</span> sigs_test, </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">threshold =</span> <span class="dv">0</span>,</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># standardize within protein celltype</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">cellwise_scaling =</span> <span class="cn">TRUE</span>, </span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">return_weighted =</span> <span class="cn">FALSE</span> </span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>d <span class="ot">=</span> <span class="fu">cbind</span>(ms.full, md)</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="co"># absolute effect size for hallmark </span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>f1 <span class="ot">=</span> <span class="fu">as.formula</span>(HALLMARK_HYPOXIA <span class="sc">~</span> <span class="dv">0</span> <span class="sc">+</span> celltype <span class="sc">+</span> age <span class="sc">+</span> sex <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>subjecctID))</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>m1 <span class="ot">=</span> lme4<span class="sc">::</span><span class="fu">lmer</span>(f1, <span class="at">data =</span> bl)</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>emm1 <span class="ot">=</span> emmeans<span class="sc">::</span><span class="fu">emmeans</span>(m1, <span class="at">specs =</span> <span class="sc">~</span>celltype)</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(emm1)</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Matthew Mulè.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
