<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>AggregateCellMetadata - take cell level metadata and collapse down to sample-level
metadata for use in pseudobuk testing. This can be useful if you do not already have
metadata for each sample in the experiment, but these data are stored in single cell
metadata. For example, metadata for the donor ID, unique sample
(e.g. donorid_timepoint), age, sex etc. Note these are all individual sample level data.
Single cell intrinsic variables like nUMI cannot be collapsed down, only variables that
are unique for each sample which are the columns of the pseudobulk data. Only include
metadata that you intend to adjust models for as covariates or random effects because all
variables will be referenced during count normalization and feature filtering. — AggregateCellMetadata • scglmmr</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="AggregateCellMetadata - take cell level metadata and collapse down to sample-level
metadata for use in pseudobuk testing. This can be useful if you do not already have
metadata for each sample in the experiment, but these data are stored in single cell
metadata. For example, metadata for the donor ID, unique sample
(e.g. donorid_timepoint), age, sex etc. Note these are all individual sample level data.
Single cell intrinsic variables like nUMI cannot be collapsed down, only variables that
are unique for each sample which are the columns of the pseudobulk data. Only include
metadata that you intend to adjust models for as covariates or random effects because all
variables will be referenced during count normalization and feature filtering. — AggregateCellMetadata" />

<meta property="og:description" content="AggregateCellMetadata - take cell level metadata and collapse down to sample-level
metadata for use in pseudobuk testing. This can be useful if you do not already have
metadata for each sample in the experiment, but these data are stored in single cell
metadata. For example, metadata for the donor ID, unique sample
(e.g. donorid_timepoint), age, sex etc. Note these are all individual sample level data.
Single cell intrinsic variables like nUMI cannot be collapsed down, only variables that
are unique for each sample which are the columns of the pseudobulk data. Only include
metadata that you intend to adjust models for as covariates or random effects because all
variables will be referenced during count normalization and feature filtering." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scglmmr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/pseudobulk_mixed_effects.html">Pseudobulk differential expression with nested group repeated measures single cell experiment designs</a>
    </li>
    <li>
      <a href="../articles/scglmmr_singlecell.html">Single cell gene and module differential expression with nested group repeated measures experiment designs</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/MattPM/scglmmr">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>AggregateCellMetadata - take cell level metadata and collapse down to sample-level
metadata for use in pseudobuk testing. This can be useful if you do not already have
metadata for each sample in the experiment, but these data are stored in single cell
metadata. For example, metadata for the donor ID, unique sample
(e.g. donorid_timepoint), age, sex etc. Note these are all individual sample level data.
Single cell intrinsic variables like nUMI cannot be collapsed down, only variables that
are unique for each sample which are the columns of the pseudobulk data. Only include
metadata that you intend to adjust models for as covariates or random effects because all
variables will be referenced during count normalization and feature filtering.</h1>
    <small class="dont-index">Source: <a href='https://github.com/MattPM/scglmmr/blob/master/R/pseudobulk_helpers.r'><code>R/pseudobulk_helpers.r</code></a></small>
    <div class="hidden name"><code>AggregateCellMetadata.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>AggregateCellMetadata - take cell level metadata and collapse down to sample-level
metadata for use in pseudobuk testing. This can be useful if you do not already have
metadata for each sample in the experiment, but these data are stored in single cell
metadata. For example, metadata for the donor ID, unique sample
(e.g. donorid_timepoint), age, sex etc. Note these are all individual sample level data.
Single cell intrinsic variables like nUMI cannot be collapsed down, only variables that
are unique for each sample which are the columns of the pseudobulk data. Only include
metadata that you intend to adjust models for as covariates or random effects because all
variables will be referenced during count normalization and feature filtering.</p>
    
    </div>

    <pre class="usage"><span class='fu'>AggregateCellMetadata</span>(
  <span class='no'>cell.metadata</span>,
  <span class='no'>sample_column</span>,
  <span class='no'>variable_columns</span>,
  <span class='no'>pseudobulk.List</span>
)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>cell.metadata</th>
      <td><p>dataframe of meta data for cells-rows as columns i.e. ColData or
Seurat@meta.data.</p></td>
    </tr>
    <tr>
      <th>sample_column</th>
      <td><p>quoted character e.g. "sample". This should indicate the variable corresponding
to the rows of the pseudobulk gene expression data.</p></td>
    </tr>
    <tr>
      <th>variable_columns</th>
      <td><p>experiment variables coresponding to sample level data. For example:
c('SubjectID', 'timepoint', 'sex', 'age').</p></td>
    </tr>
    <tr>
      <th>pseudobulklist</th>
      <td><p>the object output from PseudobulkList. used to check the columns of the aggregated
metadata match the columns of the elements of the Pseudobulk list.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Aggregated metadata</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='no'>samplemd</span> <span class='kw'>=</span> <span class='fu'>AggregateCellMetadata</span>(<span class='kw'>cell.metadata</span> <span class='kw'>=</span> <span class='no'>s</span>@<span class='kw'>meta.data</span>, <span class='kw'>sample_column</span> <span class='kw'>=</span> <span class='st'>'sample'</span>, <span class='kw'>variable_columns</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>'subjectID'</span>, <span class='st'>'timepoint'</span>, <span class='st'>'age'</span>, <span class='st'>'sex'</span>), <span class='kw'>pseudobulk.List</span> <span class='kw'>=</span> <span class='no'>pb</span>)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Matthew Mulè.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


