<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="FitLmer - This is a simpler version of FitLmerContrast that makes no assumptions about structure of underlying
 data except that it can accomodate a mixed effects model formula and that there are multiple cell types to be
  separately fitted. fit a single cell mixed effects linear model. Designed to fit an aggregated gene module score.
  Fits a model to each cell cluster to test a grouping, treatment or combination factor; returns model fits
  for maximum flexibility"><title>FitLmer - This is a simpler version of FitLmerContrast that makes no assumptions about structure of underlying
 data except that it can accomodate a mixed effects model formula and that there are multiple cell types to be
  separately fitted. fit a single cell mixed effects linear model. Designed to fit an aggregated gene module score.
  Fits a model to each cell cluster to test a grouping, treatment or combination factor; returns model fits
  for maximum flexibility — FitLmer • scglmmr</title><script src="../deps/jquery-3.5.1/jquery-3.5.1.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="FitLmer - This is a simpler version of FitLmerContrast that makes no assumptions about structure of underlying
 data except that it can accomodate a mixed effects model formula and that there are multiple cell types to be
  separately fitted. fit a single cell mixed effects linear model. Designed to fit an aggregated gene module score.
  Fits a model to each cell cluster to test a grouping, treatment or combination factor; returns model fits
  for maximum flexibility — FitLmer"><meta property="og:description" content="FitLmer - This is a simpler version of FitLmerContrast that makes no assumptions about structure of underlying
 data except that it can accomodate a mixed effects model formula and that there are multiple cell types to be
  separately fitted. fit a single cell mixed effects linear model. Designed to fit an aggregated gene module score.
  Fits a model to each cell cluster to test a grouping, treatment or combination factor; returns model fits
  for maximum flexibility"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scglmmr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/pseudobulk_mixed_effects.html">Pseudobulk differential expression with nested group repeated measures single cell experiment designs</a>
    <a class="dropdown-item" href="../articles/scglmmr_singlecell.html">Single cell gene and module differential expression with nested group repeated measures experiment designs</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MattPM/scglmmr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>FitLmer - This is a simpler version of FitLmerContrast that makes no assumptions about structure of underlying
 data except that it can accomodate a mixed effects model formula and that there are multiple cell types to be
  separately fitted. fit a single cell mixed effects linear model. Designed to fit an aggregated gene module score.
  Fits a model to each cell cluster to test a grouping, treatment or combination factor; returns model fits
  for maximum flexibility</h1>
      <small class="dont-index">Source: <a href="https://github.com/MattPM/scglmmr/blob/HEAD/R/SCGroupContrast_modulefunctions.r" class="external-link"><code>R/SCGroupContrast_modulefunctions.r</code></a></small>
      <div class="d-none name"><code>FitLmer.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>FitLmer - This is a simpler version of FitLmerContrast that makes no assumptions about structure of underlying
 data except that it can accomodate a mixed effects model formula and that there are multiple cell types to be
  separately fitted. fit a single cell mixed effects linear model. Designed to fit an aggregated gene module score.
  Fits a model to each cell cluster to test a grouping, treatment or combination factor; returns model fits
  for maximum flexibility</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">FitLmer</span><span class="op">(</span>
  <span class="va">module_data_frame</span>,
  celltype_column <span class="op">=</span> <span class="st">"celltype"</span>,
  <span class="va">metadata</span>,
  lme4.formula <span class="op">=</span> <span class="cn">NULL</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>module_data_frame</dt>
<dd><p>data for each cell to model -- designed to be scores for modules (columns) returned by scglmmr::WeightedModuleScore</p></dd>
<dt>celltype_column</dt>
<dd><p>the column in metadata with the cluster / celltype designation for each cell</p></dd>
<dt>lme4metadata</dt>
<dd><p>metadata for model fit</p></dd>
<dt>f1</dt>
<dd><p>model furmula</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in"><span class="va">s</span> <span class="op">=</span> <span class="fu">SubsetData</span><span class="op">(</span><span class="va">s</span>, ident.use <span class="op">=</span> <span class="va">tc</span><span class="op">)</span></span>
<span class="r-in"><span class="va">s</span> <span class="op">=</span> <span class="fu">NormalizeData</span><span class="op">(</span><span class="va">s</span>,normalization.method <span class="op">=</span> <span class="st">'LogNormalize'</span>,assay <span class="op">=</span> <span class="st">'RNA'</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># format metadata as factors group_id is order leveled for:</span></span>
<span class="r-in"><span class="co"># contrast_fit = contrast(emm1, method = list( (c21 - c20) - (c11 - c10) ))</span></span>
<span class="r-in"><span class="va">md</span> <span class="op">=</span> <span class="va">s</span><span class="op">@</span><span class="va">meta.data</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">timepoint</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu">mutate</span><span class="op">(</span>group_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">group</span>, <span class="va">timepoint</span>, sep <span class="op">=</span> <span class="st">'_'</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu">mutate</span><span class="op">(</span>group_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">group_id</span>,  levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'0_0'</span>, <span class="st">'0_1'</span>, <span class="st">'1_0'</span>, <span class="st">'1_1'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu">mutate</span><span class="op">(</span>subjectid <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">sampleid</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu">select</span><span class="op">(</span><span class="va">celltype</span>, <span class="va">subjectid</span>, <span class="va">age</span>, <span class="va">group_id</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu">mutate</span><span class="op">(</span>age <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">age</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu"><a href="https://rdrr.io/r/base/droplevels.html" class="external-link">droplevels</a></span><span class="op">(</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># qc data to remove celltypes with no cells for some subhects at both timepoints</span></span>
<span class="r-in"><span class="co"># keeps MLE more stable for the estimates of random intercept</span></span>
<span class="r-in"><span class="va">ct.si</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">md</span><span class="op">$</span><span class="va">celltype</span>, <span class="va">md</span><span class="op">$</span><span class="va">subjectid</span><span class="op">)</span> , <span class="fl">1</span>, <span class="va">min</span><span class="op">)</span></span>
<span class="r-in"><span class="va">c.keep</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ct.si</span><span class="op">[</span><span class="va">ct.si</span> <span class="op">&gt;</span> <span class="fl">7</span><span class="op">]</span><span class="op">)</span></span>
<span class="r-in"><span class="va">md</span> <span class="op">=</span> <span class="va">md</span><span class="op">[</span><span class="va">md</span><span class="op">$</span><span class="va">celltype</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">c.keep</span>, <span class="op">]</span></span>
<span class="r-in"></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># add single cell weighted module scores</span></span>
<span class="r-in"><span class="co"># split to standardize within cell type</span></span>
<span class="r-in"><span class="va">ct.md</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">md</span>, f <span class="op">=</span> <span class="va">md</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span class="r-in"><span class="va">mod_scores</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">ct.md</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span class="r-in">  <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="WeightedCellModuleScore.html">WeightedCellModuleScore</a></span><span class="op">(</span>gene_matrix <span class="op">=</span> <span class="va">s</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">data</span><span class="op">[</span> ,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">]</span>,</span>
<span class="r-in">                                   module_list <span class="op">=</span> <span class="va">mtor.sigs</span>,</span>
<span class="r-in">                                   threshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span class="r-in">                                   <span class="co"># standardize within protein celltype</span></span>
<span class="r-in">                                   cellwise_scaling <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span class="r-in">                                   return_weighted <span class="op">=</span> <span class="cn">FALSE</span> <span class="op">)</span></span>
<span class="r-in"><span class="op">}</span><span class="op">)</span></span>
<span class="r-in"><span class="va">ms</span> <span class="op">=</span> <span class="fu">bind_rows</span><span class="op">(</span><span class="va">mod_scores</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># correctly order rows after the split.</span></span>
<span class="r-in"><span class="va">ms</span> <span class="op">=</span> <span class="va">ms</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">md</span><span class="op">)</span>, table <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ms</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ms</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">md</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># specify model</span></span>
<span class="r-in"><span class="va">f1</span> <span class="op">=</span> <span class="st">'modulescore ~ group_id + age + (1|subjectid)'</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># fit sc mod mixed model on ewighted module scores.</span></span>
<span class="r-in"><span class="va">mm_res.m1</span> <span class="op">=</span> <span class="fu">scglmmr</span><span class="fu">::</span><span class="fu"><a href="FitLmerContrast.html">FitLmerContrast</a></span><span class="op">(</span>module_data_frame <span class="op">=</span> <span class="va">ms</span>,</span>
<span class="r-in">                                     celltype_column <span class="op">=</span> <span class="st">'celltype'</span>,</span>
<span class="r-in">                                     metadata <span class="op">=</span> <span class="va">md</span>,</span>
<span class="r-in">                                     lmer_formula <span class="op">=</span> <span class="va">f1</span>,</span>
<span class="r-in">                                     plotdatqc <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span class="r-in">                                     fixed_effects <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span class="r-in">                                     figpath <span class="op">=</span> <span class="va">plot_savepath</span><span class="op">)</span></span>
<span class="r-in"><span class="op">}</span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Matthew Mulè.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer></div>

  

  

  </body></html>

