<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference • scglmmr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scglmmr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/scglmmr_pseudobulk.html">Pseudobulk differential expression with nested group repeated measures single cell experiment designs using scglmmr</a>
    </li>
    <li>
      <a href="../articles/scglmmr_singlecell.html">Single cell gene and module level differential expression with nested group repeated measures experiment designs using scglmmr</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/MattPM/scglmmr/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="AverageSampleModuleZscore.html">AverageSampleModuleZscore()</a></code> </p>
        </td>
        <td><p>AverageSampleModuleZscore apply the function  calc_avg_module_zscore to a pseudobulklist.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="BulkDesignMatrix.html">BulkDesignMatrix()</a></code> </p>
        </td>
        <td><p>BulkDesignMatrix - designmatrix for the main experiment factor</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="CombineResults.html">CombineResults()</a></code> </p>
        </td>
        <td><p>CombineResults - For all cell types merge the gsea leading edge genes with their contrast model coefficieint and p value from limma / dream</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GSEABarPlot.html">GSEABarPlot()</a></code> </p>
        </td>
        <td><p>GSEABarPlot - plot gsea results for a single cell type</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GSEABubblePlot.html">GSEABubblePlot()</a></code> </p>
        </td>
        <td><p>GSEABubblePlot plot gsea results for all cell types</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GetContrastResults.html">GetContrastResults()</a></code> </p>
        </td>
        <td><p>GetContrastResults - return results from a contrast fit on list of celltypes from RunVoomLimma using topTable</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GetContrastResultsRaw.html">GetContrastResultsRaw()</a></code> </p>
        </td>
        <td><p>GetContrastResultsRaw - calculate p values and return contrast results from modelfit with dreamMixedModel</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GetGeneMatrix.html">GetGeneMatrix()</a></code> </p>
        </td>
        <td><p>GetGeneMatrix get a gene matric for plotting genes by celltypes statistic from pseudobulk model results e.g. using heatmap pheatmap or complexheatmap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GetLeadingEdgeFull.html">GetLeadingEdgeFull()</a></code> </p>
        </td>
        <td><p>GetLeadingEdgeFull Get a tidy dataframe of ALL Leading Edge Genes from gene set enrichment for all cell types</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GetLeadingEdgeGenes.html">GetLeadingEdgeGenes()</a></code> </p>
        </td>
        <td><p>GetLeadingEdgeGenes get the leading edge genes from a single cell type / module combination</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GetRankResults.html">GetRankResults()</a></code> </p>
        </td>
        <td><p>GetRankResults get list of gene ranks by t stat for fGSEA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GetRankResultsRaw.html">GetRankResultsRaw()</a></code> </p>
        </td>
        <td><p>GetRankResultsRaw get list of gene ranks by t stat for fGSEA. for results returned by dreamMixedModel</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GetTidySummary.html">GetTidySummary()</a></code> </p>
        </td>
        <td><p>GetTidySummary - tidy data summary for a single cell type</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="LeadEdgeSampleHeatmap.html">LeadEdgeSampleHeatmap()</a></code> </p>
        </td>
        <td><p>LeadEdgeSampleHeatmap make a heatmap of average expression for top or leading edge genes returned by LeadEdgeTidySampleExprs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="LeadEdgeTidySampleExprs.html">LeadEdgeTidySampleExprs()</a></code> </p>
        </td>
        <td><p>LeadEdgeTidySampleExprs - convert a PseudobulkList into a tidy dataframe for each sample across cell types of the leading edge genes from a gsea list</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="NormalizePseudobulk.html">NormalizePseudobulk()</a></code> </p>
        </td>
        <td><p>NormalizePseudobulk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PlotGeneDistCohort.html">PlotGeneDistCohort()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PlotHypergeometric.html">PlotHypergeometric()</a></code> </p>
        </td>
        <td><p>PlotHypergeometric - plot results returned by RunHypergeometricTest</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PseudobulkList.html">PseudobulkList()</a></code> </p>
        </td>
        <td><p>PseudobulkList - make a pseudobuk summed or average dataset for each samplexcelltype</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="RbindGseaResultList.html">RbindGseaResultList()</a></code> </p>
        </td>
        <td><p>RbindGseaResultList - prepare gsea result list for visualization funcitons GseaBubblePlot or GseaBarPlot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="RunFgseaOnRankList.html">RunFgseaOnRankList()</a></code> </p>
        </td>
        <td><p>RunFgseaOnRankList - wrapper around fast gene set enrichment analysis with the fgsea R package https://bioconductor.org/packages/release/bioc/html/fgsea.html</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="RunHypergeometricTest.html">RunHypergeometricTest()</a></code> </p>
        </td>
        <td><p>RunHypergeometricTest - run a hypergeometric test on results returned by GetContrastResults or GetContrastResultsRaw</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="RunVoomLimma.html">RunVoomLimma()</a></code> </p>
        </td>
        <td><p>RunVoomLimma</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="RunVoomLimma_covar.html">RunVoomLimma_covar()</a></code> </p>
        </td>
        <td><p>RunVoomLimma_covar RunLimmaVoom with an option to adjust for a covariate within mRNA.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="SCGroupContrastGLMM.html">SCGroupContrastGLMM()</a></code> </p>
        </td>
        <td><p>SCGroupContrastGLMM - within each cell type contrast difference in fold change between groups, baseline difference, and fold change across groups of module scores.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="SCMixedPoisson.html">SCMixedPoisson()</a></code> </p>
        </td>
        <td><p>SCMixedPoisson</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="SubjectCelltypeTable.html">SubjectCelltypeTable()</a></code> </p>
        </td>
        <td><p>SubjectCelltypeTable - make a QC contingency table of cells by samples beore making pseudobulk data.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="TopGenesTidySampleExprs.html">TopGenesTidySampleExprs()</a></code> </p>
        </td>
        <td><p>TopGenesTidySampleExprs convert a PseudobulkList into a tidy dataframe for each sample across cell types of the top differentially expressed genes for a contrast</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="WeightedCellModuleScore.html">WeightedCellModuleScore()</a></code> </p>
        </td>
        <td><p>WeightedCellModuleScore - calculate the average or weighted aveage +/- scaling return dataframe of cells by average module score</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="WilcoxWithinCluster.html">WilcoxWithinCluster()</a></code> </p>
        </td>
        <td><p>WilcoxWithinCluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calc_avg_module_zscore.html">calc_avg_module_zscore()</a></code> </p>
        </td>
        <td><p>calc_avg_module_zscore calculate average module z score of list of modules on a PseudobulkList</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dreamMixedModel.html">dreamMixedModel()</a></code> </p>
        </td>
        <td><p>dreamMixedModel - run dream mixed model
note due to this issue must require variancepartition https://github.com/GabrielHoffman/variancePartition/issues/17</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Matthew Mulè.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


