<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>SCMixedPoisson - Fit a gene level Poisson count mixed generalized linear model. — SCMixedPoisson • scglmmr</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="SCMixedPoisson - Fit a gene level Poisson count mixed generalized linear model. — SCMixedPoisson"><meta property="og:description" content="SCMixedPoisson - Fit a gene level Poisson count mixed generalized linear model."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scglmmr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/pseudobulk_mixed_effects.html">Pseudobulk differential expression with nested group repeated measures single cell experiment design</a>
    </li>
    <li>
      <a href="../articles/scglmmr_singlecell.html">Single cell gene and module differential expression with nested group repeated measures experiment designs</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/MattPM/scglmmr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>SCMixedPoisson - Fit a gene level Poisson count mixed generalized linear model.</h1>
    <small class="dont-index">Source: <a href="https://github.com/MattPM/scglmmr/blob/HEAD/R/SCMixedPoisson.r" class="external-link"><code>R/SCMixedPoisson.r</code></a></small>
    <div class="hidden name"><code>SCMixedPoisson.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>SCMixedPoisson - Fit a gene level Poisson count mixed generalized linear model.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">SCMixedPoisson</span><span class="op">(</span>
  <span class="va">gene_data</span>,
  <span class="va">metadata</span>,
  model_formula <span class="op">=</span> <span class="st">"gene ~ offset(log(nUMI)) + timepoint + (1|subjectid)"</span>,
  test_variable <span class="op">=</span> <span class="cn">NULL</span>,
  covariate_variables <span class="op">=</span> <span class="cn">NULL</span>,
  celltype_genes_test <span class="op">=</span> <span class="cn">NULL</span>,
  <span class="va">save_path</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>gene_data</dt>
<dd><p>matrix of raw single cell UMI counts genes as COLUMNS and cells as ROWS; i.e. a transpose t() of Bioconductor or Seurat raw.data slots if genes_test is specified, columns must contain all genes from genes_test i.e. umi_matrix = t(seurat_raw_counts[unique(unlist((genes_test)), ])</p></dd>
<dt>metadata</dt>
<dd><p>metadata for each cell used for fitting model</p></dd>
<dt>model_formula</dt>
<dd><p>example: 'gene ~ offset(log(nUMI)) + timepoint + (1|subjectid)' the 'gene~' and '(1|subjectid)' variables should be kept the same. The variable 'timepoint' can be altered to be any treatment or perturbation effect being tested. User can also specify 'covariate_variables' to make formula gene~offset(nUMI) + covariate1 + covariate2 + test_variable + (1|subjectid) automatically.</p></dd>
<dt>test_variable</dt>
<dd><p>the column of metadata coding the perturbation variable. e.g. 'timepoint' for formula: gene ~ timepoint + (1|subjectid)'</p></dd>
<dt>covariate_variables</dt>
<dd><p>a vector of variables in metadata to add to the model as covariates; this is only used if model_formula is NULL; otherwise speify directly</p></dd>
<dt>celltype_genes_test</dt>
<dd><p>An R list indexed by cell type: the subset of genes to test for each cell type.</p></dd>
<dt>save_path</dt>
<dd><p>file path to save intermediate results for each cell type.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>a results matrix for all genes across all cell types tested.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="va">results</span> <span class="op">=</span> <span class="fu">SCMixedPoisson</span><span class="op">(</span>gene_data <span class="op">=</span> <span class="va">gene_data</span>,</span>
<span class="r-in">  lme4metadata <span class="op">=</span> <span class="va">meta_data</span>,</span>
<span class="r-in">  model_formula <span class="op">=</span> <span class="st">'gene ~ offset(log(nUMI)) + timepoint + (1|sampleid)'</span>,</span>
<span class="r-in">  celltype_genes_test <span class="op">=</span> <span class="va">celltype_indexed_gene_vector</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in SCMixedPoisson(gene_data = gene_data, lme4metadata = meta_data,     model_formula = "gene ~ offset(log(nUMI)) + timepoint + (1|sampleid)",     celltype_genes_test = celltype_indexed_gene_vector):</span> unused argument (lme4metadata = meta_data)</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Matthew Mulè.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer></div>

  


  

  </body></html>

